{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 03\n",
    "# Estimate significant epistatic gene-gene interactions contributing to AMR phenotypes \n",
    "1. Filter out initial set of gene-gene interactions that were generated by `02_ML_ensemble.ipynb`\n",
    "2. Build gene-gene logistic-regression models of the filtered gene-gene interactions and identify significant epistatic interactions\n",
    "2. Generate figures that intuitively describe the identified epistatic interactions\n",
    "\n",
    "### Load _M. tuberculosis_ pan-genome dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/anaconda2/lib/python2.7/site-packages/statsmodels/compat/pandas.py:56: FutureWarning: The pandas.core.datetools module is deprecated and will be removed in a future version. Please use the pandas.tseries module instead.\n",
      "  from pandas.core import datetools\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from tqdm import tqdm\n",
    "from svm_epistasis_funcs import *\n",
    "\n",
    "import statsmodels\n",
    "import statsmodels.api as sm\n",
    "import patsy\n",
    "\n",
    "import warnings\n",
    "\n",
    "pd.options.mode.chained_assignment = None  # default='warn'\n",
    "\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "### Define locations and filenames for the pan-genome matrices.\n",
    "DATA_DIR = \"../data/\" \n",
    "SUPP_DIR = \"../supplement/\"\n",
    "FINAL_SUPP_DIR = SUPP_DIR+\"supplement_final/\"\n",
    "\n",
    "# load key data matrices\n",
    "CLUSTER_INFO_FILE = \"cluster_info.csv\"\n",
    "STRAIN_INFO_FILE = \"PATRIC_genome_info.csv\"\n",
    "X_PANGEN_CLUSTERS_FILE = \"pangen_cluster_df.csv\"  # the species-cluster (each cluster represents a \"gene\")\n",
    "X_PANGEN_ALLELES_FILE = \"pangen_allele_df.csv\"  # the species-allele matrix\n",
    "# Y_PHENOTYPES_FILE = \"y_drug_TB.csv\"  # the species-phenotype matrix\n",
    "Y_PHENOTYPES_FILE = \"resistance_data.csv\"  # the species-phenotype matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load resistance data\n",
    "resistance_data = pd.read_csv(DATA_DIR+Y_PHENOTYPES_FILE, index_col = 0) #resistance_data\n",
    "drug_list = list(resistance_data.columns.map(str))\n",
    "\n",
    "# Load pan-genome in both cluster and allele form\n",
    "pan_cluster_df = pd.read_csv(DATA_DIR+X_PANGEN_CLUSTERS_FILE, index_col = 0) #pan_cluster_df\n",
    "pan_variant_df = pd.read_csv(DATA_DIR+X_PANGEN_ALLELES_FILE, index_col = 0) #clust_variant_df\n",
    "\n",
    "# Load cluster information file\n",
    "clust_info_df = pd.read_csv(DATA_DIR+CLUSTER_INFO_FILE, index_col = 0) #test_df_final\n",
    "\n",
    "# Load strain information file\n",
    "strain_info_df = pd.read_csv(DATA_DIR+STRAIN_INFO_FILE, index_col = 0) #...."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# define these dictionaries for ease of mapping...\n",
    "\n",
    "# clust_to_rv_fromJson = load_json_obj(DATA_DIR+\"clust_to_rv.json\")\n",
    "clust_to_rv = pd.Series(clust_info_df.clust_to_rv.values,index=clust_info_df.index).to_dict()\n",
    "rv_to_clust = pd.Series(clust_info_df.index,index=clust_info_df.clust_to_rv.values).to_dict()\n",
    "rv_to_name = pd.Series(clust_info_df.gene_name.values,index=clust_info_df.clust_to_rv.values).to_dict()\n",
    "clust_to_prod = pd.Series(clust_info_df[\"product\"].values,index=clust_info_df.index).to_dict()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Filter out Mobile elements, PE family proteins, and hypotheticals.\n",
    "Should drop the total number of 15367 variants (columns) to 9975"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "pan_variant_df = pan_variant_df.fillna(0).copy()\n",
    "\n",
    "# remove_cols = []\n",
    "# for col in tqdm(pan_variant_df.columns):\n",
    "#     col_name = col.split(\"_\")[0].replace(\"ter\", \"ter \")\n",
    "#     clust_to_RV = clust_info_df.loc[col_name, \"clust_to_rv\"]\n",
    "#     gene_name = clust_info_df.loc[col_name, \"gene_name\"]\n",
    "#     product = clust_info_df.loc[col_name, \"product\"]\n",
    "    \n",
    "#     if clust_to_RV == \"0\" and gene_name == \"0\":\n",
    "#         remove_cols.append(col)\n",
    "#     if product == \"Mobile element protein\" or product == \"hypothetical protein\":\n",
    "#         remove_cols.append(col)\n",
    "#     if \"PE-PGRS\" in product or \"PPE family\" in product or \"PE family protein\" in product:\n",
    "#         remove_cols.append(col)\n",
    "#     if \"hypothetical protein\" in product or \"Hypothetical protein\" in product:\n",
    "#         remove_cols.append(col)\n",
    "#     if \"Transposase\" in product:\n",
    "#         remove_cols.append(col)\n",
    "\n",
    "# remove_cols = list(set(remove_cols))\n",
    "# print \"Number of alleles (variants) before filtering...\", len(pan_variant_df.columns)\n",
    "# pan_variant_df.drop(remove_cols, axis=1, inplace=True)\n",
    "# print \"Number of alleles after filtering...\",len(pan_variant_df.columns)\n",
    "# print \"Number of alleles removed...\",len(remove_cols)\n",
    "data_to_plot = pan_variant_df.copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 04.A \n",
    "### Determine initial set of potential epistatic gene-gene interactions from the ensemble SVM calculations\n",
    "- Store the potential interactions for each AMR phenotype in the dictinoary, $\\hbox{pairwise_table_groups}$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "SVM_ENSEMBLE_DIR = SUPP_DIR + \"svm_ensemble_data/\"\n",
    "COOCCURANCE_TABLE_EXCEL_DIR = SUPP_DIR+\"cooccurence_table_excel/\"\n",
    "COOCCURANCE_TABLE_FIGURES_DIR = SUPP_DIR+\"cooccurence_table_figures/\"\n",
    "GENETIC_INTERACTION_PDF_DIR = FINAL_SUPP_DIR+\"Supplementary_Data_File_4/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "clster_manual_map = {\n",
    "    \"Rv1334\": \"mec\",\n",
    "    \"Rv1184c\": \"Chp2\",\n",
    "    \"Rv3806c\": \"ubiA\",\n",
    "    \"Rv3124\": \"moaR1\",\n",
    "    \"Rv0265c\": \"fecB2\",\n",
    "    \"Rv0669c\": \"NCDase\",\n",
    "    \"Rv1675c\": \"cmr\",\n",
    "    \"Rv3447c\": \"eccC4\",\n",
    "    \"Rv3796\": \"atsH\",\n",
    "    \"Rv1942c\": \"mazF5\",\n",
    "}\n",
    "\n",
    "seqDF_dict = {\n",
    "            \"pyrazinamide\": (\"SVM\", \"SGD\"),\n",
    "             \"ethambutol\": (\"SVM\", \"SGD\"),\n",
    "             \"rifampicin\": (\"SVM\", \"SGD\"),\n",
    "             \"isoniazid\": (\"SVM\", \"SGD\"),\n",
    "             \"ethionamide\": (\"SVM\", \"SGD\"), # 79% AUC\n",
    "             \"4-aminosalicylic_acid\": (\"SVM\", \"SGD\"), # 80% AUC\n",
    "             \"ofloxacin\": (\"SVM\", \"SGD\"),\n",
    "             \"streptomycin\": (\"SVM\", \"SGD\"),\n",
    "             \"XDR\": (\"SVM\", \"SGD\"),\n",
    "             \"MDR\": (\"SVM\", \"SGD\"),\n",
    "             }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "------------- isoniazid -------------\n",
      "------------- ethambutol -------------\n",
      "------------- rifampicin -------------\n",
      "------------- pyrazinamide -------------\n",
      "------------- ethionamide -------------\n",
      "NO GOOD:  Cluster 2122Rv3854c, ethA, Monooxygenase EthA (EC 1.-.-.-)_9 Cluster 87260, 0, Hydrogen peroxide-inducible genes activator_1\n",
      "NO GOOD:  Cluster 87260, 0, Hydrogen peroxide-inducible genes activator_1 Cluster 9540Rv1888A, Rv1888A, O-Methyltransferase involved in polyketide biosynthesis_1\n",
      "NO GOOD:  Cluster 3373Rv1184c, Rv1184c, FIG01386146: Possible exported protein, Rv1184c_2 Cluster 87260, 0, Hydrogen peroxide-inducible genes activator_1\n",
      "NO GOOD:  Cluster 87260, 0, Hydrogen peroxide-inducible genes activator_1 Cluster 4070Rv2151c, ftsQ, Cell division protein FtsQ_2\n",
      "NO GOOD:  Cluster 5664Rv3421c, Rv3421c, TsaB protein, required for threonylcarbamoyladenosine (t(6)A) formation in tRNA_2 Cluster 87260, 0, Hydrogen peroxide-inducible genes activator_1\n",
      "NO GOOD:  Cluster 87260, 0, Hydrogen peroxide-inducible genes activator_1 Cluster 3373Rv1184c, Rv1184c, FIG01386146: Possible exported protein, Rv1184c_5\n",
      "NO GOOD:  Cluster 87260, 0, Hydrogen peroxide-inducible genes activator_1 Cluster 2227Rv0758, phoR, Two-component system response phosphate sensor kinase, PhoR_2\n",
      "NO GOOD:  Cluster 87260, 0, Hydrogen peroxide-inducible genes activator_1 Cluster 7317Rv2656c, Rv2656c, Probable phiRv1 phage protein_3\n",
      "NO GOOD:  Cluster 87260, 0, Hydrogen peroxide-inducible genes activator_1 Cluster 2117Rv2680, Rv2680, Uncharacterized protein Q1 colocalized with Q_6\n",
      "NO GOOD:  Cluster 6929Rv1334, mec, CysO-cysteine peptidase_1 Cluster 87260, 0, Hydrogen peroxide-inducible genes activator_1\n",
      "NO GOOD:  Cluster 87260, 0, Hydrogen peroxide-inducible genes activator_1 Cluster 1916Rv2996c, serA1, D-3-phosphoglycerate dehydrogenase (EC 1.1.1.95)_4\n",
      "------------- ofloxacin -------------\n",
      "------------- streptomycin -------------\n",
      "------------- 4-aminosalicylic_acid -------------\n",
      "------------- MDR -------------\n",
      "------------- XDR -------------\n"
     ]
    }
   ],
   "source": [
    "top_allele_pairs = 100 # 100, to 50, (50- total of 50 gene-gene), 60\n",
    "top_allele_weighted = 25 # 30, (15- total of 50 gene-gene), 30\n",
    "allele_correlation_count = 20\n",
    "drug_sig_gene_interact_dict = {}\n",
    "drug_sig_cluster_interact_dict = {}\n",
    "\n",
    "# for drug_name in [\"ethambutol\"]:\n",
    "# for drug_name in [\"isoniazid\",\"rifampicin\", \"ethambutol\", \"ofloxacin\", \"pyrazinamide\"]: # , \n",
    "for drug_name in [\"isoniazid\",\"ethambutol\",\"rifampicin\",\"pyrazinamide\",\"ethionamide\",\n",
    "                  \"ofloxacin\",\"streptomycin\",\"4-aminosalicylic_acid\",\"MDR\",\"XDR\"]:\n",
    "    \n",
    "    # --- Load the ensemble SVM data matrices --- allele correlation matrix and allele hyperplane weight matrix\n",
    "    print \"-------------\", drug_name, \"-------------\" \n",
    "    estimator_type, SGD_or_GD = seqDF_dict[drug_name]\n",
    "    drug_allele_correlation_df = pd.read_csv(SVM_ENSEMBLE_DIR+drug_name+\"_\"+estimator_type+\"_\"+SGD_or_GD+\"_top60_correlation_matrix.csv\", index_col=0)\n",
    "    # drug_allele_correlation_df = pd.read_csv(SVM_ENSEMBLE_DIR+drug_name+\"_\"+estimator_type+\"_\"+SGD_or_GD+\"_top40_correlation_matrix.csv\", index_col=0)\n",
    "    drug_allele_topWeights_df = pd.read_csv(SVM_ENSEMBLE_DIR+drug_name+\"_\"+estimator_type+\"_\"+SGD_or_GD+\"_top_clusts.csv\", index_col=0)\n",
    "    top_weighted_alleles = list(drug_allele_topWeights_df.index[:top_allele_weighted])\n",
    "    \n",
    "    sig_gene_interact_list = []\n",
    "    \n",
    "    drug_sig_gene_interact_dict[drug_name] = {}\n",
    "    drug_sig_cluster_interact_dict[drug_name] = {}\n",
    "    \n",
    "    # put dataframe in form to look at top correlations\n",
    "    sim_corr_df = drug_allele_correlation_df.loc[top_weighted_alleles, top_weighted_alleles].copy()\n",
    "    abs_corr_df = sim_corr_df.abs()\n",
    "    os = (abs_corr_df.where(np.triu(np.ones(abs_corr_df.shape), k=1).astype(np.bool))\n",
    "                 .stack()\n",
    "                 .sort_values(ascending=False))\n",
    "\n",
    "    count = 0\n",
    "    for (allele_1, allele_2) in os.index[:top_allele_pairs]:\n",
    "        var1, var2 = allele_1.split(\"_\")[-1], allele_2.split(\"_\")[-1]\n",
    "        try:\n",
    "            clst_1,prod_1, clst_2,prod_2 = allele_1.split(\"Rv\")[0],allele_1.split(\", \")[1], allele_2.split(\"Rv\")[0],allele_2.split(\", \")[1]\n",
    "            rv_1, rv_2= \"Rv\"+allele_1.split(\"Rv\")[1].split(\", \")[0], \"Rv\"+allele_2.split(\"Rv\")[1].split(\", \")[0]\n",
    "            if count < allele_correlation_count:\n",
    "                if clst_1 != clst_2:\n",
    "                    if (rv_1, rv_2) not in sig_gene_interact_list and (rv_2, rv_1) not in sig_gene_interact_list:\n",
    "                        # if os.loc[(allele_1, allele_2)] > 0.30:\n",
    "                        if allele_1 in top_weighted_alleles or allele_2 in top_weighted_alleles:\n",
    "                            sig_gene_interact_list.append((rv_1, rv_2))\n",
    "                            drug_sig_gene_interact_dict[drug_name].update({(rv_1, rv_2): os.loc[(allele_1, allele_2)]})\n",
    "                            drug_sig_cluster_interact_dict[drug_name].update({(clst_1, clst_2): os.loc[(allele_1, allele_2)]})\n",
    "                            count = count + 1\n",
    "\n",
    "        except:\n",
    "            print \"NO GOOD: \", allele_1, allele_2\n",
    "                \n",
    "pairwise_table_groups = drug_sig_cluster_interact_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The cluster groups with high SVM correlations may be indicative of epistatic interactions. In some of these cases, the resistance conferring mutation may be independent of the other mutation. We therefore performed classical epistatic analysis through logistic regression models."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Cluster 893Rv0006, gyrA, DNA gyrase subunit A (EC 5.99.1.3)_5',\n",
       " 'Cluster 893Rv0006, gyrA, DNA gyrase subunit A (EC 5.99.1.3)_13',\n",
       " 'Cluster 491Rv0667, rpoB, DNA-directed RNA polymerase beta subunit (EC 2.7.7.6)_9',\n",
       " 'Cluster 4244Rv3806c, ubiA, 5-Phosphoribosyl diphosphate (PRPP): decaprenyl-phosphate 5-phosphoribosyltransferase (EC 2.4.2.45)_2',\n",
       " 'Cluster 3930Rv2043c, pncA, Nicotinamidase (EC 3.5.1.19)_10',\n",
       " 'Cluster 893Rv0006, gyrA, DNA gyrase subunit A (EC 5.99.1.3)_9',\n",
       " 'Cluster 7926Rv1102c, mazF3, Death on curing protein, Doc toxin_6',\n",
       " 'Cluster 893Rv0006, gyrA, DNA gyrase subunit A (EC 5.99.1.3)_8',\n",
       " 'Cluster 3909Rv3848, Rv3848, FIG037441: Conserved transmembrane protein_2',\n",
       " 'Cluster 1277Rv0959, Rv0959, FIG019045: long form Mg-chelase associated protein with vWA domain_2',\n",
       " 'Cluster 7125Rv2757c, vapC21, toxin VapC18_7',\n",
       " 'Cluster 893Rv0006, gyrA, DNA gyrase subunit A (EC 5.99.1.3)_12',\n",
       " 'Cluster 2122Rv3854c, ethA, Monooxygenase EthA (EC 1.-.-.-)_9',\n",
       " 'Cluster 8038Rv3923c, rnpA, Ribonuclease P protein component (EC 3.1.26.5)_4',\n",
       " 'Cluster 4476Rv0633c, Rv0633c, putative membrane protein_5',\n",
       " 'Cluster 1116Rv1908c, katG, Catalase (EC 1.11.1.6) / Peroxidase (EC 1.11.1.7)_7',\n",
       " 'Cluster 2227Rv0758, phoR, Two-component system response phosphate sensor kinase, PhoR_2',\n",
       " 'Cluster 491Rv0667, rpoB, DNA-directed RNA polymerase beta subunit (EC 2.7.7.6)_4',\n",
       " 'Cluster 5780Rv1972, Rv1972, FIG00821219: MCE associated membrane protein_4',\n",
       " 'Cluster 4244Rv3806c, ubiA, 5-Phosphoribosyl diphosphate (PRPP): decaprenyl-phosphate 5-phosphoribosyltransferase (EC 2.4.2.45)_3',\n",
       " 'Cluster 4486Rv2764c, thyA, Thymidylate synthase (EC 2.1.1.45)_4',\n",
       " \"Cluster 1153Rv2112c, dop, Pup ligase PafA' paralog, possible component of postulated heterodimer PafA-PafA'_5\",\n",
       " 'Cluster 1058Rv0092, ctpA, Lead, cadmium, zinc and mercury transporting ATPase (EC 3.6.3.3) (EC 3.6.3.5); Copper-translocating P-type ATPase (EC 3.6.3.4)_3',\n",
       " 'Cluster 3598Rv3124, moaR1, transcriptional regulator, AfsR/DnrI/RedD family_4',\n",
       " \"Cluster 5948Rv2604c, snoP, Pyridoxal 5'-phosphate synthase (glutamine hydrolyzing), glutaminase subunit (EC 4.3.3.6)_1\"]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "top_weighted_alleles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>pair_id</th>\n",
       "      <th>pair_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>(Cluster 491, Cluster 1206)</th>\n",
       "      <td>0.473710</td>\n",
       "      <td>(Rv0667, Rv2721c)</td>\n",
       "      <td>(rpoB, Rv2721c)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(Cluster 491, Cluster 1237)</th>\n",
       "      <td>0.440304</td>\n",
       "      <td>(Rv0667, Rv1877)</td>\n",
       "      <td>(rpoB, Rv1877)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(Cluster 491, Cluster 5025)</th>\n",
       "      <td>0.386533</td>\n",
       "      <td>(Rv0667, Rv1597)</td>\n",
       "      <td>(rpoB, Rv1597)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(Cluster 491, Cluster 2117)</th>\n",
       "      <td>0.362769</td>\n",
       "      <td>(Rv0667, Rv2680)</td>\n",
       "      <td>(rpoB, Rv2680)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(Cluster 491, Cluster 1068)</th>\n",
       "      <td>0.342714</td>\n",
       "      <td>(Rv0667, Rv1633)</td>\n",
       "      <td>(rpoB, uvrB)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(Cluster 491, Cluster 3865)</th>\n",
       "      <td>0.313824</td>\n",
       "      <td>(Rv0667, Rv1244)</td>\n",
       "      <td>(rpoB, lpqZ)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(Cluster 1237, Cluster 2117)</th>\n",
       "      <td>0.299835</td>\n",
       "      <td>(Rv1877, Rv2680)</td>\n",
       "      <td>(Rv1877, Rv2680)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(Cluster 2117, Cluster 2754)</th>\n",
       "      <td>0.291671</td>\n",
       "      <td>(Rv2680, Rv1526c)</td>\n",
       "      <td>(Rv2680, Rv1526c)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(Cluster 1237, Cluster 5025)</th>\n",
       "      <td>0.281297</td>\n",
       "      <td>(Rv1877, Rv1597)</td>\n",
       "      <td>(Rv1877, Rv1597)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(Cluster 491, Cluster 2754)</th>\n",
       "      <td>0.279692</td>\n",
       "      <td>(Rv0667, Rv1526c)</td>\n",
       "      <td>(rpoB, Rv1526c)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(Cluster 1237, Cluster 1396)</th>\n",
       "      <td>0.262605</td>\n",
       "      <td>(Rv1877, Rv0669c)</td>\n",
       "      <td>(Rv1877, Rv0669c)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(Cluster 491, Cluster 577)</th>\n",
       "      <td>0.261348</td>\n",
       "      <td>(Rv0667, Rv1683)</td>\n",
       "      <td>(rpoB, Rv1683)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(Cluster 1237, Cluster 1068)</th>\n",
       "      <td>0.247750</td>\n",
       "      <td>(Rv1877, Rv1633)</td>\n",
       "      <td>(Rv1877, uvrB)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(Cluster 1206, Cluster 2117)</th>\n",
       "      <td>0.247696</td>\n",
       "      <td>(Rv2721c, Rv2680)</td>\n",
       "      <td>(Rv2721c, Rv2680)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(Cluster 491, Cluster 415)</th>\n",
       "      <td>0.239576</td>\n",
       "      <td>(Rv0667, Rv0668)</td>\n",
       "      <td>(rpoB, rpoC)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(Cluster 491, Cluster 4738)</th>\n",
       "      <td>0.229855</td>\n",
       "      <td>(Rv0667, Rv1484)</td>\n",
       "      <td>(rpoB, inhA)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(Cluster 491, Cluster 3178)</th>\n",
       "      <td>0.224965</td>\n",
       "      <td>(Rv0667, Rv2777c)</td>\n",
       "      <td>(rpoB, Rv2777c)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(Cluster 491, Cluster 4237)</th>\n",
       "      <td>0.204748</td>\n",
       "      <td>(Rv0667, Rv3516)</td>\n",
       "      <td>(rpoB, echA19)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(Cluster 4738, Cluster 3178)</th>\n",
       "      <td>0.203836</td>\n",
       "      <td>(Rv1484, Rv2777c)</td>\n",
       "      <td>(inhA, Rv2777c)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(Cluster 1237, Cluster 619)</th>\n",
       "      <td>0.201763</td>\n",
       "      <td>(Rv1877, Rv1664)</td>\n",
       "      <td>(Rv1877, pks9)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     0            pair_id          pair_name\n",
       "(Cluster 491, Cluster 1206)   0.473710  (Rv0667, Rv2721c)    (rpoB, Rv2721c)\n",
       "(Cluster 491, Cluster 1237)   0.440304   (Rv0667, Rv1877)     (rpoB, Rv1877)\n",
       "(Cluster 491, Cluster 5025)   0.386533   (Rv0667, Rv1597)     (rpoB, Rv1597)\n",
       "(Cluster 491, Cluster 2117)   0.362769   (Rv0667, Rv2680)     (rpoB, Rv2680)\n",
       "(Cluster 491, Cluster 1068)   0.342714   (Rv0667, Rv1633)       (rpoB, uvrB)\n",
       "(Cluster 491, Cluster 3865)   0.313824   (Rv0667, Rv1244)       (rpoB, lpqZ)\n",
       "(Cluster 1237, Cluster 2117)  0.299835   (Rv1877, Rv2680)   (Rv1877, Rv2680)\n",
       "(Cluster 2117, Cluster 2754)  0.291671  (Rv2680, Rv1526c)  (Rv2680, Rv1526c)\n",
       "(Cluster 1237, Cluster 5025)  0.281297   (Rv1877, Rv1597)   (Rv1877, Rv1597)\n",
       "(Cluster 491, Cluster 2754)   0.279692  (Rv0667, Rv1526c)    (rpoB, Rv1526c)\n",
       "(Cluster 1237, Cluster 1396)  0.262605  (Rv1877, Rv0669c)  (Rv1877, Rv0669c)\n",
       "(Cluster 491, Cluster 577)    0.261348   (Rv0667, Rv1683)     (rpoB, Rv1683)\n",
       "(Cluster 1237, Cluster 1068)  0.247750   (Rv1877, Rv1633)     (Rv1877, uvrB)\n",
       "(Cluster 1206, Cluster 2117)  0.247696  (Rv2721c, Rv2680)  (Rv2721c, Rv2680)\n",
       "(Cluster 491, Cluster 415)    0.239576   (Rv0667, Rv0668)       (rpoB, rpoC)\n",
       "(Cluster 491, Cluster 4738)   0.229855   (Rv0667, Rv1484)       (rpoB, inhA)\n",
       "(Cluster 491, Cluster 3178)   0.224965  (Rv0667, Rv2777c)    (rpoB, Rv2777c)\n",
       "(Cluster 491, Cluster 4237)   0.204748   (Rv0667, Rv3516)     (rpoB, echA19)\n",
       "(Cluster 4738, Cluster 3178)  0.203836  (Rv1484, Rv2777c)    (inhA, Rv2777c)\n",
       "(Cluster 1237, Cluster 619)   0.201763   (Rv1877, Rv1664)     (Rv1877, pks9)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x1c166c0f50>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXcAAAD8CAYAAACMwORRAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMS4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvNQv5yAAAEbJJREFUeJzt3W2MXGd5h/HrxjYhzQYTcLREjmGDcKuGWG3wKkRCbWeBKiFA3A9BCk1DjEAWKAHUBiFDq6AGIV6qgECJSF0lIiDEhgKt3NgtioAl5ENS1qmJ47gpTnCbtaMAduqwYEIX7n7Ys2QYdj1ndma8M4+unzTyOTPPM+e+5+z89+yZF0dmIkkqy3NWugBJUu8Z7pJUIMNdkgpkuEtSgQx3SSqQ4S5JBTLcJalAhrskFchwl6QCrV6pDa9bty7HxsY6mvPTn/6UM844oz8FDQD7G272N9yGpb89e/b8ODPPbjduxcJ9bGyM6enpjuZMTU3RaDT6U9AAsL/hZn/DbVj6i4j/rjPO0zKSVCDDXZIKZLhLUoEMd0kqkOEuSQVqG+4R8byI+PeI+F5E7I+Iv11kzGkRcWdEHIyI+yNirB/FSpLqqXPk/gzwmsz8A+APgUsj4uKWMW8HnsrMlwOfAj7e2zIlSZ1oG+45b7ZaXVNdWv9vvi3AHdXyV4DXRkT0rEpJUkdqnXOPiFURsRf4IXB3Zt7fMmQ98DhAZs4Bx4EX9bJQSVJ90cl/kB0RLwD+CXh3Zj7UdP1+4JLMnKnWHwUuysyjLfO3AdsARkdHN09OTnZU7OzsLCMjI+w7fLyjea02rV/b1fx+WeivVPY33OxvMExMTOzJzPF24zr6+oHM/N+ImAIuBR5qumkG2ADMRMRqYC1wbJH5O4AdAOPj49npR30XPh68dfuujua1OnRVZ9s9VYbl48/LZX/Dzf6GS513y5xdHbETEacDrwP+s2XYTuCaavkK4JvZyZ8EkqSeqnPkfg5wR0SsYv6XwZcz866IuBGYzsydwG3AFyLiIPNH7Ff2rWJJUlttwz0zHwQuXOT6G5qWfw68ubelSZKWy0+oSlKBDHdJKpDhLkkFMtwlqUCGuyQVyHCXpAIZ7pJUIMNdkgpkuEtSgQx3SSqQ4S5JBTLcJalAhrskFchwl6QCGe6SVCDDXZIKZLhLUoEMd0kqkOEuSQUy3CWpQIa7JBXIcJekAhnuklQgw12SCmS4S1KBDHdJKlDbcI+IDRHxrYg4EBH7I+K9i4xpRMTxiNhbXW7oT7mSpDpW1xgzB1yfmQ9ExJnAnoi4OzMfbhn3ncx8Y+9LlCR1qu2Re2Y+kZkPVMs/AQ4A6/tdmCRp+SIz6w+OGAPuAS7IzKebrm8AXwVmgCPA+zJz/yLztwHbAEZHRzdPTk52VOzs7CwjIyPsO3y8o3mtNq1f29X8flnor1T2N9zsbzBMTEzsyczxduNqh3tEjADfBj6SmV9rue35wK8yczYiLgM+nZkbT3Z/4+PjOT09XWvbC6ampmg0Goxt39XRvFaHPvaGrub3y0J/pbK/4WZ/gyEiaoV7rXfLRMQa5o/Mv9ga7ACZ+XRmzlbLu4E1EbGuw5olST1S590yAdwGHMjMTy4x5sXVOCLioup+j/ayUElSfXXeLfNq4GpgX0Tsra77IPASgMy8FbgCeFdEzAEngCuzk5P5kqSeahvumXkvEG3G3Azc3KuiJEnd8ROqklQgw12SCmS4S1KBDHdJKpDhLkkFMtwlqUCGuyQVyHCXpAIZ7pJUIMNdkgpkuEtSgQx3SSqQ4S5JBTLcJalAhrskFchwl6QCGe6SVCDDXZIKZLhLUoEMd0kqkOEuSQUy3CWpQIa7JBXIcJekAhnuklSgtuEeERsi4lsRcSAi9kfEexcZExHxmYg4GBEPRsQr+1OuJKmO1TXGzAHXZ+YDEXEmsCci7s7Mh5vGvB7YWF1eBXy2+leStALaHrln5hOZ+UC1/BPgALC+ZdgW4PM57z7gBRFxTs+rlSTV0tE594gYAy4E7m+5aT3weNP6DL/9C0CSdIpEZtYbGDECfBv4SGZ+reW2XcBHM/Peav0bwPszc0/LuG3ANoDR0dHNk5OTHRU7OzvLyMgI+w4f72heq03r1y57brfbPpnR0+HJE327+2Xr5vFqtrD/SmV/w21Y+puYmNiTmePtxtU5505ErAG+CnyxNdgrM8CGpvVzgSOtgzJzB7ADYHx8PBuNRp3N/9rU1BSNRoOt23d1NK/Voas6226zbrd9MtdvmuOmfbV2ySnVzePVbGH/lcr+hltp/dV5t0wAtwEHMvOTSwzbCby1etfMxcDxzHyih3VKkjpQ5zDx1cDVwL6I2Ftd90HgJQCZeSuwG7gMOAj8DHhb70uVJNXVNtyr8+jRZkwC1/aqKElSd/yEqiQVyHCXpAIZ7pJUIMNdkgpkuEtSgQx3SSqQ4S5JBTLcJalAhrskFchwl6QCGe6SVCDDXZIKZLhLUoEMd0kqkOEuSQUy3CWpQIa7JBXIcJekAhnuklQgw12SCmS4S1KBDHdJKpDhLkkFMtwlqUCGuyQVyHCXpAK1DfeIuD0ifhgRDy1xeyMijkfE3upyQ+/LlCR1YnWNMZ8DbgY+f5Ix38nMN/akIklS19oeuWfmPcCxU1CLJKlHIjPbD4oYA+7KzAsWua0BfBWYAY4A78vM/UvczzZgG8Do6OjmycnJjoqdnZ1lZGSEfYePdzSv1ab1a5c9t9ttn8zo6fDkib7d/bJ183g1W9h/pbK/4TYs/U1MTOzJzPF243oR7s8HfpWZsxFxGfDpzNzY7j7Hx8dzenq67babTU1N0Wg0GNu+q6N5rQ597A3Lntvttk/m+k1z3LSvzpmyU6ubx6vZwv4rlf0Nt2HpLyJqhXvX75bJzKczc7Za3g2siYh13d6vJGn5ug73iHhxRES1fFF1n0e7vV9J0vK1PQcQEV8CGsC6iJgBPgSsAcjMW4ErgHdFxBxwArgy65zrkST1Tdtwz8y3tLn9ZubfKilJGhB+QlWSCmS4S1KBDHdJKpDhLkkFMtwlqUCGuyQVyHCXpAIZ7pJUIMNdkgpkuEtSgQx3SSqQ4S5JBTLcJalAhrskFchwl6QCGe6SVCDDXZIKZLhLUoEMd0kqkOEuSQUy3CWpQIa7JBXIcJekAhnuklQgw12SCtQ23CPi9oj4YUQ8tMTtERGfiYiDEfFgRLyy92VKkjpR58j9c8ClJ7n99cDG6rIN+Gz3ZUmSutE23DPzHuDYSYZsAT6f8+4DXhAR5/SqQElS53pxzn098HjT+kx1nSRphURmth8UMQbclZkXLHLbLuCjmXlvtf4N4P2ZuWeRsduYP3XD6Ojo5snJyY6KnZ2dZWRkhH2Hj3c0b1iMng5PnljpKvrnVPa3af3aruYv52esF/11U3c3z4s62114/g2SXvbcaX/9fryXMjExsSczx9uNW73sLTxrBtjQtH4ucGSxgZm5A9gBMD4+no1Go6MNTU1N0Wg02Lp91/IqHXDXb5rjpn292CWD6VT2d+iqRlfzl/Mz1ov+uqm7m+dFne0uPP8GSS977rS/fj/e3erFaZmdwFurd81cDBzPzCd6cL+SpGVqe5gREV8CGsC6iJgBPgSsAcjMW4HdwGXAQeBnwNv6VawkqZ624Z6Zb2lzewLX9qwiSVLX/ISqJBXIcJekAhnuklQgw12SCmS4S1KBDHdJKpDhLkkFMtwlqUCGuyQVyHCXpAIZ7pJUIMNdkgpkuEtSgQx3SSqQ4S5JBTLcJalAhrskFchwl6QCGe6SVCDDXZIKZLhLUoEMd0kqkOEuSQUy3CWpQIa7JBXIcJekAtUK94i4NCIeiYiDEbF9kdu3RsSPImJvdXlH70uVJNW1ut2AiFgF3AL8KTADfDcidmbmwy1D78zM6/pQoySpQ3WO3C8CDmbmY5n5C2AS2NLfsiRJ3YjMPPmAiCuASzPzHdX61cCrmo/SI2Ir8FHgR8B/AX+ZmY8vcl/bgG0Ao6OjmycnJzsqdnZ2lpGREfYdPt7RvGExejo8eWKlq+ifU9nfpvVru5q/nJ+xXvTXTd3dPC/qbHfh+TdIetlzp/31+/FeysTExJ7MHG83ru1pGSAWua71N8K/AF/KzGci4p3AHcBrfmtS5g5gB8D4+Hg2Go0am3/W1NQUjUaDrdt3dTRvWFy/aY6b9tXZJcPpVPZ36KpGV/OX8zPWi/66qbub50Wd7S48/wZJL3vutL9+P97dqnNaZgbY0LR+LnCkeUBmHs3MZ6rVfwA296Y8SdJy1An37wIbI+K8iHgucCWws3lARJzTtHo5cKB3JUqSOtX2b8jMnIuI64CvA6uA2zNzf0TcCExn5k7gPRFxOTAHHAO29rFmSVIbtU4QZuZuYHfLdTc0LX8A+EBvS5MkLZefUJWkAhnuklQgw12SCmS4S1KBDHdJKpDhLkkFMtwlqUCGuyQVyHCXpAIZ7pJUIMNdkgpkuEtSgQx3SSqQ4S5JBTLcJalAhrskFchwl6QCGe6SVCDDXZIKZLhLUoEMd0kqkOEuSQUy3CWpQIa7JBXIcJekAtUK94i4NCIeiYiDEbF9kdtPi4g7q9vvj4ixXhcqSaqvbbhHxCrgFuD1wPnAWyLi/JZhbweeysyXA58CPt7rQiVJ9dU5cr8IOJiZj2XmL4BJYEvLmC3AHdXyV4DXRkT0rkxJUifqhPt64PGm9ZnqukXHZOYccBx4US8KlCR1LjLz5AMi3gxckpnvqNavBi7KzHc3jdlfjZmp1h+txhxtua9twLZq9feARzqsdx3w4w7nDBP7G272N9yGpb+XZubZ7QatrnFHM8CGpvVzgSNLjJmJiNXAWuBY6x1l5g5gR41tLioipjNzfLnzB539DTf7G26l9VfntMx3gY0RcV5EPBe4EtjZMmYncE21fAXwzWz3J4EkqW/aHrln5lxEXAd8HVgF3J6Z+yPiRmA6M3cCtwFfiIiDzB+xX9nPoiVJJ1fntAyZuRvY3XLdDU3LPwfe3NvSFrXsUzpDwv6Gm/0Nt6L6a/uCqiRp+Pj1A5JUoIEJ9xpfcfBXEfFwRDwYEd+IiJc23XZNRHy/ulzTOncQdNnfLyNib3VpfTF7INTo750Rsa/q4d7mTzlHxAeqeY9ExCWntvJ6lttfRIxFxImm/Xfrqa++vXb9NY27IiIyIsabrhv6/dc07jf6G5b9t6jMXPEL8y/UPgq8DHgu8D3g/JYxE8DvVMvvAu6sll8IPFb9e1a1fNZK99Sr/qr12ZXuoQf9Pb9p+XLg36rl86vxpwHnVfezaqV76mF/Y8BDK91Dt/1V484E7gHuA8ZL2n8n6W/g999Sl0E5cm/7FQeZ+a3M/Fm1eh/z77cHuAS4OzOPZeZTwN3Apaeo7rq66W8Y1Onv6abVM4CFF3u2AJOZ+Uxm/gA4WN3fIOmmv2FQ5ytGAD4MfAL4edN1Rey/ymL9Da1BCfc6X3HQ7O3Avy5z7kropj+A50XEdETcFxF/1o8Cu1Srv4i4tvr08ieA93Qyd4V10x/AeRHxHxHx7Yj4o/6Wuixt+4uIC4ENmXlXp3MHQDf9weDvv0XVeivkKbDYl4wteuQTEX8BjAN/0uncFdRNfwAvycwjEfEy4JsRsS8zH+1DnctVq7/MvAW4JSL+HPgb5j/4Vsz+W6K/J5jff0cjYjPwzxHxipYj/ZV20v4i4jnMf9vr1k7nDohu+huG/beoQTlyr/MVB0TE64C/Bi7PzGc6mbvCuumPzDxS/fsYMAVc2M9il6HTfTAJLPwFUsz+a/Lr/qrTFUer5T3Mn/v93T7VuVzt+jsTuACYiohDwMXAzupFxxL235L9Dcn+W9xKn/SvXrRYzfwLoefx7Aser2gZcyHzD+zGlutfCPyA+RdTz6qWX7jSPfWwv7OA06rldcD3WeTFoCHob2PT8puY/3QzwCv4zRfkHmPwXpDrpr+zF/ph/gW9w8P489kyfopnX3AsYv+dpL+B339LXQbitEzW+4qDvwNGgH+M+a+K/5/MvDwzj0XEh5n/DhyAGzPzt760bCV10x/w+8DfR8SvmP9L62OZ+fCKNLKEmv1dV/1l8n/AU1TfRVSN+zLwMDAHXJuZv1yRRpbQTX/AHwM3RsQc8EvgnUP687nU3FL231IGfv8txU+oSlKBBuWcuySphwx3SSqQ4S5JBTLcJalAhrskFchwl6QCGe6SVCDDXZIK9P+r2mKniizyHwAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x1c1586dc90>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "num_view = 100\n",
    "for antibiotic in [\"rifampicin\"]: # \"ethambutol\", \"rifampicin\", \"isoniazid\", \"ofloxacin\", \"pyrazinamide\"\n",
    "    anti_df = pd.DataFrame(pairwise_table_groups[antibiotic].values(), \n",
    "             index=pairwise_table_groups[antibiotic].keys()).sort_values([0],ascending=False)\n",
    "    anti_df[\"pair_id\"] = anti_df.index.map(lambda x: (clust_to_rv[x[0]],clust_to_rv[x[1]]))\n",
    "    print len(anti_df.index)\n",
    "    anti_df[\"pair_name\"] = anti_df[\"pair_id\"].map(lambda x: (rv_to_name[x[0]],rv_to_name[x[1]]))\n",
    "    display(anti_df[:num_view]), display(anti_df[0].hist(bins=20))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 04.B \n",
    "### Determine which of the potential gene-gene interactions are significant using logistic regression models.\n",
    "1. Compute all significant gene-gene interactions.\n",
    "2. Create a co-occurrance table for each gene-gene interaction\n",
    "\n",
    "Parameters\n",
    "- **color_measure**  - Metric used to color the allele co-occurrance table (oddsRatio, percent, oddsRatio\", pVal). $\\fbox{Default is oddsRatio}$\n",
    "- **pVal_test_type** - Type of significance test (chi2, f_classif). Note that f_classif is only useful if color_measure = \"pVal\". $\\fbox{Default is chi2}$\n",
    "- **add_zeroCell** - Adds this value to each cell in the odds ratio table in order to account for the cases where there is a 0. $\\fbox{Default is 0.5}$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set parameters\n",
    "color_measure = \"oddsRatio\" \n",
    "pVal_test_type = \"chi2\" \n",
    "add_zeroCell = 0.5 \n",
    "pVal_threshold_epistasis = 0.05"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Since each optimization inside the profile likelihood generates a warning (due to the random slope variance being close to zero), we turn off the warnings here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "def drop_collinearity_variables(X_alleles, Y_alleles, interaction_terms_list, X_logReg_df, verbose):\n",
    "    \"\"\"\n",
    "    Filter out allelic variables in epistatic logistic regression model according to the\n",
    "    following criteria.\n",
    "        Criteria: \n",
    "            - x == xy, drop allele xy\n",
    "            - \n",
    "    \"\"\"\n",
    "    drop_collinearity_alleles = []\n",
    "    for allele_test_x in X_alleles:\n",
    "        for allele_test_y in Y_alleles:\n",
    "            allele_test_xy = allele_test_x+\":\"+allele_test_y\n",
    "            if allele_test_xy in interaction_terms_list:\n",
    "                strains_collin_x_xy = X_logReg_df[X_logReg_df[allele_test_x] == X_logReg_df[allele_test_xy]]\n",
    "                strains_collin_y_xy = X_logReg_df[X_logReg_df[allele_test_y] == X_logReg_df[allele_test_xy]]\n",
    "                strains_collin_x_y = X_logReg_df[X_logReg_df[allele_test_y] == X_logReg_df[allele_test_x]]\n",
    "                if len(strains_collin_x_xy) == len(X_logReg_df.index):\n",
    "                    if verbose==True:\n",
    "                        print \"collinearity...x_xy ... dropping: \", allele_test_x, len(strains_collin_x_xy)\n",
    "                    \n",
    "                    drop_collinearity_alleles.append(allele_test_xy) # off before.... # got oxcA\n",
    "                    # drop_collinearity_alleles.append(allele_test_x) # original\n",
    "                    \n",
    "                elif len(strains_collin_y_xy) == len(X_logReg_df.index):\n",
    "                    if verbose==True:\n",
    "                        print \"collinearity...y_xy ... dropping: \", allele_test_y, len(strains_collin_x_xy)\n",
    "                    drop_collinearity_alleles.append(allele_test_xy) # off before.... # got oxcA\n",
    "                    # drop_collinearity_alleles.append(allele_test_y) # original\n",
    "                    \n",
    "                elif len(strains_collin_x_y) == len(X_logReg_df.index):\n",
    "                    if verbose==True:\n",
    "                        print \"collinearity...x_y ... dropping: \", allele_test_x, allele_test_y, allele_test_xy, len(strains_collin_x_xy)\n",
    "                    # ---- If the presence absence vectors of allele x and allele y are exactly the same,\n",
    "                    # drop these alleles along with their interaction term.\n",
    "                    drop_collinearity_alleles.extend([allele_test_x, allele_test_y, allele_test_xy]) # , allele_test_xy\n",
    "    \n",
    "    # ---- Check for the case where 3 interaction terms should be removed if contained by a single allele.\n",
    "    \n",
    "    for allele_test_x in X_alleles:\n",
    "        allele_test_x_ALL_y = [str(x) for x in interaction_terms_list if str(allele_test_x)+\":\" in x]\n",
    "        strains_collin_x_ALL_xy = X_logReg_df[X_logReg_df[allele_test_x] == X_logReg_df[allele_test_x_ALL_y].sum(axis=1)].index\n",
    "        \n",
    "        if len(strains_collin_x_ALL_xy) == len(X_logReg_df.index):\n",
    "            # -- new drop code... remove allele\n",
    "            if allele_test_x not in drop_collinearity_alleles:\n",
    "                if verbose==True:\n",
    "                    print \"collinearity...x_ALL_y ... dropping: \", allele_test_x\n",
    "                drop_collinearity_alleles.append(allele_test_x) # original\n",
    "                \n",
    "#             for z in allele_test_x_ALL_y:\n",
    "#                 if z not in drop_collinearity_alleles:\n",
    "#                     drop_collinearity_alleles.append(z)\n",
    "\n",
    "    for allele_test_y in Y_alleles:\n",
    "        allele_test_y_ALL_x = [str(x) for x in interaction_terms_list if \":\"+str(allele_test_y) in x]\n",
    "        strains_collin_y_ALL_xy = X_logReg_df[X_logReg_df[allele_test_y] == X_logReg_df[allele_test_y_ALL_x].sum(axis=1)].index\n",
    "        \n",
    "        if len(strains_collin_y_ALL_xy) == len(X_logReg_df.index):\n",
    "            # -- new drop code\n",
    "            if allele_test_y not in drop_collinearity_alleles:\n",
    "                if verbose==True:\n",
    "                    print \"collinearity...y_ALL_x ... dropping: \", allele_test_y\n",
    "                drop_collinearity_alleles.append(allele_test_y) # original\n",
    "            \n",
    "#             for z in allele_test_y_ALL_x:\n",
    "#                 if z not in drop_collinearity_alleles:\n",
    "#                     drop_collinearity_alleles.append(z)\n",
    "    if verbose==True:\n",
    "        print \"Dropped variables in locus-locus logistic model: \", drop_collinearity_alleles\n",
    "    \n",
    "    return drop_collinearity_alleles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'4-aminosalicylic_acid': {('Cluster 1404',\n",
       "   'Cluster 3289'): 0.23559419604031806,\n",
       "  ('Cluster 2211', 'Cluster 1227'): 0.21053891353448315,\n",
       "  ('Cluster 2211', 'Cluster 4806'): 0.20238197992761114,\n",
       "  ('Cluster 2954', 'Cluster 2698'): 0.1847280207848692,\n",
       "  ('Cluster 2954', 'Cluster 3005'): 0.17896017865231467,\n",
       "  ('Cluster 2954', 'Cluster 3289'): 0.2220104229221012,\n",
       "  ('Cluster 3316', 'Cluster 4053'): 0.20251337309512146,\n",
       "  ('Cluster 4053', 'Cluster 2698'): 0.28854781614341035,\n",
       "  ('Cluster 4486', 'Cluster 1404'): 0.19009926329445892,\n",
       "  ('Cluster 4486', 'Cluster 594'): 0.2236130205091492,\n",
       "  ('Cluster 4603', 'Cluster 2698'): 0.2328057750397428,\n",
       "  ('Cluster 4603', 'Cluster 4536'): 0.2624069630390036,\n",
       "  ('Cluster 4806', 'Cluster 1916'): 0.19975571026094766,\n",
       "  ('Cluster 594', 'Cluster 5949'): 0.29472337858732633,\n",
       "  ('Cluster 5949', 'Cluster 4486'): 0.18402371564621164,\n",
       "  ('Cluster 5949', 'Cluster 4603'): 0.1876917132930228,\n",
       "  ('Cluster 7466', 'Cluster 4053'): 0.18552759346863096,\n",
       "  ('Cluster 7466', 'Cluster 5385'): 0.18074631269048047,\n",
       "  ('Cluster 8101', 'Cluster 2331'): 0.2198548838622581,\n",
       "  ('Cluster 8101', 'Cluster 4806'): 0.1908672706768136},\n",
       " 'MDR': {('Cluster 1116', 'Cluster 2122'): 0.2490294848620745,\n",
       "  ('Cluster 1116', 'Cluster 3881'): 0.27604479152009537,\n",
       "  ('Cluster 1116', 'Cluster 415'): 0.1765576168573769,\n",
       "  ('Cluster 1116', 'Cluster 4244'): 0.2592230061243553,\n",
       "  ('Cluster 1116', 'Cluster 5240'): 0.14999702690394606,\n",
       "  ('Cluster 1116', 'Cluster 551'): 0.20520425456563585,\n",
       "  ('Cluster 1116', 'Cluster 893'): 0.2616619731277934,\n",
       "  ('Cluster 2122', 'Cluster 5240'): 0.1625342368505905,\n",
       "  ('Cluster 3930', 'Cluster 3881'): 0.16332132746185835,\n",
       "  ('Cluster 3930', 'Cluster 551'): 0.1868431670570461,\n",
       "  ('Cluster 415', 'Cluster 4244'): 0.18056509940210227,\n",
       "  ('Cluster 415', 'Cluster 7435'): 0.1562566028861843,\n",
       "  ('Cluster 491', 'Cluster 1116'): 0.2696829165345013,\n",
       "  ('Cluster 491', 'Cluster 3881'): 0.370171467623778,\n",
       "  ('Cluster 491', 'Cluster 3930'): 0.17723145711851462,\n",
       "  ('Cluster 491', 'Cluster 893'): 0.2991209175864579,\n",
       "  ('Cluster 5240', 'Cluster 3881'): 0.1651663121167852,\n",
       "  ('Cluster 551', 'Cluster 893'): 0.2307224595544471,\n",
       "  ('Cluster 7435', 'Cluster 2178'): 0.1440162575268822,\n",
       "  ('Cluster 893', 'Cluster 3881'): 0.1991286398848917},\n",
       " 'XDR': {('Cluster 3930', 'Cluster 2122'): 0.24597083126856284,\n",
       "  ('Cluster 3930', 'Cluster 5948'): 0.2641268686245499,\n",
       "  ('Cluster 4244', 'Cluster 3930'): 0.2931968402821196,\n",
       "  ('Cluster 4244', 'Cluster 7125'): 0.2416618910875225,\n",
       "  ('Cluster 4476', 'Cluster 4244'): 0.29830242881951513,\n",
       "  ('Cluster 491', 'Cluster 3909'): 0.285661168843912,\n",
       "  ('Cluster 491', 'Cluster 3930'): 0.24819155601912696,\n",
       "  ('Cluster 491', 'Cluster 4244'): 0.8854370924134465,\n",
       "  ('Cluster 7926', 'Cluster 4244'): 0.39577272157297866,\n",
       "  ('Cluster 8038', 'Cluster 4244'): 0.2682407164888891,\n",
       "  ('Cluster 893', 'Cluster 1058'): 0.2650909269181163,\n",
       "  ('Cluster 893', 'Cluster 1116'): 0.238948517196205,\n",
       "  ('Cluster 893', 'Cluster 2122'): 0.40941030227292896,\n",
       "  ('Cluster 893', 'Cluster 3930'): 0.4567860309924842,\n",
       "  ('Cluster 893', 'Cluster 4244'): 0.4676724459249829,\n",
       "  ('Cluster 893', 'Cluster 4486'): 0.29224036407370313,\n",
       "  ('Cluster 893', 'Cluster 491'): 0.3955240613305333,\n",
       "  ('Cluster 893', 'Cluster 5948'): 0.2500600882243797,\n",
       "  ('Cluster 893', 'Cluster 7125'): 0.2915348185390512,\n",
       "  ('Cluster 893', 'Cluster 8038'): 0.2596692933427464},\n",
       " 'ethambutol': {('Cluster 2954', 'Cluster 521'): 0.14515518536136215,\n",
       "  ('Cluster 3849', 'Cluster 4660'): 0.15254716171741894,\n",
       "  ('Cluster 3909', 'Cluster 2954'): 0.17650700036379474,\n",
       "  ('Cluster 3909', 'Cluster 3849'): 0.22269462778344534,\n",
       "  ('Cluster 4244', 'Cluster 1039'): 0.14754847667668247,\n",
       "  ('Cluster 4244', 'Cluster 3176'): 0.20613618708864265,\n",
       "  ('Cluster 551', 'Cluster 1039'): 0.20392209650154644,\n",
       "  ('Cluster 551', 'Cluster 1396'): 0.17337788954268787,\n",
       "  ('Cluster 551', 'Cluster 2212'): 0.15874847305682094,\n",
       "  ('Cluster 551', 'Cluster 2954'): 0.3396306522934076,\n",
       "  ('Cluster 551', 'Cluster 3849'): 0.217356245368699,\n",
       "  ('Cluster 551', 'Cluster 3909'): 0.33628277066540313,\n",
       "  ('Cluster 551', 'Cluster 4244'): 0.3247330020054669,\n",
       "  ('Cluster 551', 'Cluster 4287'): 0.2348499190866419,\n",
       "  ('Cluster 551', 'Cluster 4417'): 0.16825180604134754,\n",
       "  ('Cluster 551', 'Cluster 521'): 0.3134184899515743,\n",
       "  ('Cluster 551', 'Cluster 6096'): 0.15933892219009926,\n",
       "  ('Cluster 551', 'Cluster 6118'): 0.17057313063343968,\n",
       "  ('Cluster 551', 'Cluster 969'): 0.1590612068094471,\n",
       "  ('Cluster 6118', 'Cluster 6096'): 0.18958262730077446},\n",
       " 'ethionamide': {('Cluster 2122', 'Cluster 2227'): 0.2855564699522895,\n",
       "  ('Cluster 2122', 'Cluster 2980'): 0.22988264090599614,\n",
       "  ('Cluster 2122', 'Cluster 3373'): 0.4296888472019012,\n",
       "  ('Cluster 2122', 'Cluster 3925'): 0.2451926181760136,\n",
       "  ('Cluster 2122', 'Cluster 5664'): 0.3777866594347404,\n",
       "  ('Cluster 2122', 'Cluster 6929'): 0.26008277305644995,\n",
       "  ('Cluster 2122', 'Cluster 7317'): 0.26433472839103833,\n",
       "  ('Cluster 2122', 'Cluster 9540'): 0.2657001828049426,\n",
       "  ('Cluster 2227', 'Cluster 6999'): 0.23611254070702606,\n",
       "  ('Cluster 2227', 'Cluster 7317'): 0.227410787673574,\n",
       "  ('Cluster 2980', 'Cluster 7362'): 0.2381007105614023,\n",
       "  ('Cluster 3373', 'Cluster 2227'): 0.2398858503478637,\n",
       "  ('Cluster 3373', 'Cluster 4070'): 0.2968958312842144,\n",
       "  ('Cluster 3373', 'Cluster 5664'): 0.2447987709698808,\n",
       "  ('Cluster 3373', 'Cluster 7317'): 0.4025333582505795,\n",
       "  ('Cluster 5593', 'Cluster 3925'): 0.3189700962131681,\n",
       "  ('Cluster 5593', 'Cluster 4070'): 0.22956991309298286,\n",
       "  ('Cluster 5593', 'Cluster 7362'): 0.284886251878436,\n",
       "  ('Cluster 7317', 'Cluster 7362'): 0.2545524678207745,\n",
       "  ('Cluster 9540', 'Cluster 6316'): 0.26233586290708005},\n",
       " 'isoniazid': {('Cluster 1116', 'Cluster 1237'): 0.3822040087101239,\n",
       "  ('Cluster 1116', 'Cluster 1258'): 0.2548593519571469,\n",
       "  ('Cluster 1116', 'Cluster 1396'): 0.3326845521649363,\n",
       "  ('Cluster 1116', 'Cluster 1601'): 0.2636750534128579,\n",
       "  ('Cluster 1116', 'Cluster 203'): 0.2306719650038284,\n",
       "  ('Cluster 1116', 'Cluster 2178'): 0.37691036553274215,\n",
       "  ('Cluster 1116', 'Cluster 3849'): 0.246215921719424,\n",
       "  ('Cluster 1116', 'Cluster 4498'): 0.4250544135087802,\n",
       "  ('Cluster 1116', 'Cluster 4738'): 0.4737769272525543,\n",
       "  ('Cluster 1116', 'Cluster 5285'): 0.23684563416761556,\n",
       "  ('Cluster 1116', 'Cluster 5659'): 0.27885778053891713,\n",
       "  ('Cluster 1116', 'Cluster 7026'): 0.24571820355125054,\n",
       "  ('Cluster 1237', 'Cluster 1258'): 0.236053897190466,\n",
       "  ('Cluster 1237', 'Cluster 4498'): 0.29417936079295964,\n",
       "  ('Cluster 2178', 'Cluster 4498'): 0.24040389468715384,\n",
       "  ('Cluster 4498', 'Cluster 5285'): 0.2527706997469589,\n",
       "  ('Cluster 4738', 'Cluster 1601'): 0.3101189253964035,\n",
       "  ('Cluster 4738', 'Cluster 2178'): 0.3058897593114856,\n",
       "  ('Cluster 4738', 'Cluster 4010'): 0.3218724856982849,\n",
       "  ('Cluster 4738', 'Cluster 4498'): 0.3610092936096218},\n",
       " 'ofloxacin': {('Cluster 2227', 'Cluster 4252'): 0.2477017249031455,\n",
       "  ('Cluster 2980', 'Cluster 5003'): 0.3167975369444971,\n",
       "  ('Cluster 4931', 'Cluster 2212'): 0.2791086911045591,\n",
       "  ('Cluster 5003', 'Cluster 4252'): 0.37175990880871795,\n",
       "  ('Cluster 5346', 'Cluster 2227'): 0.2692521162788322,\n",
       "  ('Cluster 5346', 'Cluster 4252'): 0.2871890119666135,\n",
       "  ('Cluster 5346', 'Cluster 6993'): 0.2479313603974468,\n",
       "  ('Cluster 5962', 'Cluster 4931'): 0.3023787409859545,\n",
       "  ('Cluster 5962', 'Cluster 4953'): 0.2637228610711749,\n",
       "  ('Cluster 893', 'Cluster 1163'): 0.2607758856323386,\n",
       "  ('Cluster 893', 'Cluster 1736'): 0.2800652760376305,\n",
       "  ('Cluster 893', 'Cluster 2212'): 0.297947730328164,\n",
       "  ('Cluster 893', 'Cluster 2227'): 0.2689115287786467,\n",
       "  ('Cluster 893', 'Cluster 2980'): 0.6212291354515127,\n",
       "  ('Cluster 893', 'Cluster 4476'): 0.3075443372713961,\n",
       "  ('Cluster 893', 'Cluster 4931'): 0.2653631623671479,\n",
       "  ('Cluster 893', 'Cluster 5003'): 0.3640083337624398,\n",
       "  ('Cluster 893', 'Cluster 5962'): 0.37083566333407103,\n",
       "  ('Cluster 893', 'Cluster 6770'): 0.30045445408953136,\n",
       "  ('Cluster 893', 'Cluster 6993'): 0.2360254316381424},\n",
       " 'pyrazinamide': {('Cluster 3074', 'Cluster 3682'): 0.1830667177205288,\n",
       "  ('Cluster 3074', 'Cluster 896'): 0.2788142411233915,\n",
       "  ('Cluster 3151', 'Cluster 1209'): 0.180812926119798,\n",
       "  ('Cluster 3151', 'Cluster 3935'): 0.18309603866336588,\n",
       "  ('Cluster 3289', 'Cluster 521'): 0.17853897293214102,\n",
       "  ('Cluster 3713', 'Cluster 643'): 0.2091855272722957,\n",
       "  ('Cluster 3930', 'Cluster 1784'): 0.16248086690353047,\n",
       "  ('Cluster 3930', 'Cluster 3713'): 0.16114765197026365,\n",
       "  ('Cluster 3930', 'Cluster 3935'): 0.18822664802344816,\n",
       "  ('Cluster 3930', 'Cluster 4054'): 0.2108681380326454,\n",
       "  ('Cluster 3930', 'Cluster 4459'): 0.17093764744562845,\n",
       "  ('Cluster 3935', 'Cluster 921'): 0.1822650351269968,\n",
       "  ('Cluster 4054', 'Cluster 4389'): 0.19172423226843432,\n",
       "  ('Cluster 4477', 'Cluster 4459'): 0.14750752734320716,\n",
       "  ('Cluster 4477', 'Cluster 643'): 0.20778838815100886,\n",
       "  ('Cluster 4606', 'Cluster 896'): 0.21720315407201574,\n",
       "  ('Cluster 4606', 'Cluster 921'): 0.15871407905355536,\n",
       "  ('Cluster 521', 'Cluster 1209'): 0.14693859177816973,\n",
       "  ('Cluster 521', 'Cluster 3074'): 0.1696339556189609,\n",
       "  ('Cluster 521', 'Cluster 4905'): 0.15591790243465645},\n",
       " 'rifampicin': {('Cluster 1206', 'Cluster 2117'): 0.24769629766842985,\n",
       "  ('Cluster 1237', 'Cluster 1068'): 0.2477501115334089,\n",
       "  ('Cluster 1237', 'Cluster 1396'): 0.26260480418745874,\n",
       "  ('Cluster 1237', 'Cluster 2117'): 0.2998346659903906,\n",
       "  ('Cluster 1237', 'Cluster 5025'): 0.2812968413690147,\n",
       "  ('Cluster 1237', 'Cluster 619'): 0.2017632633755999,\n",
       "  ('Cluster 2117', 'Cluster 2754'): 0.29167119057178315,\n",
       "  ('Cluster 4738', 'Cluster 3178'): 0.20383625867917746,\n",
       "  ('Cluster 491', 'Cluster 1068'): 0.3427140374802672,\n",
       "  ('Cluster 491', 'Cluster 1206'): 0.4737099530335359,\n",
       "  ('Cluster 491', 'Cluster 1237'): 0.4403043781085342,\n",
       "  ('Cluster 491', 'Cluster 2117'): 0.36276864302565615,\n",
       "  ('Cluster 491', 'Cluster 2754'): 0.2796924500840848,\n",
       "  ('Cluster 491', 'Cluster 3178'): 0.22496505253931284,\n",
       "  ('Cluster 491', 'Cluster 3865'): 0.3138238259858893,\n",
       "  ('Cluster 491', 'Cluster 415'): 0.23957586298596706,\n",
       "  ('Cluster 491', 'Cluster 4237'): 0.2047476871756909,\n",
       "  ('Cluster 491', 'Cluster 4738'): 0.2298548978345629,\n",
       "  ('Cluster 491', 'Cluster 5025'): 0.3865327987545475,\n",
       "  ('Cluster 491', 'Cluster 577'): 0.2613479177420209},\n",
       " 'streptomycin': {('Cluster 1237', 'Cluster 2032'): 0.2285736575227355,\n",
       "  ('Cluster 1237', 'Cluster 4389'): 0.2536539488071272,\n",
       "  ('Cluster 3010', 'Cluster 2032'): 0.2373346463639975,\n",
       "  ('Cluster 5240', 'Cluster 1884'): 0.3642145271996363,\n",
       "  ('Cluster 5240', 'Cluster 2979'): 0.33760073220929804,\n",
       "  ('Cluster 5240', 'Cluster 4980'): 0.2465711436652033,\n",
       "  ('Cluster 5240', 'Cluster 7636'): 0.22723460907928855,\n",
       "  ('Cluster 5965', 'Cluster 2979'): 0.23821111408774004,\n",
       "  ('Cluster 7435', 'Cluster 1206'): 0.2319206676333053,\n",
       "  ('Cluster 7435', 'Cluster 1237'): 0.34042642359571706,\n",
       "  ('Cluster 7435', 'Cluster 1884'): 0.2636359218313877,\n",
       "  ('Cluster 7435', 'Cluster 2633'): 0.3157850630823899,\n",
       "  ('Cluster 7435', 'Cluster 4056'): 0.2690517889852478,\n",
       "  ('Cluster 7435', 'Cluster 4389'): 0.2627578139763519,\n",
       "  ('Cluster 7435', 'Cluster 4980'): 0.35448440356507704,\n",
       "  ('Cluster 7435', 'Cluster 5240'): 0.5921029503297671,\n",
       "  ('Cluster 7435', 'Cluster 7636'): 0.3086336551320165,\n",
       "  ('Cluster 7636', 'Cluster 2032'): 0.24264475334460225,\n",
       "  ('Cluster 7636', 'Cluster 3010'): 0.26737019972278,\n",
       "  ('Cluster 7636', 'Cluster 4980'): 0.2834384794384045}}"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pairwise_table_groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# pairwise_table_groups = {\"pyrazinamide\": {('Cluster 3930', 'Cluster 3289'): 0.2327053738607056}}\n",
    "# pairwise_table_groups = {\"ethambutol\": {('Cluster 551', 'Cluster 4244'): 0.2327053738607056}}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Full run - make both tables and run the epistatic interaction significance tests."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "slideshow": {
     "slide_type": "-"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "0it [00:00, ?it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "S: 578 | R: 983\n",
      "rifampicin\n",
      "removing...  92 clusters\n",
      "rifampicin__Cluster491_Cluster2754_\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "svm_epistasis_funcs.py:533: FutureWarning: \n",
      "Passing list-likes to .loc or [] with any missing label will raise\n",
      "KeyError in the future, you can use .reindex() as an alternative.\n",
      "\n",
      "See the documentation here:\n",
      "http://pandas.pydata.org/pandas-docs/stable/indexing.html#deprecate-loc-reindex-listlike\n",
      "  two_allele_df = X_all.loc[:, [all_1,all_2]].copy()\n",
      "/anaconda2/lib/python2.7/site-packages/pandas/core/indexing.py:1367: FutureWarning: \n",
      "Passing list-likes to .loc or [] with any missing label will raise\n",
      "KeyError in the future, you can use .reindex() as an alternative.\n",
      "\n",
      "See the documentation here:\n",
      "http://pandas.pydata.org/pandas-docs/stable/indexing.html#deprecate-loc-reindex-listlike\n",
      "  return self._getitem_tuple(key)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "rifampicin__Cluster491_Cluster4738_\n",
      "rifampicin__Cluster491_Cluster1206_\n",
      "rifampicin__Cluster491_Cluster3865_\n",
      "rifampicin__Cluster1237_Cluster619_\n",
      "rifampicin__Cluster2117_Cluster2754_\n",
      "rifampicin__Cluster491_Cluster3178_\n",
      "rifampicin__Cluster491_Cluster415_\n",
      "rifampicin__Cluster491_Cluster577_\n",
      "rifampicin__Cluster1237_Cluster2117_\n",
      "rifampicin__Cluster491_Cluster5025_\n",
      "rifampicin__Cluster1206_Cluster2117_\n",
      "rifampicin__Cluster1237_Cluster1396_\n",
      "rifampicin__Cluster491_Cluster1068_\n",
      "rifampicin__Cluster4738_Cluster3178_\n",
      "rifampicin__Cluster1237_Cluster1068_\n",
      "rifampicin__Cluster1237_Cluster5025_\n",
      "rifampicin__Cluster491_Cluster4237_\n",
      "rifampicin__Cluster491_Cluster2117_\n",
      "rifampicin__Cluster491_Cluster1237_\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "1it [01:09, 69.66s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "S: 506 | R: 1057\n",
      "isoniazid\n",
      "removing...  107 clusters\n",
      "isoniazid__Cluster1116_Cluster1601_\n",
      "isoniazid__Cluster2178_Cluster4498_\n",
      "isoniazid__Cluster1116_Cluster203_\n",
      "isoniazid__Cluster1116_Cluster2178_\n",
      "isoniazid__Cluster4738_Cluster4498_\n",
      "isoniazid__Cluster4738_Cluster1601_\n",
      "isoniazid__Cluster1237_Cluster1258_\n",
      "isoniazid__Cluster1116_Cluster4498_\n",
      "isoniazid__Cluster4738_Cluster2178_\n",
      "isoniazid__Cluster1116_Cluster1396_\n",
      "isoniazid__Cluster1116_Cluster3849_\n",
      "isoniazid__Cluster4498_Cluster5285_\n",
      "isoniazid__Cluster1116_Cluster1258_\n",
      "isoniazid__Cluster1116_Cluster5285_\n",
      "isoniazid__Cluster4738_Cluster4010_\n",
      "isoniazid__Cluster1116_Cluster1237_\n",
      "isoniazid__Cluster1116_Cluster4738_\n",
      "isoniazid__Cluster1116_Cluster7026_\n",
      "isoniazid__Cluster1237_Cluster4498_\n",
      "isoniazid__Cluster1116_Cluster5659_\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2it [01:58, 59.47s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "S: 92 | R: 137\n",
      "pyrazinamide\n",
      "removing...  97 clusters\n",
      "pyrazinamide__Cluster3074_Cluster896_\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "svm_epistasis_funcs.py:348: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  resist_string = \"(\"+resist_string +\")\"+ \": \" + str(round(resist_alleles[allele].sum()/float(allele_df[allele].sum()),2))\n",
      "svm_epistasis_funcs.py:351: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  allele_to_resistPercent_dict.update({allele: round(resist_alleles[allele].sum()/float(allele_df[allele].sum()),2)})\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pyrazinamide__Cluster3074_Cluster3682_\n",
      "pyrazinamide__Cluster3151_Cluster3935_\n",
      "pyrazinamide__Cluster3930_Cluster1784_\n",
      "pyrazinamide__Cluster4477_Cluster4459_\n",
      "pyrazinamide__Cluster3935_Cluster921_\n",
      "pyrazinamide__Cluster3151_Cluster1209_\n",
      "pyrazinamide__Cluster3930_Cluster3935_\n",
      "pyrazinamide__Cluster4477_Cluster643_\n",
      "pyrazinamide__Cluster3930_Cluster4459_\n",
      "pyrazinamide__Cluster3930_Cluster3713_\n",
      "pyrazinamide__Cluster521_Cluster1209_\n",
      "pyrazinamide__Cluster3930_Cluster4054_\n",
      "pyrazinamide__Cluster4606_Cluster896_\n",
      "pyrazinamide__Cluster4606_Cluster921_\n",
      "pyrazinamide__Cluster3289_Cluster521_\n",
      "pyrazinamide__Cluster3713_Cluster643_\n",
      "pyrazinamide__Cluster521_Cluster4905_\n",
      "pyrazinamide__Cluster521_Cluster3074_\n",
      "pyrazinamide__Cluster4054_Cluster4389_\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "3it [02:29, 49.69s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "S: 353 | R: 209\n",
      "ethionamide\n",
      "removing...  101 clusters\n",
      "ethionamide__Cluster5593_Cluster3925_\n",
      "ethionamide__Cluster3373_Cluster4070_\n",
      "ethionamide__Cluster2122_Cluster9540_\n",
      "ethionamide__Cluster5593_Cluster4070_\n",
      "ethionamide__Cluster3373_Cluster7317_\n",
      "ethionamide__Cluster2122_Cluster2980_\n",
      "ethionamide__Cluster2227_Cluster7317_\n",
      "ethionamide__Cluster3373_Cluster2227_\n",
      "ethionamide__Cluster7317_Cluster7362_\n",
      "ethionamide__Cluster2122_Cluster6929_\n",
      "ethionamide__Cluster3373_Cluster5664_\n",
      "ethionamide__Cluster2122_Cluster7317_\n",
      "ethionamide__Cluster2122_Cluster2227_\n",
      "ethionamide__Cluster2122_Cluster3373_\n",
      "ethionamide__Cluster2122_Cluster5664_\n",
      "ethionamide__Cluster5593_Cluster7362_\n",
      "ethionamide__Cluster2227_Cluster6999_\n",
      "ethionamide__Cluster2122_Cluster3925_\n",
      "ethionamide__Cluster2980_Cluster7362_\n",
      "ethionamide__Cluster9540_Cluster6316_\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "4it [03:02, 45.55s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "S: 295 | R: 80\n",
      "4-aminosalicylic_acid\n",
      "removing...  109 clusters\n",
      "4-aminosalicylic_acid__Cluster1404_Cluster3289_\n",
      "PAS__Cluster2211_Cluster4806_\n",
      "PAS__Cluster3316_Cluster4053_\n",
      "PAS__Cluster4053_Cluster2698_\n",
      "PAS__Cluster594_Cluster5949_\n",
      "PAS__Cluster4603_Cluster4536_\n",
      "PAS__Cluster7466_Cluster5385_\n",
      "PAS__Cluster5949_Cluster4603_\n",
      "PAS__Cluster8101_Cluster2331_\n",
      "PAS__Cluster5949_Cluster4486_\n",
      "PAS__Cluster4603_Cluster2698_\n",
      "PAS__Cluster4486_Cluster1404_\n",
      "PAS__Cluster4806_Cluster1916_\n",
      "PAS__Cluster2954_Cluster3289_\n",
      "PAS__Cluster4486_Cluster594_\n",
      "PAS__Cluster8101_Cluster4806_\n",
      "PAS__Cluster7466_Cluster4053_\n",
      "PAS__Cluster2954_Cluster3005_\n",
      "PAS__Cluster2954_Cluster2698_\n",
      "PAS__Cluster2211_Cluster1227_\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "5it [03:28, 41.77s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "S: 848 | R: 492\n",
      "ethambutol\n",
      "removing...  99 clusters\n",
      "ethambutol__Cluster551_Cluster521_\n",
      "ethambutol__Cluster4244_Cluster1039_\n",
      "ethambutol__Cluster551_Cluster2212_\n",
      "ethambutol__Cluster3909_Cluster2954_\n",
      "ethambutol__Cluster2954_Cluster521_\n",
      "ethambutol__Cluster551_Cluster4287_\n",
      "ethambutol__Cluster551_Cluster6118_\n",
      "ethambutol__Cluster551_Cluster6096_\n",
      "ethambutol__Cluster551_Cluster3849_\n",
      "ethambutol__Cluster551_Cluster1396_\n",
      "ethambutol__Cluster551_Cluster969_\n",
      "ethambutol__Cluster551_Cluster1039_\n",
      "ethambutol__Cluster551_Cluster4244_\n",
      "ethambutol__Cluster3909_Cluster3849_\n",
      "ethambutol__Cluster551_Cluster4417_\n",
      "ethambutol__Cluster551_Cluster3909_\n",
      "ethambutol__Cluster551_Cluster2954_\n",
      "ethambutol__Cluster4244_Cluster3176_\n",
      "ethambutol__Cluster6118_Cluster6096_\n",
      "ethambutol__Cluster3849_Cluster4660_\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "6it [04:20, 43.42s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# RR 62\n",
      "# MDR 1136\n",
      "# XDR 192\n",
      "number of  XDR   192\n",
      "number of strains available for classification 1565\n",
      "186\n",
      "XDR__Cluster893_Cluster7125_\n",
      "XDR__Cluster893_Cluster5948_\n",
      "XDR__Cluster893_Cluster8038_\n",
      "XDR__Cluster491_Cluster4244_\n",
      "XDR__Cluster491_Cluster3930_\n",
      "XDR__Cluster4244_Cluster3930_\n",
      "XDR__Cluster491_Cluster3909_\n",
      "XDR__Cluster8038_Cluster4244_\n",
      "XDR__Cluster893_Cluster1116_\n",
      "XDR__Cluster893_Cluster4486_\n",
      "XDR__Cluster893_Cluster2122_\n",
      "XDR__Cluster893_Cluster3930_\n",
      "XDR__Cluster893_Cluster1058_\n",
      "XDR__Cluster4476_Cluster4244_\n",
      "XDR__Cluster7926_Cluster4244_\n",
      "XDR__Cluster4244_Cluster7125_\n",
      "XDR__Cluster893_Cluster491_\n",
      "XDR__Cluster3930_Cluster5948_\n",
      "XDR__Cluster893_Cluster4244_\n",
      "XDR__Cluster3930_Cluster2122_\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "7it [05:21, 45.88s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# RR 62\n",
      "# MDR 1136\n",
      "# XDR 192\n",
      "number of  MDR   1328\n",
      "number of strains available for classification 1565\n",
      "946\n",
      "MDR__Cluster415_Cluster4244_\n",
      "MDR__Cluster1116_Cluster893_\n",
      "MDR__Cluster2122_Cluster5240_\n",
      "MDR__Cluster7435_Cluster2178_\n",
      "MDR__Cluster491_Cluster3930_\n",
      "MDR__Cluster491_Cluster893_\n",
      "MDR__Cluster1116_Cluster551_\n",
      "MDR__Cluster491_Cluster1116_\n",
      "MDR__Cluster3930_Cluster3881_\n",
      "MDR__Cluster1116_Cluster415_\n",
      "MDR__Cluster1116_Cluster2122_\n",
      "MDR__Cluster551_Cluster893_\n",
      "MDR__Cluster491_Cluster3881_\n",
      "MDR__Cluster1116_Cluster5240_\n",
      "MDR__Cluster893_Cluster3881_\n",
      "MDR__Cluster3930_Cluster551_\n",
      "MDR__Cluster1116_Cluster3881_\n",
      "MDR__Cluster5240_Cluster3881_\n",
      "MDR__Cluster1116_Cluster4244_\n",
      "MDR__Cluster415_Cluster7435_\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "8it [06:39, 49.98s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "S: 732 | R: 663\n",
      "streptomycin\n",
      "removing...  98 clusters\n",
      "streptomycin__Cluster7435_Cluster4389_\n",
      "streptomycin__Cluster1237_Cluster4389_\n",
      "streptomycin__Cluster7435_Cluster5240_\n",
      "streptomycin__Cluster5240_Cluster4980_\n",
      "streptomycin__Cluster7435_Cluster1206_\n",
      "streptomycin__Cluster7435_Cluster1884_\n",
      "streptomycin__Cluster7435_Cluster4980_\n",
      "streptomycin__Cluster5240_Cluster1884_\n",
      "streptomycin__Cluster7636_Cluster2032_\n",
      "streptomycin__Cluster7435_Cluster2633_\n",
      "streptomycin__Cluster1237_Cluster2032_\n",
      "streptomycin__Cluster7435_Cluster7636_\n",
      "streptomycin__Cluster5240_Cluster2979_\n",
      "streptomycin__Cluster7636_Cluster4980_\n",
      "streptomycin__Cluster7435_Cluster4056_\n",
      "streptomycin__Cluster3010_Cluster2032_\n",
      "streptomycin__Cluster7435_Cluster1237_\n",
      "streptomycin__Cluster7636_Cluster3010_\n",
      "streptomycin__Cluster5240_Cluster7636_\n",
      "streptomycin__Cluster5965_Cluster2979_\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "9it [07:44, 51.61s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "S: 554 | R: 302\n",
      "ofloxacin\n",
      "removing...  102 clusters\n",
      "ofloxacin__Cluster893_Cluster4476_\n",
      "ofloxacin__Cluster893_Cluster2227_\n",
      "ofloxacin__Cluster2980_Cluster5003_\n",
      "ofloxacin__Cluster5003_Cluster4252_\n",
      "ofloxacin__Cluster893_Cluster5962_\n",
      "ofloxacin__Cluster893_Cluster1736_\n",
      "ofloxacin__Cluster5962_Cluster4953_\n",
      "ofloxacin__Cluster893_Cluster4931_\n",
      "ofloxacin__Cluster893_Cluster5003_\n",
      "ofloxacin__Cluster893_Cluster1163_\n",
      "ofloxacin__Cluster5346_Cluster6993_\n",
      "ofloxacin__Cluster893_Cluster6770_\n",
      "ofloxacin__Cluster5346_Cluster2227_\n",
      "ofloxacin__Cluster893_Cluster2980_\n",
      "ofloxacin__Cluster5346_Cluster4252_\n",
      "ofloxacin__Cluster4931_Cluster2212_\n",
      "ofloxacin__Cluster5962_Cluster4931_\n",
      "ofloxacin__Cluster2227_Cluster4252_\n",
      "ofloxacin__Cluster893_Cluster6993_\n",
      "ofloxacin__Cluster893_Cluster2212_\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "10it [08:27, 50.71s/it]\n"
     ]
    }
   ],
   "source": [
    "antibiotic_gene_gene_interactions = {}\n",
    "clust_alleles = [\"what\"] # dummy variable\n",
    "\n",
    "for antibiotic_name, cluster_groups in tqdm(pairwise_table_groups.iteritems()):\n",
    "    \n",
    "    antibiotic_gene_gene_interactions[antibiotic_name] = {}\n",
    "    \n",
    "    row_index_location = 3\n",
    "    \n",
    "    # \", \"ethionamide\", \"ofloxacin\", \"amikacin\" \"4-aminosalicylic_acid\"\n",
    "    if antibiotic_name == \"MDR\" or antibiotic_name == \"XDR\":\n",
    "        X1, y1 = get_MDR_training(resistance_data, data_to_plot, antibiotic_name)\n",
    "    else:\n",
    "        y1, X1 = get_target_data(data_to_plot, resistance_data, antibiotic_name)\n",
    "        X1 = remove_key_antibiotic_clusters(X1, antibiotic_name)\n",
    "        \n",
    "    \n",
    "    for (clust1, clust2) in cluster_groups.keys():\n",
    "        \n",
    "        rv_id_1, rv_id_2 = clust_to_rv[clust1], clust_to_rv[clust2]\n",
    "        clust1 = clust1.replace(\"ter \", \"ter\")+\"_\"\n",
    "        clust2 = clust2.replace(\"ter \", \"ter\")+\"_\"\n",
    "        # write code for specific case\n",
    "        \n",
    "        two_clust_df, two_clustPercent_df = get_2allele_resist_dataframes(clust1, clust2, antibiotic_name, \"All\",\n",
    "                                                                          y1,pan_variant_df.loc[y1.index, :], \n",
    "                                                                          pan_variant_df, clust_alleles, False, \n",
    "                                                                          COOCCURANCE_TABLE_EXCEL_DIR)\n",
    "        if color_measure == \"oddsRatio\":\n",
    "            two_alleleOdds_df = return_2allele_oddsRatio_df(two_clustPercent_df, two_clust_df, \n",
    "                                           pan_variant_df.loc[y1.index, :], y1, add_zeroCell, log_or_no=True)\n",
    "            two_allelePval_df = pd.DataFrame()\n",
    "        \n",
    "        # note that the last input is either pVal or percent depending on how to color the cell.\n",
    "        if antibiotic_name == '4-aminosalicylic_acid':\n",
    "            antibiotic_name = \"PAS\"\n",
    "        save_occur_df, labeled_2cluster_df = make_table_figure(clust1, clust2, antibiotic_name, two_clust_df, \n",
    "                                                               two_clustPercent_df, \n",
    "                                                               two_allelePval_df, two_alleleOdds_df,\n",
    "                                                               clust_to_rv, \n",
    "                                                               rv_to_name, clster_manual_map, color_measure,\n",
    "                                                               COOCCURANCE_TABLE_FIGURES_DIR,\n",
    "                                                               color_bar_or_not = True)\n",
    "        \n",
    "        X_alleles = [x for x in two_alleleOdds_df.index if x[-1]!=\"_\"]\n",
    "        Y_alleles = [x for x in two_alleleOdds_df.columns if x[-1]!=\"_\"]\n",
    "        \n",
    "        X1 = X1.fillna(0)\n",
    "        two_gene_interact_alleles = X1.loc[:,X_alleles+Y_alleles].copy()\n",
    "\n",
    "        # ---- start loop here.... -----\n",
    "#         interaction_terms_list = []\n",
    "#         eq_terms = []\n",
    "#         for allele_x in X_alleles:\n",
    "#             for allele_y in Y_alleles:\n",
    "#                 if two_alleleOdds_df.loc[allele_x, allele_y]!=0:\n",
    "#                     interaction_terms_list.append(allele_x+\":\"+allele_y)\n",
    "                    \n",
    "#                     # --- Write allele-allele logistic regression: \n",
    "#                     # ---- Y ~ a + a1*x1 + a2*x2 + a3*x1*x2\n",
    "#                     eq_terms.extend([allele_x+\":\"+allele_y])\n",
    "#                     # logReg_string_equation = allele_x + \" + \" + allele_y + \" + \" + allele_x+\":\"+allele_y\n",
    "#                 eq_terms.extend([allele_x, allele_y])\n",
    "\n",
    "#         eq_terms = list(set(eq_terms))\n",
    "#         logReg_string_equation = \" + \".join(eq_terms)\n",
    "#         logReg_string_equation = logReg_string_equation.strip(\" + \")\n",
    "#         logReg_string_equation = \"AMR_phenotype ~ \"+logReg_string_equation\n",
    "#         # print logReg_string_equation\n",
    "\n",
    "\n",
    "        # ---- start loop here.... -----\n",
    "        interaction_terms_list = []\n",
    "        for allele_x in X_alleles:\n",
    "            for allele_y in Y_alleles:\n",
    "                if two_alleleOdds_df.loc[allele_x, allele_y]!=0:\n",
    "                    interaction_terms_list.append(allele_x+\":\"+allele_y)\n",
    "\n",
    "        logReg_string_equation = \"\"\n",
    "        for col in two_gene_interact_alleles.columns:\n",
    "            logReg_string_equation = logReg_string_equation + col + \" + \"\n",
    "        for interact_term in interaction_terms_list:\n",
    "            logReg_string_equation = logReg_string_equation + interact_term + \" + \"\n",
    "\n",
    "        logReg_string_equation = logReg_string_equation.strip(\" + \")\n",
    "        logReg_string_equation = \"AMR_phenotype ~ \"+logReg_string_equation\n",
    "\n",
    "        \n",
    "        # \n",
    "        two_gene_interact_alleles[\"AMR_phenotype\"] = y1\n",
    "\n",
    "        # Use patsy package to turn this string equation into something usable by python.\n",
    "        y_logReg, X_logReg = patsy.dmatrices(logReg_string_equation, two_gene_interact_alleles)\n",
    "\n",
    "        # print \n",
    "        # print \"single:\", len(two_gene_interact_alleles.columns), \"| interactions:\", len(interaction_terms_list)\n",
    "        \n",
    "        y_logReg_df = pd.DataFrame(np.asarray(y_logReg), index = y1.index, columns=[\"AMR_phenotype\"])\n",
    "        X_logReg_df = pd.DataFrame(np.asarray(X_logReg), index = X1.index, \n",
    "                                   columns=X_logReg.design_info.column_names)\n",
    "        X_logReg_df.drop([\"Intercept\"], axis=1, inplace=True)\n",
    "        \n",
    "        # ----- GET RID OF MAJOR COLLINEARITY VARIABLES-----\n",
    "        drop_co_alleles = drop_collinearity_variables(X_alleles, Y_alleles, interaction_terms_list, X_logReg_df, False)\n",
    "        # print X_logReg_df.shape\n",
    "        X_logReg_df.drop(drop_co_alleles, axis=1, inplace=True)\n",
    "        # print X_logReg_df.shape\n",
    "        \n",
    "        try:\n",
    "            # Since each optimization inside the profile likelihood generates a warning \n",
    "            # (due to the random slope variance being close to zero), we turn off the warnings here.\n",
    "            with warnings.catch_warnings():\n",
    "                warnings.filterwarnings(\"ignore\")\n",
    "                logit = sm.Logit(y_logReg_df[\"AMR_phenotype\"], X_logReg_df)\n",
    "                result = logit.fit(maxiter=100, method='bfgs',disp=False)\n",
    "\n",
    "            pVal_thresh_series = result.pvalues\n",
    "            pVal_thresh_df = pd.DataFrame(pVal_thresh_series)\n",
    "            pVal_thresh_df.columns = [\"pVal\"]\n",
    "            pVal_thresh_df[\"aic\"], pVal_thresh_df[\"bic\"], pVal_thresh_df[\"pseudo_r2\"] = result.aic, result.bic, result.prsquared\n",
    "            antibiotic_gene_gene_interactions[antibiotic_name][(clust1, clust2)] = dict(pVal_thresh_df.T)\n",
    "            # antibiotic_gene_gene_interactions[antibiotic_name][(clust1, clust2)] = dict(result.pvalues[result.pvalues<pVal_threshold_epistasis])\n",
    "        except:\n",
    "            continue\n",
    "            # print \"optimization error... likely collinearity...\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- at one run... ['Cluster3930_12', 'Cluster3930_7', 'Cluster3930_9', 'Cluster3930_14', 'Cluster3930_2', 'Cluster3930_5', 'Cluster3930_6']\n",
    "- ['Cluster3930_12', 'Cluster3930_7', 'Cluster3930_9', 'Cluster3289_3', 'Cluster3930_11', 'Cluster3930_14', 'Cluster3930_2', 'Cluster3930_4', 'Cluster3930_5', 'Cluster3930_6']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# params = result.params\n",
    "# conf = result.conf_int()\n",
    "# conf['OR'] = params\n",
    "# conf[\"pVals\"] = result.pvalues\n",
    "# conf.columns = ['2.5%', '97.5%', 'OR', \"pVals\"]\n",
    "# conf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "rifampicin\n",
      "# of unique gene-gene interactions 8\n",
      "# of unique allele-allele interactions 12\n",
      "isoniazid\n",
      "# of unique gene-gene interactions 12\n",
      "# of unique allele-allele interactions 27\n",
      "pyrazinamide\n",
      "# of unique gene-gene interactions 3\n",
      "# of unique allele-allele interactions 6\n",
      "ethionamide\n",
      "# of unique gene-gene interactions 6\n",
      "# of unique allele-allele interactions 7\n",
      "ethambutol\n",
      "# of unique gene-gene interactions 8\n",
      "# of unique allele-allele interactions 11\n",
      "XDR\n",
      "# of unique gene-gene interactions 17\n",
      "# of unique allele-allele interactions 90\n",
      "MDR\n",
      "# of unique gene-gene interactions 17\n",
      "# of unique allele-allele interactions 64\n",
      "streptomycin\n",
      "# of unique gene-gene interactions 12\n",
      "# of unique allele-allele interactions 38\n",
      "ofloxacin\n",
      "# of unique gene-gene interactions 15\n",
      "# of unique allele-allele interactions 52\n",
      "----------------------------------------\n",
      "# of unique gene-gene interactions 97\n",
      "# of unique allele-allele interactions 306\n"
     ]
    }
   ],
   "source": [
    "# new_pVal_thresh = 0.005\n",
    "clust_pair_pVal_sum_dict = {}\n",
    "epistasis_final_dict = {}\n",
    "tmp_df = pd.DataFrame()\n",
    "\n",
    "all_Antibiotic_tmp_df = pd.DataFrame()\n",
    "all_Antibiotic_pVal_correct_df = pd.DataFrame()\n",
    "\n",
    "interaction_count = 0\n",
    "for antibiotic, cluster_pair_dict in antibiotic_gene_gene_interactions.iteritems():\n",
    "    if len(cluster_pair_dict)!=0:\n",
    "        rv_pair_list, name_pair_list = [], []\n",
    "        print antibiotic\n",
    "        clust_pair_pVal_sum = 0\n",
    "        epistasis_final_dict[antibiotic] = {}\n",
    "        tmp_df = pd.DataFrame()\n",
    "        for (clst1, clst2), interaction_pair_dict in cluster_pair_dict.iteritems():\n",
    "            # add_to_clusterPairdict = False\n",
    "\n",
    "            for allele_allele, row_info in interaction_pair_dict.iteritems():\n",
    "                pVal = row_info[\"pVal\"]\n",
    "                # if \":\" in allele_allele #and pVal<new_pVal_thresh:\n",
    "\n",
    "                if (clst1, clst2) not in epistasis_final_dict[antibiotic].keys():\n",
    "                    epistasis_final_dict[antibiotic][(clst1, clst2)] = {allele_allele: pVal}\n",
    "                else:\n",
    "                    epistasis_final_dict[antibiotic][(clst1, clst2)].update({allele_allele: pVal})\n",
    "\n",
    "                if \":\" in allele_allele:\n",
    "                    single_or_double_allele = \"double\"\n",
    "                    rv_id1= clust_to_rv[allele_allele.split(\":\")[0].split(\"_\")[0].replace(\"ter\", \"ter \")]\n",
    "                    rv_id2= clust_to_rv[allele_allele.split(\":\")[1].split(\"_\")[0].replace(\"ter\", \"ter \")]\n",
    "                else:\n",
    "                    single_or_double_allele = \"single\"\n",
    "                # print (rv_id1, rv_to_name[rv_id1], \"||\",rv_id2, rv_to_name[rv_id2])\n",
    "\n",
    "                if single_or_double_allele != \"single\":\n",
    "\n",
    "                    if (rv_id1, rv_id2) not in rv_pair_list:\n",
    "                        rv_pair_list.append((antibiotic, clst1, clst2 , str(rv_id1), str(rv_id2)))\n",
    "                        name_pair_list.append((antibiotic, clst1, clst2, str(rv_to_name[rv_id1]), str(rv_to_name[rv_id2])))\n",
    "                        rv_pair_list, name_pair_list = list(set(rv_pair_list)), list(set(name_pair_list))\n",
    "\n",
    "                    rv_NAME1, rv_NAME2 = str(rv_to_name[rv_id1]), str(rv_to_name[rv_id2])\n",
    "                    if rv_NAME1 == \"None\" and rv_id1 in clster_manual_map.keys():\n",
    "                        rv_NAME1 = clster_manual_map[rv_id1]\n",
    "                    elif rv_NAME1 == \"None\":\n",
    "                        rv_NAME1 = rv_id1\n",
    "\n",
    "                    if rv_NAME2 == \"None\" and rv_id2 in clster_manual_map.keys():\n",
    "                        rv_NAME2 = clster_manual_map[rv_id2]\n",
    "                    elif rv_NAME2 == \"None\":\n",
    "                        rv_NAME2 = rv_id2\n",
    "\n",
    "                    epistasis_row_df = pd.DataFrame({\"antibiotic\": antibiotic, \n",
    "                                                     \"allele_interaction\": allele_allele, \n",
    "                                                     \"single_or_double\": single_or_double_allele, \n",
    "                                                     \"pValue\": pVal, \n",
    "                                                     \"AIC\": row_info[\"aic\"],\n",
    "                                                     \"BIC\": row_info[\"bic\"],\n",
    "                                                     \"pseudo_r2\": row_info[\"pseudo_r2\"],\n",
    "                                                     \"geneID_interact\": rv_id1+\":\"+rv_id2, \n",
    "                                    \"name_interact\":rv_NAME1+\":\"+rv_NAME2},\n",
    "                                    index=[interaction_count])\n",
    "                    tmp_df = pd.concat([tmp_df, epistasis_row_df])\n",
    "                    interaction_count = interaction_count + 1\n",
    "\n",
    "        tmp_df = tmp_df[[\"antibiotic\",\"allele_interaction\", \"geneID_interact\",\"single_or_double\", \n",
    "                         \"name_interact\", \"pValue\", \"AIC\", \"BIC\", \"pseudo_r2\"]]\n",
    "\n",
    "        # Correct for multiple hypothesis testing of the interaction variables in the logistic regression model. # positively correlated # negcorr\n",
    "        # rejected_list, pvalue_corrected_list = statsmodels.sandbox.stats.multicomp.fdrcorrection0(tmp_df[\"pValue\"], alpha=0.5, method='indep', is_sorted=False)\n",
    "        # tmp_df = tmp_df[tmp_df[\"pValue\"]<.05].copy()\n",
    "        # tmp_df.reset_index(inplace=True)\n",
    "        rejected_list, pvalue_corrected_list = statsmodels.sandbox.stats.multicomp.fdrcorrection0(tmp_df[\"pValue\"], \n",
    "                                                                                                  alpha=0.5, method='indep', is_sorted=False)\n",
    "\n",
    "        tmp_df_pVal_correct = tmp_df[rejected_list].copy()\n",
    "        tmp_df_pVal_correct[\"corrected_pVal\"] = pvalue_corrected_list[rejected_list]\n",
    "        tmp_df_pVal_correct_interact = tmp_df_pVal_correct[tmp_df_pVal_correct[\"single_or_double\"]==\"double\"].copy()\n",
    "        tmp_df_pVal_correct_interact = tmp_df_pVal_correct_interact.reset_index()\n",
    "        tmp_df_pVal_correct_interact.drop([\"index\"], axis=1, inplace=True)\n",
    "        print \"# of unique gene-gene interactions\", len(tmp_df_pVal_correct_interact[\"geneID_interact\"].unique())\n",
    "        print \"# of unique allele-allele interactions\", len(tmp_df_pVal_correct_interact[\"allele_interaction\"].unique())\n",
    "\n",
    "        all_Antibiotic_tmp_df = pd.concat([tmp_df, all_Antibiotic_tmp_df])\n",
    "        all_Antibiotic_pVal_correct_df = pd.concat([tmp_df_pVal_correct_interact, all_Antibiotic_pVal_correct_df])\n",
    "\n",
    "save_interact = FINAL_SUPP_DIR+\"allele_epistasis_\"+str(top_allele_pairs)+\"_df.csv\"\n",
    "all_Antibiotic_pVal_correct_df.to_csv(save_interact)\n",
    "print \"----------------------------------------\"\n",
    "print \"# of unique gene-gene interactions\", len(all_Antibiotic_pVal_correct_df[\"geneID_interact\"].unique())\n",
    "print \"# of unique allele-allele interactions\", len(all_Antibiotic_pVal_correct_df[\"allele_interaction\"].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 299,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>antibiotic</th>\n",
       "      <th>allele_interaction</th>\n",
       "      <th>geneID_interact</th>\n",
       "      <th>single_or_double</th>\n",
       "      <th>name_interact</th>\n",
       "      <th>pValue</th>\n",
       "      <th>AIC</th>\n",
       "      <th>BIC</th>\n",
       "      <th>pseudo_r2</th>\n",
       "      <th>corrected_pVal</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ethionamide</td>\n",
       "      <td>Cluster5664_2:Cluster2122_10</td>\n",
       "      <td>Rv3421c:Rv3854c</td>\n",
       "      <td>double</td>\n",
       "      <td>Rv3421c:ethA</td>\n",
       "      <td>0.008549</td>\n",
       "      <td>707.098627</td>\n",
       "      <td>815.386173</td>\n",
       "      <td>0.114167</td>\n",
       "      <td>0.355388</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ethionamide</td>\n",
       "      <td>Cluster5664_1:Cluster2122_10</td>\n",
       "      <td>Rv3421c:Rv3854c</td>\n",
       "      <td>double</td>\n",
       "      <td>Rv3421c:ethA</td>\n",
       "      <td>0.009491</td>\n",
       "      <td>707.098627</td>\n",
       "      <td>815.386173</td>\n",
       "      <td>0.114167</td>\n",
       "      <td>0.355388</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ethionamide</td>\n",
       "      <td>Cluster5664_2:Cluster3373_1</td>\n",
       "      <td>Rv3421c:Rv1184c</td>\n",
       "      <td>double</td>\n",
       "      <td>Rv3421c:Rv1184c</td>\n",
       "      <td>0.010203</td>\n",
       "      <td>707.149272</td>\n",
       "      <td>767.790298</td>\n",
       "      <td>0.084440</td>\n",
       "      <td>0.355388</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ethionamide</td>\n",
       "      <td>Cluster7362_2:Cluster5593_7</td>\n",
       "      <td>Rv1690:Rv1222</td>\n",
       "      <td>double</td>\n",
       "      <td>lprJ:rseA</td>\n",
       "      <td>0.003981</td>\n",
       "      <td>743.437566</td>\n",
       "      <td>804.078592</td>\n",
       "      <td>0.035520</td>\n",
       "      <td>0.277359</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ethionamide</td>\n",
       "      <td>Cluster2980_4:Cluster2122_9</td>\n",
       "      <td>Rv2780:Rv3854c</td>\n",
       "      <td>double</td>\n",
       "      <td>ald:ethA</td>\n",
       "      <td>0.000409</td>\n",
       "      <td>700.001648</td>\n",
       "      <td>743.316666</td>\n",
       "      <td>0.083291</td>\n",
       "      <td>0.047687</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ethionamide</td>\n",
       "      <td>Cluster7362_2:Cluster2980_4</td>\n",
       "      <td>Rv1690:Rv2780</td>\n",
       "      <td>double</td>\n",
       "      <td>lprJ:ald</td>\n",
       "      <td>0.000456</td>\n",
       "      <td>727.667170</td>\n",
       "      <td>770.982189</td>\n",
       "      <td>0.045995</td>\n",
       "      <td>0.047687</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    antibiotic            allele_interaction  geneID_interact  \\\n",
       "0  ethionamide  Cluster5664_2:Cluster2122_10  Rv3421c:Rv3854c   \n",
       "1  ethionamide  Cluster5664_1:Cluster2122_10  Rv3421c:Rv3854c   \n",
       "2  ethionamide   Cluster5664_2:Cluster3373_1  Rv3421c:Rv1184c   \n",
       "3  ethionamide   Cluster7362_2:Cluster5593_7    Rv1690:Rv1222   \n",
       "4  ethionamide   Cluster2980_4:Cluster2122_9   Rv2780:Rv3854c   \n",
       "5  ethionamide   Cluster7362_2:Cluster2980_4    Rv1690:Rv2780   \n",
       "\n",
       "  single_or_double    name_interact    pValue         AIC         BIC  \\\n",
       "0           double     Rv3421c:ethA  0.008549  707.098627  815.386173   \n",
       "1           double     Rv3421c:ethA  0.009491  707.098627  815.386173   \n",
       "2           double  Rv3421c:Rv1184c  0.010203  707.149272  767.790298   \n",
       "3           double        lprJ:rseA  0.003981  743.437566  804.078592   \n",
       "4           double         ald:ethA  0.000409  700.001648  743.316666   \n",
       "5           double         lprJ:ald  0.000456  727.667170  770.982189   \n",
       "\n",
       "   pseudo_r2  corrected_pVal  \n",
       "0   0.114167        0.355388  \n",
       "1   0.114167        0.355388  \n",
       "2   0.084440        0.355388  \n",
       "3   0.035520        0.277359  \n",
       "4   0.083291        0.047687  \n",
       "5   0.045995        0.047687  "
      ]
     },
     "execution_count": 299,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_Antibiotic_pVal_correct_df[all_Antibiotic_pVal_correct_df[\"antibiotic\"]==\"ethionamide\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>antibiotic</th>\n",
       "      <th>allele_interaction</th>\n",
       "      <th>geneID_interact</th>\n",
       "      <th>single_or_double</th>\n",
       "      <th>name_interact</th>\n",
       "      <th>pValue</th>\n",
       "      <th>AIC</th>\n",
       "      <th>BIC</th>\n",
       "      <th>pseudo_r2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1019</th>\n",
       "      <td>ethambutol</td>\n",
       "      <td>Cluster3849_2:Cluster3909_2</td>\n",
       "      <td>Rv0265c:Rv3848</td>\n",
       "      <td>double</td>\n",
       "      <td>Rv0265c:Rv3848</td>\n",
       "      <td>0.012672</td>\n",
       "      <td>1716.306393</td>\n",
       "      <td>1763.110217</td>\n",
       "      <td>0.036099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1020</th>\n",
       "      <td>ethambutol</td>\n",
       "      <td>Cluster3849_2:Cluster3909_3</td>\n",
       "      <td>Rv0265c:Rv3848</td>\n",
       "      <td>double</td>\n",
       "      <td>Rv0265c:Rv3848</td>\n",
       "      <td>0.004405</td>\n",
       "      <td>1716.306393</td>\n",
       "      <td>1763.110217</td>\n",
       "      <td>0.036099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1032</th>\n",
       "      <td>ethambutol</td>\n",
       "      <td>Cluster3909_2:Cluster551_9</td>\n",
       "      <td>Rv3848:Rv3795</td>\n",
       "      <td>double</td>\n",
       "      <td>Rv3848:embB</td>\n",
       "      <td>0.008978</td>\n",
       "      <td>1096.392738</td>\n",
       "      <td>1200.401236</td>\n",
       "      <td>0.400428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1037</th>\n",
       "      <td>ethambutol</td>\n",
       "      <td>Cluster4660_1:Cluster3849_1</td>\n",
       "      <td>Rv0687:Rv0265c</td>\n",
       "      <td>double</td>\n",
       "      <td>Rv0687:Rv0265c</td>\n",
       "      <td>0.014784</td>\n",
       "      <td>1757.642527</td>\n",
       "      <td>1820.047625</td>\n",
       "      <td>0.016044</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1038</th>\n",
       "      <td>ethambutol</td>\n",
       "      <td>Cluster4660_1:Cluster3849_2</td>\n",
       "      <td>Rv0687:Rv0265c</td>\n",
       "      <td>double</td>\n",
       "      <td>Rv0687:Rv0265c</td>\n",
       "      <td>0.036636</td>\n",
       "      <td>1757.642527</td>\n",
       "      <td>1820.047625</td>\n",
       "      <td>0.016044</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1047</th>\n",
       "      <td>ethambutol</td>\n",
       "      <td>Cluster2212_2:Cluster551_2</td>\n",
       "      <td>Rv0346c:Rv3795</td>\n",
       "      <td>double</td>\n",
       "      <td>ansP2:embB</td>\n",
       "      <td>0.046976</td>\n",
       "      <td>1102.602691</td>\n",
       "      <td>1237.813738</td>\n",
       "      <td>0.403714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1056</th>\n",
       "      <td>ethambutol</td>\n",
       "      <td>Cluster2212_2:Cluster551_10</td>\n",
       "      <td>Rv0346c:Rv3795</td>\n",
       "      <td>double</td>\n",
       "      <td>ansP2:embB</td>\n",
       "      <td>0.000020</td>\n",
       "      <td>1102.602691</td>\n",
       "      <td>1237.813738</td>\n",
       "      <td>0.403714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1069</th>\n",
       "      <td>ethambutol</td>\n",
       "      <td>Cluster4244_3:Cluster551_11</td>\n",
       "      <td>Rv3806c:Rv3795</td>\n",
       "      <td>double</td>\n",
       "      <td>ubiA:embB</td>\n",
       "      <td>0.001727</td>\n",
       "      <td>1099.037189</td>\n",
       "      <td>1239.448661</td>\n",
       "      <td>0.406873</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1094</th>\n",
       "      <td>ethambutol</td>\n",
       "      <td>Cluster1039_2:Cluster4244_2</td>\n",
       "      <td>Rv1563c:Rv3806c</td>\n",
       "      <td>double</td>\n",
       "      <td>treY:ubiA</td>\n",
       "      <td>0.000046</td>\n",
       "      <td>1579.450041</td>\n",
       "      <td>1662.656839</td>\n",
       "      <td>0.121720</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1095</th>\n",
       "      <td>ethambutol</td>\n",
       "      <td>Cluster1039_2:Cluster4244_3</td>\n",
       "      <td>Rv1563c:Rv3806c</td>\n",
       "      <td>double</td>\n",
       "      <td>treY:ubiA</td>\n",
       "      <td>0.034125</td>\n",
       "      <td>1579.450041</td>\n",
       "      <td>1662.656839</td>\n",
       "      <td>0.121720</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1097</th>\n",
       "      <td>ethambutol</td>\n",
       "      <td>Cluster1039_1:Cluster4244_1</td>\n",
       "      <td>Rv1563c:Rv3806c</td>\n",
       "      <td>double</td>\n",
       "      <td>treY:ubiA</td>\n",
       "      <td>0.010290</td>\n",
       "      <td>1579.450041</td>\n",
       "      <td>1662.656839</td>\n",
       "      <td>0.121720</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1179</th>\n",
       "      <td>ethambutol</td>\n",
       "      <td>Cluster6096_3:Cluster6118_2</td>\n",
       "      <td>Rv1031:Rv2390c</td>\n",
       "      <td>double</td>\n",
       "      <td>kdpC:Rv2390c</td>\n",
       "      <td>0.003997</td>\n",
       "      <td>1637.337390</td>\n",
       "      <td>1694.542063</td>\n",
       "      <td>0.083190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1182</th>\n",
       "      <td>ethambutol</td>\n",
       "      <td>Cluster6096_1:Cluster6118_1</td>\n",
       "      <td>Rv1031:Rv2390c</td>\n",
       "      <td>double</td>\n",
       "      <td>kdpC:Rv2390c</td>\n",
       "      <td>0.019374</td>\n",
       "      <td>1637.337390</td>\n",
       "      <td>1694.542063</td>\n",
       "      <td>0.083190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1193</th>\n",
       "      <td>ethambutol</td>\n",
       "      <td>Cluster3176_3:Cluster4244_2</td>\n",
       "      <td>Rv1267c:Rv3806c</td>\n",
       "      <td>double</td>\n",
       "      <td>embR:ubiA</td>\n",
       "      <td>0.000183</td>\n",
       "      <td>1593.296565</td>\n",
       "      <td>1676.503363</td>\n",
       "      <td>0.113861</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      antibiotic           allele_interaction  geneID_interact  \\\n",
       "1019  ethambutol  Cluster3849_2:Cluster3909_2   Rv0265c:Rv3848   \n",
       "1020  ethambutol  Cluster3849_2:Cluster3909_3   Rv0265c:Rv3848   \n",
       "1032  ethambutol   Cluster3909_2:Cluster551_9    Rv3848:Rv3795   \n",
       "1037  ethambutol  Cluster4660_1:Cluster3849_1   Rv0687:Rv0265c   \n",
       "1038  ethambutol  Cluster4660_1:Cluster3849_2   Rv0687:Rv0265c   \n",
       "1047  ethambutol   Cluster2212_2:Cluster551_2   Rv0346c:Rv3795   \n",
       "1056  ethambutol  Cluster2212_2:Cluster551_10   Rv0346c:Rv3795   \n",
       "1069  ethambutol  Cluster4244_3:Cluster551_11   Rv3806c:Rv3795   \n",
       "1094  ethambutol  Cluster1039_2:Cluster4244_2  Rv1563c:Rv3806c   \n",
       "1095  ethambutol  Cluster1039_2:Cluster4244_3  Rv1563c:Rv3806c   \n",
       "1097  ethambutol  Cluster1039_1:Cluster4244_1  Rv1563c:Rv3806c   \n",
       "1179  ethambutol  Cluster6096_3:Cluster6118_2   Rv1031:Rv2390c   \n",
       "1182  ethambutol  Cluster6096_1:Cluster6118_1   Rv1031:Rv2390c   \n",
       "1193  ethambutol  Cluster3176_3:Cluster4244_2  Rv1267c:Rv3806c   \n",
       "\n",
       "     single_or_double   name_interact    pValue          AIC          BIC  \\\n",
       "1019           double  Rv0265c:Rv3848  0.012672  1716.306393  1763.110217   \n",
       "1020           double  Rv0265c:Rv3848  0.004405  1716.306393  1763.110217   \n",
       "1032           double     Rv3848:embB  0.008978  1096.392738  1200.401236   \n",
       "1037           double  Rv0687:Rv0265c  0.014784  1757.642527  1820.047625   \n",
       "1038           double  Rv0687:Rv0265c  0.036636  1757.642527  1820.047625   \n",
       "1047           double      ansP2:embB  0.046976  1102.602691  1237.813738   \n",
       "1056           double      ansP2:embB  0.000020  1102.602691  1237.813738   \n",
       "1069           double       ubiA:embB  0.001727  1099.037189  1239.448661   \n",
       "1094           double       treY:ubiA  0.000046  1579.450041  1662.656839   \n",
       "1095           double       treY:ubiA  0.034125  1579.450041  1662.656839   \n",
       "1097           double       treY:ubiA  0.010290  1579.450041  1662.656839   \n",
       "1179           double    kdpC:Rv2390c  0.003997  1637.337390  1694.542063   \n",
       "1182           double    kdpC:Rv2390c  0.019374  1637.337390  1694.542063   \n",
       "1193           double       embR:ubiA  0.000183  1593.296565  1676.503363   \n",
       "\n",
       "      pseudo_r2  \n",
       "1019   0.036099  \n",
       "1020   0.036099  \n",
       "1032   0.400428  \n",
       "1037   0.016044  \n",
       "1038   0.016044  \n",
       "1047   0.403714  \n",
       "1056   0.403714  \n",
       "1069   0.406873  \n",
       "1094   0.121720  \n",
       "1095   0.121720  \n",
       "1097   0.121720  \n",
       "1179   0.083190  \n",
       "1182   0.083190  \n",
       "1193   0.113861  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_Antibiotic_tmp_df[(all_Antibiotic_tmp_df[\"antibiotic\"]==\"ethambutol\") & (all_Antibiotic_tmp_df[\"pValue\"]<0.05)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "100"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(all_Antibiotic_tmp_df[all_Antibiotic_tmp_df[\"pValue\"]<0.05][\"name_interact\"].unique())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 04.C\n",
    "### Save visual allele co-occurence tables of the identified significant gene-gene interactions.\n",
    "- Saving all interactions in tmp_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "GENETIC_INTERACTION_PDF_DIR = FINAL_SUPP_DIR+\"Supplementary_Data_File_4/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# generate the dictionary for the genetic \n",
    "pdf_cluster_output_dict = {}\n",
    "# for antibiotic, row in tmp_df.groupby([\"antibiotic\"]):\n",
    "for antibiotic, row in all_Antibiotic_pVal_correct_df.groupby([\"antibiotic\"]):\n",
    "    pdf_cluster_output_dict[antibiotic] = []\n",
    "    for clst_clst in row[\"allele_interaction\"]:\n",
    "        clst1 = clst_clst.split(\":\")[0].split(\"_\")[0].replace(\"ter\", \"ter \")\n",
    "        clst2 = clst_clst.split(\":\")[1].split(\"_\")[0].replace(\"ter\", \"ter \")\n",
    "        if (clst2, clst1) not in pdf_cluster_output_dict[antibiotic]:\n",
    "            pdf_cluster_output_dict[antibiotic].append((clst2, clst1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "../supplement/cooccurence_table_figures/rifampicin_Rv1877__Rv2680_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/rifampicin_Rv1877__Rv1597_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/rifampicin_rpoB__echA19_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/rifampicin_Rv2680__Rv1526c_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/rifampicin_rpoB__Rv1597_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/rifampicin_Rv1877__pks9_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/rifampicin_rpoB__Rv1526c_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/rifampicin_rpoB__rpoC_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/isoniazid_katG__gdh_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/isoniazid_katG__Rv0669c_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/isoniazid_pbpA__lpqR_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/isoniazid_Rv1877__priA_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/isoniazid_inhA__oxcA_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/isoniazid_katG__Rv0265c_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/isoniazid_katG__pbpA_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/isoniazid_katG__oxcA_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/isoniazid_katG__inhA_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/isoniazid_katG__lpqR_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/isoniazid_katG__Rv2704_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/isoniazid_katG__priA_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/pyrazinamide_Rv3843c__xthA_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/pyrazinamide_Rv0625c__dppC_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/pyrazinamide_pncA__ercc3_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/ethionamide_ethA__Rv3421c_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/ethionamide_Rv1184c__Rv3421c_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/ethionamide_rseA__lprJ_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/ethionamide_ethA__mec_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/ethionamide_ethA__ald_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/ethionamide_ald__lprJ_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/ethambutol_Rv3848__Rv0265c_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/ethambutol_embB__Rv3848_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/ethambutol_Rv0265c__Rv0687_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/ethambutol_embB__ansP2_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/ethambutol_embB__ubiA_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/ethambutol_ubiA__treY_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/ethambutol_Rv2390c__kdpC_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/ethambutol_ubiA__embR_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/XDR_gyrA__katG_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/XDR_rnpA__ubiA_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/XDR_rpoB__pncA_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/XDR_gyrA__ubiA_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/XDR_gyrA__pncA_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/XDR_gyrA__snoP_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/XDR_gyrA__rnpA_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/XDR_pncA__ethA_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/XDR_gyrA__thyA_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/XDR_rpoB__Rv3848_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/XDR_ubiA__vapC21_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/XDR_gyrA__rpoB_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/XDR_pncA__snoP_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/XDR_Rv0633c__ubiA_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/XDR_mazF3__ubiA_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/XDR_rpoB__ubiA_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/XDR_gyrA__ctpA_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/MDR_rpoC__rpsL_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/MDR_rpoB__Rv3312c_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/MDR_katG__embB_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/MDR_pncA__embB_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/MDR_rpoB__pncA_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/MDR_katG__Rv3312c_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/MDR_katG__gid_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/MDR_rpoB__katG_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/MDR_rpoC__ubiA_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/MDR_katG__gyrA_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/MDR_gyrA__Rv3312c_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/MDR_katG__ethA_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/MDR_ethA__gid_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/MDR_katG__rpoC_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/MDR_gid__Rv3312c_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/MDR_katG__ubiA_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/MDR_rpoB__gyrA_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/streptomycin_rpsL__mshD_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/streptomycin_rpsL__xthA_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/streptomycin_Rv3773c__Rv0688_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/streptomycin_rpsL__gid_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/streptomycin_yajC__Rv1591_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/streptomycin_rpsL__Rv0449c_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/streptomycin_gid__Rv0515_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/streptomycin_Rv1877__xthA_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/streptomycin_yajC__moeW_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/streptomycin_rpsL__Rv1877_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/streptomycin_Rv1877__dnaA_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/streptomycin_rpsL__Rv1591_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/ofloxacin_gyrA__ansP2_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/ofloxacin_gyrA__ptrBb_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/ofloxacin_gyrA__lprA_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/ofloxacin_Rv1115__Rv0880_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/ofloxacin_Rv1115__phoR_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/ofloxacin_gyrA__rplV_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/ofloxacin_phoR__Rv0428c_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/ofloxacin_gyrA__Rv1081c_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/ofloxacin_gyrA__accD5_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/ofloxacin_gyrA__Rv0071_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/ofloxacin_gyrA__phoR_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/ofloxacin_lprA__ansP2_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/ofloxacin_gyrA__Rv0633c_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/ofloxacin_Rv0071__Rv0428c_oddsRatio_occurance.png\n",
      "../supplement/cooccurence_table_figures/ofloxacin_gyrA__Rv0880_oddsRatio_occurance.png\n",
      "rifampicin ['../supplement/cooccurence_table_figures/rifampicin_Rv1877__Rv2680_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/rifampicin_Rv1877__Rv1597_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/rifampicin_rpoB__echA19_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/rifampicin_Rv2680__Rv1526c_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/rifampicin_rpoB__Rv1597_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/rifampicin_Rv1877__pks9_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/rifampicin_rpoB__Rv1526c_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/rifampicin_rpoB__rpoC_oddsRatio_occurance.png']\n",
      "isoniazid ['../supplement/cooccurence_table_figures/isoniazid_katG__gdh_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/isoniazid_katG__Rv0669c_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/isoniazid_pbpA__lpqR_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/isoniazid_Rv1877__priA_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/isoniazid_inhA__oxcA_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/isoniazid_katG__Rv0265c_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/isoniazid_katG__pbpA_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/isoniazid_katG__oxcA_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/isoniazid_katG__inhA_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/isoniazid_katG__lpqR_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/isoniazid_katG__Rv2704_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/isoniazid_katG__priA_oddsRatio_occurance.png']\n",
      "XDR ['../supplement/cooccurence_table_figures/XDR_gyrA__katG_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/XDR_rnpA__ubiA_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/XDR_rpoB__pncA_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/XDR_gyrA__ubiA_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/XDR_gyrA__pncA_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/XDR_gyrA__snoP_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/XDR_gyrA__rnpA_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/XDR_pncA__ethA_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/XDR_gyrA__thyA_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/XDR_rpoB__Rv3848_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/XDR_ubiA__vapC21_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/XDR_gyrA__rpoB_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/XDR_pncA__snoP_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/XDR_Rv0633c__ubiA_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/XDR_mazF3__ubiA_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/XDR_rpoB__ubiA_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/XDR_gyrA__ctpA_oddsRatio_occurance.png']\n",
      "MDR ['../supplement/cooccurence_table_figures/MDR_rpoC__rpsL_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/MDR_rpoB__Rv3312c_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/MDR_katG__embB_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/MDR_pncA__embB_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/MDR_rpoB__pncA_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/MDR_katG__Rv3312c_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/MDR_katG__gid_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/MDR_rpoB__katG_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/MDR_rpoC__ubiA_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/MDR_katG__gyrA_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/MDR_gyrA__Rv3312c_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/MDR_katG__ethA_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/MDR_ethA__gid_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/MDR_katG__rpoC_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/MDR_gid__Rv3312c_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/MDR_katG__ubiA_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/MDR_rpoB__gyrA_oddsRatio_occurance.png']\n",
      "streptomycin ['../supplement/cooccurence_table_figures/streptomycin_rpsL__mshD_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/streptomycin_rpsL__xthA_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/streptomycin_Rv3773c__Rv0688_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/streptomycin_rpsL__gid_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/streptomycin_yajC__Rv1591_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/streptomycin_rpsL__Rv0449c_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/streptomycin_gid__Rv0515_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/streptomycin_Rv1877__xthA_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/streptomycin_yajC__moeW_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/streptomycin_rpsL__Rv1877_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/streptomycin_Rv1877__dnaA_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/streptomycin_rpsL__Rv1591_oddsRatio_occurance.png']\n",
      "ofloxacin ['../supplement/cooccurence_table_figures/ofloxacin_gyrA__ansP2_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/ofloxacin_gyrA__ptrBb_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/ofloxacin_gyrA__lprA_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/ofloxacin_Rv1115__Rv0880_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/ofloxacin_Rv1115__phoR_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/ofloxacin_gyrA__rplV_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/ofloxacin_phoR__Rv0428c_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/ofloxacin_gyrA__Rv1081c_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/ofloxacin_gyrA__accD5_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/ofloxacin_gyrA__Rv0071_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/ofloxacin_gyrA__phoR_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/ofloxacin_lprA__ansP2_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/ofloxacin_gyrA__Rv0633c_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/ofloxacin_Rv0071__Rv0428c_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/ofloxacin_gyrA__Rv0880_oddsRatio_occurance.png']\n",
      "ethambutol ['../supplement/cooccurence_table_figures/ethambutol_Rv3848__Rv0265c_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/ethambutol_embB__Rv3848_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/ethambutol_Rv0265c__Rv0687_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/ethambutol_embB__ansP2_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/ethambutol_embB__ubiA_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/ethambutol_ubiA__treY_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/ethambutol_Rv2390c__kdpC_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/ethambutol_ubiA__embR_oddsRatio_occurance.png']\n",
      "ethionamide ['../supplement/cooccurence_table_figures/ethionamide_ethA__Rv3421c_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/ethionamide_Rv1184c__Rv3421c_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/ethionamide_rseA__lprJ_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/ethionamide_ethA__mec_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/ethionamide_ethA__ald_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/ethionamide_ald__lprJ_oddsRatio_occurance.png']\n",
      "pyrazinamide ['../supplement/cooccurence_table_figures/pyrazinamide_Rv3843c__xthA_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/pyrazinamide_Rv0625c__dppC_oddsRatio_occurance.png', '../supplement/cooccurence_table_figures/pyrazinamide_pncA__ercc3_oddsRatio_occurance.png']\n"
     ]
    }
   ],
   "source": [
    "from fpdf import FPDF\n",
    "import img2pdf\n",
    "\n",
    "which_df_2_color = \"oddsRatio\"\n",
    "antibiotic_to_imagelist_dict = {}\n",
    "# for antibiotic_name, cluster_groups in pairwise_table_groups.iteritems():\n",
    "for antibiotic_name, cluster_groups in pdf_cluster_output_dict.iteritems():\n",
    "\n",
    "    imagelist = []\n",
    "    \n",
    "    for (clust1, clust2) in cluster_groups:\n",
    "        \n",
    "        rv_id_1, rv_id_2 = clust_to_rv[clust1], clust_to_rv[clust2]\n",
    "        name1, name2 = rv_to_name[rv_id_1], rv_to_name[rv_id_2]\n",
    "        \n",
    "        if name1 == None and rv_id_1 in clster_manual_map.keys():\n",
    "            name1 = clster_manual_map[rv_id_1]\n",
    "        elif name1 == None:\n",
    "            name1 = rv_id_1\n",
    "\n",
    "        if name2 == None and rv_id_2 in clster_manual_map.keys():\n",
    "            name2 = clster_manual_map[rv_id_2]\n",
    "        elif name2 == None:\n",
    "            name2 = rv_id_2\n",
    "        \n",
    "        # print name1, name2\n",
    "        \n",
    "        if antibiotic_name == '4-aminosalicylic_acid':\n",
    "            antibiotic_name = \"PAS\"\n",
    "        \n",
    "        png_interaction_file = COOCCURANCE_TABLE_FIGURES_DIR+antibiotic_name+\"_\"+name1+\"__\"+name2+\"_\"+which_df_2_color+\"_occurance.png\"\n",
    "        print png_interaction_file\n",
    "        imagelist.append(png_interaction_file)\n",
    "        \n",
    "    imagelist = [str(x) for x in imagelist]\n",
    "    antibiotic_to_imagelist_dict.update({antibiotic_name: imagelist})\n",
    "    \n",
    "    \n",
    "# Generate antibiotic genetic interaction files.\n",
    "for antibiotic_name, imagelist in antibiotic_to_imagelist_dict.iteritems():\n",
    "    # pdf_file_name = GENETIC_INTERACTION_PDF_DIR+antibiotic_name+\"_top\"+str(correlation_allele_count)+\"_epistasis.pdf\"\n",
    "    pdf_file_name = GENETIC_INTERACTION_PDF_DIR+antibiotic_name+\"_epistasis.pdf\"\n",
    "    print antibiotic_name, imagelist\n",
    "    with open(pdf_file_name,\"wb\") as f:\n",
    "        f.write(img2pdf.convert(imagelist))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "----"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
