{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.6.1\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "from pickle import *\n",
    "from tqdm import tqdm\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import random\n",
    "from collections import Counter\n",
    "import sys\n",
    "from sklearn.utils import resample\n",
    "import scipy.stats as ss\n",
    "%matplotlib inline\n",
    "\n",
    "sys.path.append('/home/user/notebook/erol/TB Project/Modules')\n",
    "from naked import *\n",
    "from model_troubleshooting import *\n",
    "from svm_epistasis_funcs import *\n",
    "import networkx as nx\n",
    "\n",
    "from mtb_functions import *\n",
    "from sklearn.linear_model import SGDClassifier\n",
    "from sklearn.feature_selection import SelectFromModel, SelectPercentile, VarianceThreshold\n",
    "\n",
    "from sklearn.model_selection import cross_val_score, StratifiedKFold, train_test_split\n",
    "from sklearn.pipeline import make_pipeline, Pipeline\n",
    "from sklearn.model_selection import cross_val_score\n",
    "from sklearn import svm\n",
    "\n",
    "import ast\n",
    "from itertools import cycle\n",
    "import uniprot\n",
    "import urllib,urllib2\n",
    "from bs4 import BeautifulSoup\n",
    "from sklearn.model_selection import GridSearchCV, RepeatedStratifiedKFold\n",
    "from sklearn.linear_model import LogisticRegression\n",
    "from sklearn.metrics import roc_curve, auc\n",
    "from sklearn.feature_selection import chi2, f_classif\n",
    "from sklearn.svm import LinearSVC"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Load jon's TB database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "resubmission_file_loc = \"resubmission_file/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 11039/11039 [01:15<00:00, 145.80it/s]\n"
     ]
    }
   ],
   "source": [
    "iEK_gene_names = session.query(GeneCluster.cluster, Gene.refseq_locus_tag, \n",
    "                               Gene.gene_name).filter(GeneCluster.genome_id==\"83332.12\").join(Gene).join(Sequence)\n",
    "cluster_genes = pd.read_sql(iEK_gene_names.statement, iEK_gene_names.session.bind)\n",
    "test_gene_names = session.query(Gene.refseq_locus_tag, GeneCluster.cluster, \n",
    "                                Gene.gene_name).filter(Gene.genome_id==\"83332.12\").join(GeneCluster).join(Sequence)\n",
    "test_gene_names = pd.read_sql(test_gene_names.statement, test_gene_names.session.bind)\n",
    "all_clusters = session.query(Cluster).all()\n",
    "\n",
    "clust_to_prod = {}\n",
    "for clust in tqdm(all_clusters):\n",
    "    clust_name = str(clust.cluster)\n",
    "    clust_to_prod.update({clust_name: str(clust.gene.product)})\n",
    "    \n",
    "clust_to_refrv = {}\n",
    "for index, row in test_gene_names.iterrows():\n",
    "    if row.refseq_locus_tag != None:\n",
    "        clust_to_refrv.update({str(row.refseq_locus_tag): row.cluster})\n",
    "\n",
    "clust_to_rv = {}\n",
    "rv_to_name = {}\n",
    "rv_to_clust = {}\n",
    "for index, row in cluster_genes.iterrows():\n",
    "    if row.refseq_locus_tag != None:\n",
    "        clust_to_rv.update({row.cluster: row.refseq_locus_tag})\n",
    "        if row.refseq_locus_tag not in rv_to_clust.keys():\n",
    "            rv_to_clust.update({row.refseq_locus_tag: row.cluster})\n",
    "            rv_to_name.update({row.refseq_locus_tag: row.gene_name})\n",
    "        else:\n",
    "            print row.refseq_locus_tag"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "save_json_obj(clust_to_prod, resubmission_file_loc +\"clust_to_prod.json\")\n",
    "save_json_obj(clust_to_rv, resubmission_file_loc +\"clust_to_rv.json\")\n",
    "save_json_obj(rv_to_clust, resubmission_file_loc +\"rv_to_clust.json\")\n",
    "save_json_obj(rv_to_name, resubmission_file_loc +\"rv_to_name.json\")\n",
    "# save_json_obj(clust_to_refrv, resubmission_file_loc +\"clust_to_refrv.json\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Load pan-genome dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Getting the drug list to be removed from cluster_frame.\n",
    "resistance_data = pd.read_csv(\"../../Dataframes/resistance_frame.csv\",index_col=0,dtype={'genome_id': object})\n",
    "resistance_data[\"genome_id\"] = resistance_data[\"genome_id\"].astype(str)\n",
    "resistance_data = resistance_data.set_index(\"genome_id\")\n",
    "drug_list = list(resistance_data.columns.map(str))\n",
    "\n",
    "pan_cluster_df = pd.read_csv('../../Dataframes/tb_clust_matrix.csv',dtype={\"Unnamed: 0\": object})\n",
    "pan_cluster_df[\"Unnamed: 0\"] = pan_cluster_df[\"Unnamed: 0\"].astype(str)\n",
    "pan_cluster_df = pan_cluster_df.set_index(\"Unnamed: 0\")\n",
    "\n",
    "clust_variant_df = pd.read_csv('../../Dataframes/clust_variant_df.csv',dtype={\"Unnamed: 0\": object})\n",
    "clust_variant_df[\"Unnamed: 0\"] = clust_variant_df[\"Unnamed: 0\"].astype(str)\n",
    "clust_variant_df = clust_variant_df.set_index(\"Unnamed: 0\")\n",
    "\n",
    "test_df_final = pd.read_csv('../../Dataframes/clust_rv_name_ortho_cog.csv', dtype={\"Unnamed: 0\": object})\n",
    "test_df_final[\"Unnamed: 0\"] = test_df_final[\"Unnamed: 0\"].astype(str)\n",
    "test_df_final = test_df_final.set_index(\"Unnamed: 0\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>isoniazid</th>\n",
       "      <th>rifampicin</th>\n",
       "      <th>ethambutol</th>\n",
       "      <th>amikacin</th>\n",
       "      <th>ciprofloxacin</th>\n",
       "      <th>clofazimine</th>\n",
       "      <th>cycloserine</th>\n",
       "      <th>ethionamide</th>\n",
       "      <th>kanamycin</th>\n",
       "      <th>moxifloxacin</th>\n",
       "      <th>nicotinamide</th>\n",
       "      <th>ofloxacin</th>\n",
       "      <th>capreomycin</th>\n",
       "      <th>4-aminosalicylic_acid</th>\n",
       "      <th>amoxicillin</th>\n",
       "      <th>prothionamide</th>\n",
       "      <th>pyrazinamide</th>\n",
       "      <th>rifabutin</th>\n",
       "      <th>streptomycin</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>genome_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1295764.3</th>\n",
       "      <td>R</td>\n",
       "      <td>R</td>\n",
       "      <td>R</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R</td>\n",
       "      <td>R</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R</td>\n",
       "      <td>R</td>\n",
       "      <td>R</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1423468.3</th>\n",
       "      <td>R</td>\n",
       "      <td>R</td>\n",
       "      <td>S</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1448744.3</th>\n",
       "      <td>R</td>\n",
       "      <td>R</td>\n",
       "      <td>S</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1448649.3</th>\n",
       "      <td>S</td>\n",
       "      <td>S</td>\n",
       "      <td>S</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1773.371</th>\n",
       "      <td>R</td>\n",
       "      <td>R</td>\n",
       "      <td>R</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          isoniazid rifampicin ethambutol amikacin ciprofloxacin clofazimine  \\\n",
       "genome_id                                                                      \n",
       "1295764.3         R          R          R      NaN           NaN           S   \n",
       "1423468.3         R          R          S      NaN           NaN         NaN   \n",
       "1448744.3         R          R          S      NaN           NaN         NaN   \n",
       "1448649.3         S          S          S      NaN           NaN         NaN   \n",
       "1773.371          R          R          R      NaN           NaN         NaN   \n",
       "\n",
       "          cycloserine ethionamide kanamycin moxifloxacin nicotinamide  \\\n",
       "genome_id                                                               \n",
       "1295764.3         NaN           R         R          NaN            R   \n",
       "1423468.3         NaN         NaN       NaN          NaN          NaN   \n",
       "1448744.3         NaN         NaN       NaN          NaN          NaN   \n",
       "1448649.3         NaN         NaN       NaN          NaN          NaN   \n",
       "1773.371          NaN         NaN       NaN          NaN          NaN   \n",
       "\n",
       "          ofloxacin capreomycin 4-aminosalicylic_acid amoxicillin  \\\n",
       "genome_id                                                           \n",
       "1295764.3         R           R                   NaN         NaN   \n",
       "1423468.3         S         NaN                   NaN         NaN   \n",
       "1448744.3       NaN         NaN                   NaN         NaN   \n",
       "1448649.3       NaN         NaN                   NaN         NaN   \n",
       "1773.371        NaN         NaN                   NaN         NaN   \n",
       "\n",
       "          prothionamide pyrazinamide rifabutin streptomycin  \n",
       "genome_id                                                    \n",
       "1295764.3           NaN            R       NaN            R  \n",
       "1423468.3           NaN          NaN       NaN            R  \n",
       "1448744.3           NaN          NaN       NaN            S  \n",
       "1448649.3           NaN          NaN       NaN            S  \n",
       "1773.371            NaN          NaN       NaN            R  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "resistance_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3. Load TB transcription factor network"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(15980, 12)\n",
      "(15248, 12)\n",
      "3476\n",
      "154\n"
     ]
    }
   ],
   "source": [
    "tb_TF_network = pd.read_excel(\"tb_TF_network.xlsx\",header=2)\n",
    "\n",
    "drop_MT_rows = []\n",
    "print tb_TF_network.shape\n",
    "for index, row in tb_TF_network.iterrows():\n",
    "    if \"MT\" in row[\"Regulator\"] or \"MT\" in row[\"Gene\"]:\n",
    "        drop_MT_rows.append(index)\n",
    "tb_TF_network.drop(drop_MT_rows, inplace=True)\n",
    "print tb_TF_network.shape\n",
    "\n",
    "tb_TF_network_dict = {}\n",
    "for regulator, grp_genes in tb_TF_network.groupby([\"Regulator\"]):\n",
    "    tb_TF_network_dict.update({regulator: list(grp_genes[\"Gene\"].values)})\n",
    "    \n",
    "TF_network = pd.crosstab(index=tb_TF_network[\"Gene\"], columns=tb_TF_network[\"Regulator\"])\n",
    "gene_to_TFs_dict = {}\n",
    "for index, row in TF_network.iterrows():\n",
    "    gene_TFs = list(row[row==1].index)\n",
    "    gene_to_TFs_dict.update({index: gene_TFs})\n",
    "    \n",
    "print len(gene_to_TFs_dict.keys())\n",
    "\n",
    "all_TFs = []\n",
    "for g in gene_to_TFs_dict.keys():        \n",
    "    if g in gene_to_TFs_dict.keys():\n",
    "        all_TFs.extend(gene_to_TFs_dict[g])\n",
    "\n",
    "all_TFs_COUNTER = Counter(all_TFs)\n",
    "print len(tb_TF_network_dict.keys())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "send jon first sentence of paper and 212 slide"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 4. Load pathway mapping (optional)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import urllib2\n",
    "url = 'http://rest.kegg.jp/list/pathway/mtu'\n",
    "data = urllib2.urlopen(url, timeout=20).read()\n",
    "\n",
    "gene_to_pathway = {}\n",
    "pathway_to_genes = {}\n",
    "\n",
    "for kegg_path in tqdm(data.split(\"\\n\")):\n",
    "    if len(kegg_path)!= 0:\n",
    "        # print kegg_path.split(\"\\t\")[0].split(\":\")[1]\n",
    "        tb_kegg_path = kegg_path.split(\"\\t\")[0].split(\":\")[1]\n",
    "        # print tb_kegg_path\n",
    "        # print kegg_path.split(\"\\t\")[1].split(\" - \")[0]\n",
    "        pathway_name = kegg_path.split(\"\\t\")[1].split(\" - \")[0]\n",
    "        \n",
    "        # url = 'http://rest.kegg.jp/link/genes/' + str(tb_kegg_path)\n",
    "        url = 'http://rest.kegg.jp/link/mtu/' + str(tb_kegg_path)\n",
    "        # print url\n",
    "        data = urllib2.urlopen(url, timeout=20).read()\n",
    "        \n",
    "        pathway_to_genes[pathway_name] = []\n",
    "        \n",
    "        for g in data.split(\"\\n\"):\n",
    "            if len(g)!=0:\n",
    "                g_to_p = g.split(\"path:\" + tb_kegg_path + \"\\tmtu:\")[1]\n",
    "\n",
    "                if g_to_p not in gene_to_pathway.keys():\n",
    "                    gene_to_pathway.update({g_to_p: [pathway_name]})\n",
    "                else:\n",
    "                    gene_to_pathway[g_to_p].append(pathway_name)\n",
    "                \n",
    "                pathway_to_genes[pathway_name].extend([g_to_p])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 5. Filter out Mobile elements, PE family proteins, and hypotheticals."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 15367/15367 [00:00<00:00, 23001.02it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15367\n",
      "9975\n",
      "5392\n"
     ]
    }
   ],
   "source": [
    "clust_variant_df = clust_variant_df.fillna(0).copy()\n",
    "\n",
    "remove_cols = []\n",
    "for col in tqdm(clust_variant_df.columns):\n",
    "    col_name = col.split(\"_\")[0].replace(\"ter\", \"ter \")\n",
    "    clust_to_RV = test_df_final.loc[col_name, \"clust_to_rv\"]\n",
    "    gene_name = test_df_final.loc[col_name, \"gene_name\"]\n",
    "    product = test_df_final.loc[col_name, \"product\"]\n",
    "    \n",
    "    if clust_to_RV == \"0\" and gene_name == \"0\":\n",
    "        remove_cols.append(col)\n",
    "\n",
    "    if product == \"Mobile element protein\" or product == \"hypothetical protein\":\n",
    "        remove_cols.append(col)\n",
    "        \n",
    "    if \"PE-PGRS\" in product or \"PPE family\" in product or \"PE family protein\" in product:\n",
    "        remove_cols.append(col)\n",
    "        \n",
    "    if \"hypothetical protein\" in product or \"Hypothetical protein\" in product:\n",
    "        remove_cols.append(col)\n",
    "        \n",
    "    if \"Transposase\" in product:\n",
    "        remove_cols.append(col)\n",
    "\n",
    "\n",
    "\n",
    "remove_cols = list(set(remove_cols))\n",
    "print len(clust_variant_df.columns)\n",
    "clust_variant_df.drop(remove_cols, axis=1, inplace=True)\n",
    "print len(clust_variant_df.columns)\n",
    "print len(remove_cols)\n",
    "\n",
    "data_to_plot = clust_variant_df.copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Conclusions from analysis below.\n",
    "\n",
    "Build train an L1-norm SVM using SGD by bootstrap aggregation \n",
    "    - Different threshold values between 150 and 220\n",
    "    - We do this because setting a specific threshold greatly influences what can be picked for each SVM hyperplane. i.e. if the threshold is less than 180, embR does not appear and proC does, while above ~180 embR appears and proC does not.\n",
    "    - Do the variations in the certain allele weights correlate with the variations in other allele weights? This points to epistasis. In the case of Ethambutol, we see that the variations in ubiA (recently KNOWN) are correlated with variations in embR and proC. embR makes sense given that embR regulates embABC which modulates Ethambutol susceptibility. Since ubiA increases flux through DPA\n",
    "\n",
    "\n",
    "- Negative weights correspond to **Susceptibility**.\n",
    "- Positive weights correspond to **Resistance**.\n",
    "\n",
    "\n",
    "1. Ensemble of Support vector hyperplanes describes epistatic interactions to AMR.\n",
    "    - Uncovers relationship between *inhA* and *katG* for Isoniazid.\n",
    "    - Implicates relationship between *gyrA* and *ald* for Ofloxacin.\n",
    "        - Alanine dehydrogenase is associated with Ofloxacin AMR as opposed to D-cycloserine.\n",
    "    - Epistatic interactions between *ubiA* and *embR*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def drug_resist_type(x):\n",
    "    '''Takes rows in dataframe and creates new column '''\n",
    "    '''MDR (multi drug resistant): resistance to at least isoniazid and rifampin'''\n",
    "    '''XDR (extensively drug resistant): resistance to isoniazid and rifampin, plus any \n",
    "        fluoroquinolone and at least one of 3 injectaable second line drugs'''\n",
    "    '''XXDR (totally drug resistant): resistant to all first and second line TB drugs'''\n",
    "    # Multi Drug Resistant - MDR\n",
    "    if x.isoniazid == 'R' and x.rifampicin == 'R':\n",
    "        res_type = 'MDR'\n",
    "    elif x.rifampicin == 'R':\n",
    "        res_type = 'RR' # Rifampicin Resistant\n",
    "    else:\n",
    "        res_type = 'Susceptible'\n",
    "    # All first line drugs\n",
    "    if res_type == 'MDR' and x.ethambutol == \"R\" and x.pyrazinamide== \"R\": # and x.streptomycin == \"R\":\n",
    "        all_first_line = \"Yes\"\n",
    "    else:\n",
    "        all_first_line = \"No\"\n",
    "    # Second line aminoglycosides - at least 1.\n",
    "    if x.amikacin == 'R' or x.kanamycin == 'R' or x.capreomycin == 'R':\n",
    "        second_line = \"Yes\"\n",
    "    else:\n",
    "        second_line = \"No\"\n",
    "    # Second line Flouroquinols - \n",
    "    if x.ciprofloxacin == 'R' or x.ofloxacin == 'R' or x.moxifloxacin == 'R':\n",
    "        second_flouro = \"Yes\"\n",
    "    else:\n",
    "        second_flouro = \"No\"\n",
    "    # ALL second line drugs\n",
    "    if (x.ciprofloxacin == 'R' and x.ofloxacin == 'R' and x.moxifloxacin == 'R' and \n",
    "        x.amikacin == 'R' and x.kanamycin == 'R' and x.capreomycin == 'R' and x.cycloserine and\n",
    "        x.ethionamide == 'R' and x.prothionamide == 'R' and x[\"4-aminosalicylic_acid\"]== 'R'):\n",
    "        all_second_line = \"Yes\"\n",
    "    else:\n",
    "        all_second_line = \"No\"\n",
    "    # Extensively Drug Resistant - XDR\n",
    "    if res_type == 'MDR' and second_line == \"Yes\" and second_flouro == \"Yes\":\n",
    "        res_type = 'XDR'\n",
    "    # Totally Drug Resistant - XXDR\n",
    "    if all_first_line == \"Yes\" and all_second_line == \"Yes\":\n",
    "        res_type = 'XXDR'\n",
    "    #print x\n",
    "    return res_type\n",
    "\n",
    "def get_MDR_training(resistance_data, data_to_plot, MDR_or_XDR):\n",
    "    # MDR\n",
    "    resist_type = resistance_data.apply(lambda x: drug_resist_type(x) , axis=1)\n",
    "\n",
    "    print \"# RR\", len(resist_type[resist_type==\"RR\"].index)\n",
    "    print \"# MDR\", len(resist_type[resist_type==\"MDR\"].index)\n",
    "    print \"# XDR\", len(resist_type[resist_type==\"XDR\"].index)\n",
    "\n",
    "    resist_type[resist_type==\"Susceptible\"] = 0\n",
    "    resist_type[resist_type==\"RR\"] = 0\n",
    "    if MDR_or_XDR == \"XDR\":\n",
    "        resist_type[resist_type==\"MDR\"] = 0\n",
    "        resist_type[resist_type==\"XDR\"] = 1\n",
    "    if MDR_or_XDR == \"MDR\":\n",
    "        resist_type[resist_type==\"MDR\"] = 1\n",
    "        resist_type[resist_type==\"XDR\"] = 1\n",
    "    # resist_type[resist_type<1] = 0\n",
    "    print \"number of \",MDR_or_XDR,\" \",resist_type.sum()\n",
    "\n",
    "    shared_strains = set(list(data_to_plot.index)).intersection(set(resist_type.index))\n",
    "    print \"number of strains available for classification\",len(shared_strains)\n",
    "\n",
    "    y_MDR = resist_type[list(shared_strains)]\n",
    "    print y_MDR.sum()\n",
    "\n",
    "    X = data_to_plot.loc[shared_strains, :].copy()\n",
    "    y = y_MDR.fillna(0).copy()\n",
    "    drug_name = \"MDR\"\n",
    "    return X, y"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 6. Hyperparameterization through sklearn GRIDSEARCH_CV"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "threshold_val = 130\n",
    "\n",
    "for drug_name in [\"ethionamide\"]: # \"ethambutol\", \"pyrazinamide\", \"isoniazid\", \"pyrazinamide\", \"4-aminosalicylic_acid\"\n",
    "    \n",
    "    if drug_name == \"MDR\" or drug_name == \"XDR\":\n",
    "        X1, y1 = get_MDR_training(resistance_data, data_to_plot, drug_name)\n",
    "    else:\n",
    "        y1, X1 = get_target_data(data_to_plot, resistance_data, drug_name)\n",
    "        X1 = remove_key_antibiotic_clusters(X1, drug_name)\n",
    "    \n",
    "    print \"Before removing low variance: \", X1.shape\n",
    "    selector = VarianceThreshold(threshold=.01)\n",
    "    selector.fit_transform(X1)\n",
    "    X1 = X1[X1.columns[selector.get_support()]].copy()\n",
    "    print \"After removing low variance: \", X1.shape\n",
    "\n",
    "    # cross-validation sets\n",
    "    # cv = StratifiedKFold(n_splits = 10)\n",
    "    cv = RepeatedStratifiedKFold(n_splits=5,n_repeats=2)\n",
    "\n",
    "    # estimator\n",
    "    clf = SGDClassifier(loss=\"hinge\", penalty=\"l1\", shuffle=True, class_weight=\"balanced\", alpha=0.0001) \n",
    "    # clf = LogisticRegression(penalty = \"l1\", class_weight=\"balanced\")\n",
    "    # feature selection w/ parameter \"threshold\"\n",
    "    feat_selection = SelectFromModel(clf, threshold=threshold_val)\n",
    "    # feat_selection = SelectKBest(clf, threshold=threshold_val)\n",
    "\n",
    "    # Setup pipeline for SGD-SVM feature selection\n",
    "    model = Pipeline([\n",
    "              ('fs', feat_selection),  # SelectFromModel using SGD-SVM\n",
    "              ('clf', clf), \n",
    "            ])\n",
    "\n",
    "    # Set the parameters for optimizing cross-validation \n",
    "    params = {\n",
    "        'fs__threshold': np.linspace(120, 200, num=20) # try a range from 15 -> 150, \n",
    "        # np.linspace(.4, 1, num=10) for logistic regression\n",
    "    }\n",
    "\n",
    "\n",
    "    grid = GridSearchCV( model , params, cv=cv)\n",
    "    grid.fit(X1, y1)\n",
    "\n",
    "    print(\"The best parameters are %s with a score of %0.2f\"\n",
    "          % (grid.best_params_, grid.best_score_))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'X1' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-7-21d4ceb64b9c>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0mnew_cols\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnew_vars\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m \u001b[0;32mfor\u001b[0m \u001b[0mx\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mX1\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mgrid\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbest_estimator_\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnamed_steps\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"fs\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_support\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      5\u001b[0m     \u001b[0mnew_vars\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m     \u001b[0mnew_cols\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"_\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreplace\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"ter\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"ter \"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'X1' is not defined"
     ]
    }
   ],
   "source": [
    "clust_select_list, clust_var_list = [], []\n",
    "new_cols, new_vars = [], []\n",
    "\n",
    "for x in list(X1.columns[grid.best_estimator_.named_steps[\"fs\"].get_support()]):\n",
    "    new_vars.append(str(x))\n",
    "    new_cols.append(str(x).split(\"_\")[0].replace(\"ter\", \"ter \"))\n",
    "    \n",
    "new_cols = set(new_cols)\n",
    "new_vars = set(new_vars)\n",
    "\n",
    "clust_select_list = clust_select_list + list(new_cols)\n",
    "clust_var_list = clust_var_list + list(new_vars)\n",
    "\n",
    "print len(list(X1.columns[grid.best_estimator_.named_steps[\"fs\"].get_support()]))\n",
    "test_df_final.loc[clust_select_list]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 7. Allele AMR simulations\n",
    "1. Generates figures, and GEM-PRO dataframes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "clster_manual_map = {\n",
    "    \"Rv1334\": \"mec\",\n",
    "    \"Rv1184c\": \"Chp2\",\n",
    "    \"Rv3806c\": \"ubiA\",\n",
    "    \"Rv3124\": \"moaR1\",\n",
    "    \"Rv0265c\": \"fecB2\",\n",
    "    \"Rv0669c\": \"NCDase\",\n",
    "    \"Rv1675c\": \"cmr\",\n",
    "    \"Rv3447c\": \"eccC4\",\n",
    "    \"Rv3796\": \"atsH\",\n",
    "    \"Rv1942c\": \"mazF5\",\n",
    "}\n",
    "\n",
    "threshold_range_dict = {\n",
    "    \"ethambutol\": [.7, .9], # 160, 200 for .8 sample size, [120, 140] for maybe .3?\n",
    "    \"pyrazinamide\": [120, 140], # [160, 170], train_size = 0.6, [175, 180] for train_size = 0.3\n",
    "    \"isoniazid\": [160, 200],\n",
    "    \"ethionamide\": [190, 215],\n",
    "    \"ofloxacin\": [190, 220],\n",
    "    \"4-aminosalicylic_acid\": [160, 210],\n",
    "    \"rifampicin\": [160, 200],\n",
    "    \"MDR\": [195, 210],\n",
    "    \"XDR\": [240, 260],\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>isoniazid</th>\n",
       "      <th>rifampicin</th>\n",
       "      <th>ethambutol</th>\n",
       "      <th>amikacin</th>\n",
       "      <th>ciprofloxacin</th>\n",
       "      <th>clofazimine</th>\n",
       "      <th>cycloserine</th>\n",
       "      <th>ethionamide</th>\n",
       "      <th>kanamycin</th>\n",
       "      <th>moxifloxacin</th>\n",
       "      <th>nicotinamide</th>\n",
       "      <th>ofloxacin</th>\n",
       "      <th>capreomycin</th>\n",
       "      <th>4-aminosalicylic_acid</th>\n",
       "      <th>amoxicillin</th>\n",
       "      <th>prothionamide</th>\n",
       "      <th>pyrazinamide</th>\n",
       "      <th>rifabutin</th>\n",
       "      <th>streptomycin</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>genome_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1295764.3</th>\n",
       "      <td>R</td>\n",
       "      <td>R</td>\n",
       "      <td>R</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R</td>\n",
       "      <td>R</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R</td>\n",
       "      <td>R</td>\n",
       "      <td>R</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1423468.3</th>\n",
       "      <td>R</td>\n",
       "      <td>R</td>\n",
       "      <td>S</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1448744.3</th>\n",
       "      <td>R</td>\n",
       "      <td>R</td>\n",
       "      <td>S</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1448649.3</th>\n",
       "      <td>S</td>\n",
       "      <td>S</td>\n",
       "      <td>S</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1773.371</th>\n",
       "      <td>R</td>\n",
       "      <td>R</td>\n",
       "      <td>R</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          isoniazid rifampicin ethambutol amikacin ciprofloxacin clofazimine  \\\n",
       "genome_id                                                                      \n",
       "1295764.3         R          R          R      NaN           NaN           S   \n",
       "1423468.3         R          R          S      NaN           NaN         NaN   \n",
       "1448744.3         R          R          S      NaN           NaN         NaN   \n",
       "1448649.3         S          S          S      NaN           NaN         NaN   \n",
       "1773.371          R          R          R      NaN           NaN         NaN   \n",
       "\n",
       "          cycloserine ethionamide kanamycin moxifloxacin nicotinamide  \\\n",
       "genome_id                                                               \n",
       "1295764.3         NaN           R         R          NaN            R   \n",
       "1423468.3         NaN         NaN       NaN          NaN          NaN   \n",
       "1448744.3         NaN         NaN       NaN          NaN          NaN   \n",
       "1448649.3         NaN         NaN       NaN          NaN          NaN   \n",
       "1773.371          NaN         NaN       NaN          NaN          NaN   \n",
       "\n",
       "          ofloxacin capreomycin 4-aminosalicylic_acid amoxicillin  \\\n",
       "genome_id                                                           \n",
       "1295764.3         R           R                   NaN         NaN   \n",
       "1423468.3         S         NaN                   NaN         NaN   \n",
       "1448744.3       NaN         NaN                   NaN         NaN   \n",
       "1448649.3       NaN         NaN                   NaN         NaN   \n",
       "1773.371        NaN         NaN                   NaN         NaN   \n",
       "\n",
       "          prothionamide pyrazinamide rifabutin streptomycin  \n",
       "genome_id                                                    \n",
       "1295764.3           NaN            R       NaN            R  \n",
       "1423468.3           NaN          NaN       NaN            R  \n",
       "1448744.3           NaN          NaN       NaN            S  \n",
       "1448649.3           NaN          NaN       NaN            S  \n",
       "1773.371            NaN          NaN       NaN            R  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "resistance_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sim_allDrugs_df = pd.DataFrame()\n",
    "top_num = 40\n",
    "training_size_percents = [.8]\n",
    "score_threshold_test = .80 # .80\n",
    "n_classes = 2\n",
    "simulation_iterations = 400\n",
    "verbose_sim = False\n",
    "\n",
    "fpr, tpr, roc_auc = dict(), dict(), dict()\n",
    "estimator_type = \"SVM\" #logReg or SVM\n",
    "SGD_or_GD = \"SGD\" #SGD or GD\n",
    "\n",
    "# kf = StratifiedKFold(n_splits=2, shuffle=False)\n",
    "\n",
    "antibiotic_clusters = []\n",
    "# Done: Ethambutol, Isoniazid, Pyrazinamide, Rifampicin, Streptomycin, 4-aminosalicylic_acid\n",
    "# [\"isoniazid\", \"ethambutol\", \"rifampicin\", \"pyrazinamide\", \"ethionamide\", \"ofloxacin\", \"4-aminosalicylic_acid\", \"MDR\", \"XDR\"]\n",
    "for drug_name in [\"streptomycin\"]: # [\"isoniazid\", \"ethambutol\", \"rifampicin\", \"pyrazinamide\", \"ethionamide\", \"ofloxacin\", \"4-aminosalicylic_acid\", \"MDR\", \"XDR\"]: # \", \"ethionamide\", \"ofloxacin\", \"amikacin\" \"4-aminosalicylic_acid\"\n",
    "    \n",
    "    if drug_name == \"MDR\" or drug_name == \"XDR\":\n",
    "        X1, y1 = get_MDR_training(resistance_data, data_to_plot, drug_name)\n",
    "    else:\n",
    "        y1, X1 = get_target_data(data_to_plot, resistance_data, drug_name)\n",
    "        X1 = remove_key_antibiotic_clusters(X1, drug_name)\n",
    "    \n",
    "    print \"Before removing low variance: \", X1.shape\n",
    "    selector = VarianceThreshold(threshold=.01)\n",
    "    selector.fit_transform(X1)\n",
    "    X1 = X1[X1.columns[selector.get_support()]].copy()\n",
    "    print \"After removing low variance: \", X1.shape\n",
    "    \n",
    "    # threshold_range = threshold_range_dict[drug_name]\n",
    "\n",
    "    sim_all_df = pd.DataFrame()\n",
    "\n",
    "    for t in tqdm(range(simulation_iterations)):\n",
    "\n",
    "        # randomly choose the threshold value.\n",
    "        # threshold_val = random.randint(threshold_range[0], threshold_range[1])\n",
    "        if estimator_type == \"SVM\":\n",
    "            if SGD_or_GD == \"SGD\":\n",
    "                clf = SGDClassifier(loss=\"hinge\", penalty=\"l1\", shuffle=True, class_weight=\"balanced\" , alpha=0.0001) \n",
    "                if drug_name==\"ethionamide\":\n",
    "                    threshold_val = 160\n",
    "                elif drug_name==\"XDR\":\n",
    "                    threshold_val = 230\n",
    "                elif drug_name==\"capreomycin\":\n",
    "                    threshold_val = 190\n",
    "                elif drug_name==\"moxifloxacin\":\n",
    "                    threshold_val = 190\n",
    "                elif drug_name==\"ciprofloxacin\":\n",
    "                    threshold_val = 140\n",
    "                elif drug_name==\"rifabutin\":\n",
    "                    threshold_val = 100\n",
    "                else:\n",
    "                    threshold_val = 170\n",
    "            elif SGD_or_GD == \"GD\":\n",
    "                clf = LinearSVC(penalty = \"l1\", loss=\"squared_hinge\", dual=False, class_weight=\"balanced\")\n",
    "                threshold_val = .3\n",
    "        elif estimator_type == \"logReg\":\n",
    "            if SGD_or_GD == \"SGD\":\n",
    "                clf = SGDClassifier(loss=\"log\", penalty=\"l1\", shuffle=True, class_weight=\"balanced\" , alpha=0.0001) \n",
    "                threshold_val = 190\n",
    "            elif SGD_or_GD == \"GD\":\n",
    "                clf = LogisticRegression(penalty = \"l1\", class_weight=\"balanced\")\n",
    "                if drug_name==\"ethionamide\":\n",
    "                    threshold_val = .9\n",
    "                elif drug_name==\"cycloserine\":\n",
    "                    threshold_val = .4\n",
    "                elif drug_name==\"rifabutin\":\n",
    "                    threshold_val = .4\n",
    "                else:\n",
    "                    threshold_val = .7    \n",
    "    \n",
    "                    \n",
    "        percent_of_train = random.choice(training_size_percents)\n",
    "        train_x, X_test, train_y, y_test = train_test_split(X1, y1, \n",
    "                                                            train_size = percent_of_train, \n",
    "                                                            stratify=None) # stratify=y1\n",
    "    \n",
    "        # train_x, train_y = resample(X1, y1, n_samples=int(percent_of_train*len(X1.index)))\n",
    "\n",
    "        # feature selection w/ parameter \"threshold\"\n",
    "        feat_selection = SelectFromModel(clf, threshold=threshold_val)\n",
    "\n",
    "        # Setup pipeline for SGD-SVM feature selection\n",
    "        model = Pipeline([\n",
    "                  ('fs', feat_selection),  # SelectFromModel using SGD-SVM\n",
    "                  ('clf', clf), \n",
    "                ])\n",
    "        \n",
    "        # model.fit(X1, y1)\n",
    "        model.fit(train_x, train_y)\n",
    "\n",
    "        sim_t_df = pd.DataFrame([model.named_steps[\"clf\"].coef_[0]], \n",
    "                                columns=X1.columns[model.named_steps[\"fs\"].get_support()])\n",
    "        \n",
    "        model_score_on_test = model.score(X_test, y_test)\n",
    "        \n",
    "        if verbose_sim == True:\n",
    "            print \"\",\n",
    "            print \"# of features in simulation \"+str(t)+ \":\"+ str(len(sim_t_df.columns)),\n",
    "            print \"#R: \" + str(train_y.sum()) + \"#S: \" + str(len(train_y.index)-train_y.sum()),\n",
    "            print \"% of samples: \" + str(percent_of_train),\n",
    "            print \"Prediction Score: \", model_score_on_test\n",
    "        \n",
    "        # if model_score_on_test > score_threshold_test:\n",
    "        sim_all_df = pd.concat([sim_all_df, sim_t_df], ignore_index=True)\n",
    "        \n",
    "        # y_score = model.predict(X_test)\n",
    "        y_score = model.decision_function(X_test)\n",
    "\n",
    "        fpr[t], tpr[t], _ = roc_curve(y_test.values, y_score)\n",
    "        roc_auc[t] = auc(fpr[t], tpr[t])\n",
    "\n",
    "    \n",
    "    print \"# of cluster features...\", len(list(set(sim_all_df.columns)))\n",
    "\n",
    "    sim_all_df_T = sim_all_df.transpose().copy()\n",
    "    sim_all_df_T[\"svm_weight_sum\"] = sim_all_df_T.apply(lambda x: abs(x).sum(), axis=1)\n",
    "    sim_all_df_T_top = sim_all_df_T.sort_values(\"svm_weight_sum\", ascending=False)[:top_num]\n",
    "    \n",
    "    sim_allDrugs_df = pd.concat([sim_allDrugs_df, sim_all_df[sim_all_df_T_top.index]])\n",
    "\n",
    "    amr_clsts = []\n",
    "    for x in list(sim_all_df_T_top.index):\n",
    "        amr_clsts.append(str(x).split(\"_\")[0].replace(\"ter\", \"ter \"))\n",
    "\n",
    "    amr_clsts_set = list(set(amr_clsts))\n",
    "    antibiotic_clusters = antibiotic_clusters+amr_clsts_set\n",
    "\n",
    "    \n",
    "    clster_var_mapping_prod, clster_var_mapping_Rv = {}, {}\n",
    "    for i in sim_all_df_T.index:\n",
    "        var_index = i.split(\"_\")[1]\n",
    "        # print i\n",
    "        # clster_var_mapping_prod.update({i: clust_to_prod[str(i).split(\"_\")[0].replace(\"ter\", \"ter \")]+var_index})\n",
    "\n",
    "        try:\n",
    "            clster_var_mapping_prod.update({i: i+clust_to_prod[str(i).split(\"_\")[0].replace(\"ter\", \"ter \")]+\"_\"+var_index})\n",
    "            # clster_var_mapping_Rv.update({i: clust_to_rv[str(i).split(\"_\")[0].replace(\"ter\", \"ter \")]+var_index})\n",
    "        except:\n",
    "            clster_var_mapping_prod.update({i: i+clust_to_prod[str(i).split(\"_\")[0].replace(\"ter\", \"ter \")]+\"_\"+var_index})\n",
    "            # clster_var_mapping_Rv.update({i: clust_to_prod[str(i).split(\"_\")[0].replace(\"ter\", \"ter \")]+var_index})\n",
    "        \n",
    "    clster_to_IDENTIFIER, rv_list = clster_identifier_func(clster_manual_map, sim_all_df_T, \n",
    "                                                       clust_to_rv, rv_to_name, clust_to_prod)\n",
    "    \n",
    "    save_ROC_curve_figure(drug_name, fpr, tpr, roc_auc, simulation_iterations, estimator_type)\n",
    "    \n",
    "    sim_all_df_Prod, allele_to_colormap, clust_alleles = svm_top_weights_colormap(sim_all_df_T, clster_to_IDENTIFIER, \n",
    "                                                                              top_num, X1, y1, \n",
    "                                                                              clster_var_mapping_prod)\n",
    "    \n",
    "    sim_all_df_Prod.to_csv(\"SVM_dataframes/\"+drug_name+\"_\"+estimator_type+\"_\"+SGD_or_GD+\"_top_clusts.csv\")\n",
    "    sim_all_df_T.sort_values([\"svm_weight_sum\"], ascending=False)[:100].to_csv(\"SVM_dataframes/\"+drug_name+\"_\"+estimator_type+\"_\"+SGD_or_GD+\"_top_100alleles.csv\")\n",
    "    \n",
    "    sim_corr_df = sim_all_df_Prod.fillna(0).T.corr().copy()\n",
    "    sim_corr_df = sim_corr_df[abs(sim_corr_df)>.05].fillna(0)\n",
    "\n",
    "    g_plot = plot_svm_heatmap(sim_corr_df, (15, 15), allele_to_colormap, \"seismic\") # \"seismic\"\n",
    "    g_plot.savefig(\"SVM_figures/\"+drug_name+\"_\"+estimator_type+\"_\"+SGD_or_GD+\"_top\"+str(top_num)+\"_correlation.png\")\n",
    "    \n",
    "    sim_corr_df.to_csv(\"SVM_dataframes/\"+drug_name+\"_\"+estimator_type+\"_\"+SGD_or_GD+\"_top\"+str(top_num)+\"_correlation_matrix.csv\")\n",
    "\n",
    "#     g = sns.clustermap(sim_all_df_Prod.fillna(0), \n",
    "#                        col_cluster=False, row_cluster=True, cmap=\"seismic\",\n",
    "#                       figsize=(18,14), xticklabels=False)\n",
    "#     plt.setp(g.ax_heatmap.yaxis.get_majorticklabels(), rotation=0);\n",
    "#     plt.setp(g.ax_heatmap.yaxis.get_majorticklabels(), rotation=0);\n",
    "#     g.savefig(\"SVM_figures/\"+drug_name+\"_\"+estimator_type+\"_\"+SGD_or_GD+\"_iterations.png\")\n",
    "    \n",
    "#    general_cluster_alleles = list(set([clst_al.split(\"_\")[0].replace(\"ter\", \"ter \") for clst_al in clust_alleles]))\n",
    "\n",
    "#    GEMPro_final = get_GEMPRO_df(general_cluster_alleles, clust_to_rv, rv_to_name, clster_manual_map, clust_variant_df)\n",
    "#    GEMPro_final.to_csv(\"GEMPro_seq_dataframes/\"+drug_name+\"_\"+estimator_type+\"_\"+SGD_or_GD+\"_\"+str(top_num)+\"_SEQdf.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "general_cluster_alleles = list(set([clst_al.split(\"_\")[0].replace(\"ter\", \"ter \") for clst_al in clust_alleles]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "general_cluster_alleles = [\"Cluster 3005\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "GEMPro_final = get_GEMPRO_df(general_cluster_alleles, clust_to_rv, rv_to_name, clster_manual_map, clust_variant_df)\n",
    "\n",
    "#     GEMPro_final.to_csv(\"GEMPro_seq_dataframes/\"+drug_name+\"_\"+estimator_type+\"_\"+SGD_or_GD+\"_SEQdf.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's think about this. We should only care about the hits that have structures... For the other highly implacted genes, it will be important to see how"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 8. Generate Chi2 df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import statsmodels #.api as sm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "506\n",
      "1057\n",
      "isoniazid\n",
      "removing...  107 clusters\n",
      "Before removing low variance:  (1563, 9866)\n",
      "After removing low variance:  (1563, 8571)\n",
      "(1024, 3)\n",
      "848\n",
      "492\n",
      "ethambutol\n",
      "removing...  99 clusters\n",
      "Before removing low variance:  (1340, 9874)\n",
      "After removing low variance:  (1340, 8453)\n",
      "(921, 3)\n",
      "578\n",
      "983\n",
      "rifampicin\n",
      "removing...  92 clusters\n",
      "Before removing low variance:  (1561, 9881)\n",
      "After removing low variance:  (1561, 8581)\n",
      "(1306, 3)\n",
      "92\n",
      "137\n",
      "pyrazinamide\n",
      "removing...  97 clusters\n",
      "Before removing low variance:  (229, 9876)\n",
      "After removing low variance:  (229, 6712)\n",
      "(7, 3)\n",
      "353\n",
      "209\n",
      "ethionamide\n",
      "removing...  101 clusters\n",
      "Before removing low variance:  (562, 9872)\n",
      "After removing low variance:  (562, 7537)\n",
      "(120, 3)\n",
      "554\n",
      "302\n",
      "ofloxacin\n",
      "removing...  102 clusters\n",
      "Before removing low variance:  (856, 9871)\n",
      "After removing low variance:  (856, 7792)\n",
      "(128, 3)\n",
      "295\n",
      "80\n",
      "4-aminosalicylic_acid\n",
      "removing...  109 clusters\n",
      "Before removing low variance:  (375, 9864)\n",
      "After removing low variance:  (375, 7234)\n",
      "(132, 3)\n",
      "732\n",
      "663\n",
      "streptomycin\n",
      "removing...  98 clusters\n",
      "Before removing low variance:  (1395, 9875)\n",
      "After removing low variance:  (1395, 8338)\n",
      "(869, 3)\n",
      "257\n",
      "142\n",
      "amikacin\n",
      "removing...  115 clusters\n",
      "Before removing low variance:  (399, 9858)\n",
      "After removing low variance:  (399, 6794)\n",
      "(39, 3)\n",
      "262\n",
      "71\n",
      "cycloserine\n",
      "removing...  115 clusters\n",
      "Before removing low variance:  (333, 9858)\n",
      "After removing low variance:  (333, 6735)\n",
      "(3, 3)\n",
      "550\n",
      "278\n",
      "kanamycin\n",
      "removing...  115 clusters\n",
      "Before removing low variance:  (828, 9858)\n",
      "After removing low variance:  (828, 7753)\n",
      "(159, 3)\n",
      "237\n",
      "141\n",
      "capreomycin\n",
      "removing...  115 clusters\n",
      "Before removing low variance:  (378, 9858)\n",
      "After removing low variance:  (378, 6942)\n",
      "(57, 3)\n"
     ]
    }
   ],
   "source": [
    "# writer = pd.ExcelWriter('Supplementary_files/Chi2_allele_AMR_associations.xlsx')\n",
    "\n",
    "writer = pd.ExcelWriter('DataFile01_b.xlsx')\n",
    "\n",
    "for drug_name in [\"isoniazid\", \"ethambutol\", \"rifampicin\", \"pyrazinamide\", \"ethionamide\", \n",
    "                  \"ofloxacin\", \"4-aminosalicylic_acid\", \"streptomycin\", \"amikacin\",\"cycloserine\", \n",
    "                  \"kanamycin\", \"capreomycin\"]: # \", \"ethionamide\", \"ofloxacin\", \"amikacin\" \"4-aminosalicylic_acid\"\n",
    "    \n",
    "    chi2_test = pd.DataFrame()\n",
    "    \n",
    "    if drug_name == \"MDR\" or drug_name == \"XDR\":\n",
    "        X1, y1 = get_MDR_training(resistance_data, data_to_plot, drug_name)\n",
    "    else:\n",
    "        y1, X1 = get_target_data(data_to_plot, resistance_data, drug_name)\n",
    "        X1 = remove_key_antibiotic_clusters(X1, drug_name)\n",
    "    \n",
    "    print \"Before removing low variance: \", X1.shape\n",
    "    selector = VarianceThreshold(threshold=.01)\n",
    "    selector.fit_transform(X1)\n",
    "    X1 = X1[X1.columns[selector.get_support()]].copy()\n",
    "    print \"After removing low variance: \", X1.shape\n",
    "    \n",
    "    # chi2_test = pd.DataFrame(chi2(X1, y1)[0]).T\n",
    "    chi2_test = pd.DataFrame(list(chi2(X1, y1)))\n",
    "    chi2_test.columns = X1.columns\n",
    "    chi2_test = chi2_test.sort_values([0, 1], axis=1, ascending=False).T\n",
    "    chi2_test.columns = [\"chi2\", \"pVal\"]\n",
    "    \n",
    "    # bonferroni multiple-testing correction\n",
    "    rejected_list, pvalue_corrected_list, alphaC, alphacBonf = statsmodels.sandbox.stats.multicomp.multipletests(chi2_test[\"pVal\"], \n",
    "                                                                                          alpha=0.005, method='bonferroni', is_sorted=False)\n",
    "\n",
    "    chi2_test_corrected = chi2_test[rejected_list].copy()\n",
    "    chi2_test_corrected[\"corrected_pVal\"] = pvalue_corrected_list[rejected_list]\n",
    "    print chi2_test_corrected.shape\n",
    "    \n",
    "    chi2_test_top50_df = pd.DataFrame()\n",
    "    for allele, row in chi2_test_corrected.iloc[:40, :].iterrows():\n",
    "        cluster, allele_id = allele.split(\"_\")[0].replace(\"ter\", \"ter \"), allele.split(\"_\")[1]\n",
    "        chi2_statistic = row[\"chi2\"]\n",
    "        pVal_statistic = row[\"pVal\"]\n",
    "        pVal_corrected_stat = row[\"corrected_pVal\"]\n",
    "        # print cluster, svm_weight_score\n",
    "        cluster_series_df = test_df_final.loc[cluster, :]\n",
    "        cluster_series_df[\"chi2\"] = chi2_statistic\n",
    "        cluster_series_df[\"pVal\"] = pVal_statistic\n",
    "        cluster_series_df[\"pVal_corrected\"] = pVal_corrected_stat\n",
    "        cluster_series_df.drop([\"name_to_rv\"], inplace=True)\n",
    "        chi2_test_top50_df = chi2_test_top50_df.append(cluster_series_df)\n",
    "    \n",
    "    new_column_order = [\"chi2\", \"pVal\",\"pVal_corrected\", \"clust_to_rv\", \"gene_name\", \"product\", \"cog\", \"count\",  \"ortho\", \n",
    "                        \"pan\",  \"refseq\", \"score\"]\n",
    "    chi2_test_top50_df = chi2_test_top50_df[new_column_order]\n",
    "    \n",
    "    chi2_test_top50_df.to_excel(writer, sheet_name = drug_name+\"_chi2\", index=True)\n",
    "    \n",
    "writer.save()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 9. Generate ANOVA df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "506\n",
      "1057\n",
      "isoniazid\n",
      "removing...  107 clusters\n",
      "Before removing low variance:  (1563, 9866)\n",
      "After removing low variance:  (1563, 8571)\n",
      "(1533, 3)\n",
      "848\n",
      "492\n",
      "ethambutol\n",
      "removing...  99 clusters\n",
      "Before removing low variance:  (1340, 9874)\n",
      "After removing low variance:  (1340, 8453)\n",
      "(1523, 3)\n",
      "578\n",
      "983\n",
      "rifampicin\n",
      "removing...  92 clusters\n",
      "Before removing low variance:  (1561, 9881)\n",
      "After removing low variance:  (1561, 8581)\n",
      "(1794, 3)\n",
      "92\n",
      "137\n",
      "pyrazinamide\n",
      "removing...  97 clusters\n",
      "Before removing low variance:  (229, 9876)\n",
      "After removing low variance:  (229, 6712)\n",
      "(97, 3)\n",
      "353\n",
      "209\n",
      "ethionamide\n",
      "removing...  101 clusters\n",
      "Before removing low variance:  (562, 9872)\n",
      "After removing low variance:  (562, 7537)\n",
      "(221, 3)\n",
      "554\n",
      "302\n",
      "ofloxacin\n",
      "removing...  102 clusters\n",
      "Before removing low variance:  (856, 9871)\n",
      "After removing low variance:  (856, 7792)\n",
      "(214, 3)\n",
      "295\n",
      "80\n",
      "4-aminosalicylic_acid\n",
      "removing...  109 clusters\n",
      "Before removing low variance:  (375, 9864)\n",
      "After removing low variance:  (375, 7234)\n",
      "(226, 3)\n",
      "732\n",
      "663\n",
      "streptomycin\n",
      "removing...  98 clusters\n",
      "Before removing low variance:  (1395, 9875)\n",
      "After removing low variance:  (1395, 8338)\n",
      "(1433, 3)\n",
      "257\n",
      "142\n",
      "amikacin\n",
      "removing...  115 clusters\n",
      "Before removing low variance:  (399, 9858)\n",
      "After removing low variance:  (399, 6794)\n",
      "(247, 3)\n",
      "262\n",
      "71\n",
      "cycloserine\n",
      "removing...  115 clusters\n",
      "Before removing low variance:  (333, 9858)\n",
      "After removing low variance:  (333, 6735)\n",
      "(14, 3)\n",
      "550\n",
      "278\n",
      "kanamycin\n",
      "removing...  115 clusters\n",
      "Before removing low variance:  (828, 9858)\n",
      "After removing low variance:  (828, 7753)\n",
      "(406, 3)\n",
      "237\n",
      "141\n",
      "capreomycin\n",
      "removing...  115 clusters\n",
      "Before removing low variance:  (378, 9858)\n",
      "After removing low variance:  (378, 6942)\n",
      "(180, 3)\n"
     ]
    }
   ],
   "source": [
    "# writer = pd.ExcelWriter('Supplementary_files/ANOVA_Ftest_allele_AMR_associations.xlsx')\n",
    "writer = pd.ExcelWriter('DataFile01_c.xlsx')\n",
    "\n",
    "for drug_name in [\"isoniazid\", \"ethambutol\", \"rifampicin\", \"pyrazinamide\", \"ethionamide\", \n",
    "                  \"ofloxacin\", \"4-aminosalicylic_acid\", \"streptomycin\", \"amikacin\",\"cycloserine\", \n",
    "                  \"kanamycin\", \"capreomycin\"]: \n",
    "    ANOVA_test = pd.DataFrame()\n",
    "    \n",
    "    if drug_name == \"MDR\" or drug_name == \"XDR\":\n",
    "        X1, y1 = get_MDR_training(resistance_data, data_to_plot, drug_name)\n",
    "    else:\n",
    "        y1, X1 = get_target_data(data_to_plot, resistance_data, drug_name)\n",
    "        X1 = remove_key_antibiotic_clusters(X1, drug_name)\n",
    "    \n",
    "    print \"Before removing low variance: \", X1.shape\n",
    "    selector = VarianceThreshold(threshold=.01)\n",
    "    selector.fit_transform(X1)\n",
    "    X1 = X1[X1.columns[selector.get_support()]].copy()\n",
    "    print \"After removing low variance: \", X1.shape\n",
    "    \n",
    "    # chi2_test = pd.DataFrame(chi2(X1, y1)[0]).T\n",
    "    ANOVA_test = pd.DataFrame(list(f_classif(X1, y1)))\n",
    "    ANOVA_test.columns = X1.columns\n",
    "    ANOVA_test = ANOVA_test.sort_values([0, 1], axis=1, ascending=False).T\n",
    "    ANOVA_test.columns = [\"F_value\", \"pVal\"]\n",
    "    \n",
    "    # bonferroni multiple-testing correction\n",
    "    rejected_list, pvalue_corrected_list, alphaC, alphacBonf = statsmodels.sandbox.stats.multicomp.multipletests(ANOVA_test[\"pVal\"], \n",
    "                                                                                          alpha=0.005, method='bonferroni', is_sorted=False)\n",
    "\n",
    "    ANOVA_test_corrected = ANOVA_test[rejected_list].copy()\n",
    "    ANOVA_test_corrected[\"corrected_pVal\"] = pvalue_corrected_list[rejected_list]\n",
    "    print ANOVA_test_corrected.shape\n",
    "    \n",
    "    ANOVA_test_top50_df = pd.DataFrame()\n",
    "    for allele, row in ANOVA_test_corrected.iloc[:40, :].iterrows():\n",
    "        cluster, allele_id = allele.split(\"_\")[0].replace(\"ter\", \"ter \"), allele.split(\"_\")[1]\n",
    "        FVal_statistic = row[\"F_value\"]\n",
    "        pVal_statistic = row[\"pVal\"]\n",
    "        pVal_corrected_stat = row[\"corrected_pVal\"]\n",
    "        # print cluster, svm_weight_score\n",
    "        cluster_series_df = test_df_final.loc[cluster, :]\n",
    "        cluster_series_df[\"F_value\"] = FVal_statistic\n",
    "        cluster_series_df[\"pVal\"] = pVal_statistic\n",
    "        cluster_series_df[\"pVal_corrected\"] = pVal_corrected_stat\n",
    "        cluster_series_df.drop([\"name_to_rv\"], inplace=True)\n",
    "        ANOVA_test_top50_df = ANOVA_test_top50_df.append(cluster_series_df)\n",
    "    \n",
    "    new_column_order = [\"F_value\", \"pVal\",\"pVal_corrected\", \"clust_to_rv\", \"gene_name\", \"product\", \"cog\", \"count\",  \"ortho\", \n",
    "                        \"pan\",  \"refseq\", \"score\"]\n",
    "    ANOVA_test_top50_df = ANOVA_test_top50_df[new_column_order]\n",
    "    \n",
    "    ANOVA_test_top50_df.to_excel(writer, sheet_name = drug_name+\"_ANOVA\", index=True)\n",
    "    \n",
    "writer.save()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Write all three associations types to a single file!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "writer = pd.ExcelWriter('Supplementary_files/Supplementary_Data_File_1.xlsx')\n",
    "for drug_name in [\"isoniazid\", \n",
    "                  \"ethambutol\", \"rifampicin\", \"pyrazinamide\", \"ethionamide\", \n",
    "                  \"ofloxacin\", \"4-aminosalicylic_acid\", \"streptomycin\", \"amikacin\",\"cycloserine\", \n",
    "                  \"kanamycin\", \"capreomycin\"]: \n",
    "    \n",
    "#     if drug_name == \"4-aminosalicylic_acid\":\n",
    "#         MI_excel_df = pd.read_excel(\"DataFile01_a.xlsx\", sheet_name=\"MI_4amniosalicylic\")\n",
    "#     else:\n",
    "    MI_excel_df = pd.read_excel(\"DataFile01_a.xlsx\", sheet_name=drug_name+\"_MI\")\n",
    "        \n",
    "    Chi2_excel_df = pd.read_excel(\"DataFile01_b.xlsx\", sheet_name=drug_name+\"_chi2\")\n",
    "    ANOVA_excel_df = pd.read_excel(\"DataFile01_c.xlsx\", sheet_name=drug_name+\"_ANOVA\")\n",
    "    \n",
    "    MI_cols = [\"MI sum\", \"clust_to_rv\",\"gene_name\",\"product\"]\n",
    "    chi2_cols = [\"chi2\",\"pVal\",\"pVal_corrected\",\"clust_to_rv\",\"gene_name\",\"product\"]\n",
    "    ANOVA_cols = [\"F_value\",\"pVal\",\"pVal_corrected\",\"clust_to_rv\",\"gene_name\",\"product\"]\n",
    "    \n",
    "    sheet_header_str = [\"\",\"Mutual Information\"] + [\"\"]*5 + [\"Chi-squared\"] + [\"\"]*7 + [\"ANOVA F-test\"]\n",
    "    pd.DataFrame(sheet_header_str, columns=[\"Association\"]).T.to_excel(writer,sheet_name = drug_name, \n",
    "                                                                       header=False,\n",
    "                                                                       index=False, startrow=0)\n",
    "    MI_excel_df[MI_cols].to_excel(writer, sheet_name = drug_name, index=True, startrow=1, startcol=1)\n",
    "    Chi2_excel_df[chi2_cols].to_excel(writer, sheet_name = drug_name, index=True, startrow=1, startcol=7)\n",
    "    ANOVA_excel_df[ANOVA_cols].to_excel(writer, sheet_name = drug_name, index=True, startrow=1, startcol=15)\n",
    "    \n",
    "writer.save()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "labeled_2cluster_df.to_excel(writer, sheet_name = antibiotic_name, index=True, startrow=row_index_location,  startcol=20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Mutual Information</th>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   0\n",
       "Mutual Information  "
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame.from_dict({\"Mutual Information\": \"\"}, orient=\"index\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Association</th>\n",
       "      <td>Mutual Information</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Chi-squared</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ANOVA F-test</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             0  1  2  3  4  5  6            7  8  9  10 11 12  \\\n",
       "Association  Mutual Information                    Chi-squared                  \n",
       "\n",
       "            13            14  \n",
       "Association     ANOVA F-test  "
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame([\"Mutual Information\"] + [\"\"]*6 + [\"Chi-squared\"] + [\"\"]*6 + [\"ANOVA F-test\"], columns=[\"Association\"]).T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Mutual Information',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " 'Chi-squared',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " '',\n",
       " 'ANOVA F-test']"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "\n",
    "MI_excel_df = pd.read_excel(\"DataFile01_a.xlsx\", sheet_name=\"MI_isoniazid\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MI sum</th>\n",
       "      <th>clust_to_rv</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>ortho</th>\n",
       "      <th>cog</th>\n",
       "      <th>product</th>\n",
       "      <th>refseq</th>\n",
       "      <th>count</th>\n",
       "      <th>score</th>\n",
       "      <th>name_to_rv</th>\n",
       "      <th>pan</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Cluster 491</th>\n",
       "      <td>0.585352</td>\n",
       "      <td>RV0667</td>\n",
       "      <td>RV0667</td>\n",
       "      <td>419947.MtubH3_010100005402</td>\n",
       "      <td>K</td>\n",
       "      <td>DNA-directed RNA polymerase beta subunit (EC 2...</td>\n",
       "      <td>BA24_03142</td>\n",
       "      <td>1595</td>\n",
       "      <td>2475.4</td>\n",
       "      <td>0</td>\n",
       "      <td>Core</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 1116</th>\n",
       "      <td>0.428611</td>\n",
       "      <td>RV1908C</td>\n",
       "      <td>RV1908C</td>\n",
       "      <td>83332.Rv1908c</td>\n",
       "      <td>P</td>\n",
       "      <td>Catalase (EC 1.11.1.6) / Peroxidase (EC 1.11.1.7)</td>\n",
       "      <td>Z578_01418</td>\n",
       "      <td>1592</td>\n",
       "      <td>1674.2</td>\n",
       "      <td>0</td>\n",
       "      <td>Core</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 551</th>\n",
       "      <td>0.412052</td>\n",
       "      <td>RV3795</td>\n",
       "      <td>RV3795</td>\n",
       "      <td>419947.MRA_3835</td>\n",
       "      <td>M</td>\n",
       "      <td>Integral membrane indolylacetylinositol arabin...</td>\n",
       "      <td>V199_03742</td>\n",
       "      <td>1593</td>\n",
       "      <td>2421.4</td>\n",
       "      <td>0</td>\n",
       "      <td>Core</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 3930</th>\n",
       "      <td>0.248406</td>\n",
       "      <td>RV2043C</td>\n",
       "      <td>RV2043C</td>\n",
       "      <td>419947.MRA_2058</td>\n",
       "      <td>Q</td>\n",
       "      <td>Nicotinamidase (EC 3.5.1.19)</td>\n",
       "      <td>None</td>\n",
       "      <td>1283</td>\n",
       "      <td>414.4</td>\n",
       "      <td>0</td>\n",
       "      <td>Acces</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 7435</th>\n",
       "      <td>0.177361</td>\n",
       "      <td>RV0682</td>\n",
       "      <td>RV0682</td>\n",
       "      <td>419947.MRA_0691</td>\n",
       "      <td>J</td>\n",
       "      <td>SSU ribosomal protein S12p (S23e)</td>\n",
       "      <td>Z578_02043</td>\n",
       "      <td>1595</td>\n",
       "      <td>258.4</td>\n",
       "      <td>0</td>\n",
       "      <td>Core</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                MI sum clust_to_rv gene_name                       ortho cog  \\\n",
       "Cluster 491   0.585352      RV0667    RV0667  419947.MtubH3_010100005402   K   \n",
       "Cluster 1116  0.428611     RV1908C   RV1908C               83332.Rv1908c   P   \n",
       "Cluster 551   0.412052      RV3795    RV3795             419947.MRA_3835   M   \n",
       "Cluster 3930  0.248406     RV2043C   RV2043C             419947.MRA_2058   Q   \n",
       "Cluster 7435  0.177361      RV0682    RV0682             419947.MRA_0691   J   \n",
       "\n",
       "                                                        product      refseq  \\\n",
       "Cluster 491   DNA-directed RNA polymerase beta subunit (EC 2...  BA24_03142   \n",
       "Cluster 1116  Catalase (EC 1.11.1.6) / Peroxidase (EC 1.11.1.7)  Z578_01418   \n",
       "Cluster 551   Integral membrane indolylacetylinositol arabin...  V199_03742   \n",
       "Cluster 3930                       Nicotinamidase (EC 3.5.1.19)        None   \n",
       "Cluster 7435                  SSU ribosomal protein S12p (S23e)  Z578_02043   \n",
       "\n",
       "              count   score  name_to_rv    pan  \n",
       "Cluster 491    1595  2475.4           0   Core  \n",
       "Cluster 1116   1592  1674.2           0   Core  \n",
       "Cluster 551    1593  2421.4           0   Core  \n",
       "Cluster 3930   1283   414.4           0  Acces  \n",
       "Cluster 7435   1595   258.4           0   Core  "
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MI_excel_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chi2</th>\n",
       "      <th>pVal</th>\n",
       "      <th>pVal_corrected</th>\n",
       "      <th>clust_to_rv</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>product</th>\n",
       "      <th>cog</th>\n",
       "      <th>count</th>\n",
       "      <th>ortho</th>\n",
       "      <th>pan</th>\n",
       "      <th>refseq</th>\n",
       "      <th>score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Cluster 1116</th>\n",
       "      <td>272.915547</td>\n",
       "      <td>2.627186e-61</td>\n",
       "      <td>2.251761e-57</td>\n",
       "      <td>RV1908C</td>\n",
       "      <td>RV1908C</td>\n",
       "      <td>Catalase (EC 1.11.1.6) / Peroxidase (EC 1.11.1.7)</td>\n",
       "      <td>P</td>\n",
       "      <td>1592</td>\n",
       "      <td>83332.Rv1908c</td>\n",
       "      <td>Core</td>\n",
       "      <td>Z578_01418</td>\n",
       "      <td>1674.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 1116</th>\n",
       "      <td>230.246391</td>\n",
       "      <td>5.267552e-52</td>\n",
       "      <td>4.514819e-48</td>\n",
       "      <td>RV1908C</td>\n",
       "      <td>RV1908C</td>\n",
       "      <td>Catalase (EC 1.11.1.6) / Peroxidase (EC 1.11.1.7)</td>\n",
       "      <td>P</td>\n",
       "      <td>1592</td>\n",
       "      <td>83332.Rv1908c</td>\n",
       "      <td>Core</td>\n",
       "      <td>Z578_01418</td>\n",
       "      <td>1674.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 2387</th>\n",
       "      <td>152.831095</td>\n",
       "      <td>4.170489e-35</td>\n",
       "      <td>3.574526e-31</td>\n",
       "      <td>RV2462C</td>\n",
       "      <td>RV2462C</td>\n",
       "      <td>Cell division trigger factor (EC 5.2.1.8)</td>\n",
       "      <td>O</td>\n",
       "      <td>1595</td>\n",
       "      <td>419947.MRA_2488</td>\n",
       "      <td>Core</td>\n",
       "      <td>Z578_02233</td>\n",
       "      <td>974.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 1329</th>\n",
       "      <td>152.136980</td>\n",
       "      <td>5.914091e-35</td>\n",
       "      <td>5.068967e-31</td>\n",
       "      <td>RV1736C</td>\n",
       "      <td>RV1161</td>\n",
       "      <td>Respiratory nitrate reductase alpha chain (EC ...</td>\n",
       "      <td>C</td>\n",
       "      <td>1593</td>\n",
       "      <td>710685.MycrhN_6026</td>\n",
       "      <td>Core</td>\n",
       "      <td>Z578_00643</td>\n",
       "      <td>553.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 591</th>\n",
       "      <td>151.106004</td>\n",
       "      <td>9.936176e-35</td>\n",
       "      <td>8.516296e-31</td>\n",
       "      <td>RV0932C</td>\n",
       "      <td>RV0931C</td>\n",
       "      <td>Serine/threonine-protein kinase PknD (EC 2.7.1...</td>\n",
       "      <td>T</td>\n",
       "      <td>1595</td>\n",
       "      <td>419947.MRA_0939</td>\n",
       "      <td>Core</td>\n",
       "      <td>AO93_01423</td>\n",
       "      <td>1471.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 6719</th>\n",
       "      <td>150.644727</td>\n",
       "      <td>1.253256e-34</td>\n",
       "      <td>1.074166e-30</td>\n",
       "      <td>RV2533C</td>\n",
       "      <td>RV2533C</td>\n",
       "      <td>Transcription termination protein NusB</td>\n",
       "      <td>K</td>\n",
       "      <td>1592</td>\n",
       "      <td>419947.MRA_2561</td>\n",
       "      <td>Core</td>\n",
       "      <td>AJ54_03676</td>\n",
       "      <td>283.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 2133</th>\n",
       "      <td>150.229502</td>\n",
       "      <td>1.544530e-34</td>\n",
       "      <td>1.323817e-30</td>\n",
       "      <td>RV0143C</td>\n",
       "      <td>MMAR_0353</td>\n",
       "      <td>Chloride channel protein</td>\n",
       "      <td>P</td>\n",
       "      <td>1593</td>\n",
       "      <td>216594.MMAR_0353</td>\n",
       "      <td>Core</td>\n",
       "      <td>N126_01920</td>\n",
       "      <td>868.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 1527</th>\n",
       "      <td>148.425516</td>\n",
       "      <td>3.829271e-34</td>\n",
       "      <td>3.282068e-30</td>\n",
       "      <td>RV1069C</td>\n",
       "      <td>MT1099</td>\n",
       "      <td>Predicted membrane protein (DUF2319)</td>\n",
       "      <td>S</td>\n",
       "      <td>1594</td>\n",
       "      <td>419947.MRA_1079</td>\n",
       "      <td>Core</td>\n",
       "      <td>None</td>\n",
       "      <td>1350.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 4383</th>\n",
       "      <td>148.175464</td>\n",
       "      <td>4.342859e-34</td>\n",
       "      <td>3.722264e-30</td>\n",
       "      <td>RV3551</td>\n",
       "      <td>RV3551</td>\n",
       "      <td>Putative CoA-transferase subunit alpha Rv3551/...</td>\n",
       "      <td>I</td>\n",
       "      <td>1595</td>\n",
       "      <td>419947.MRA_3590</td>\n",
       "      <td>Core</td>\n",
       "      <td>Z578_00037</td>\n",
       "      <td>662.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 464</th>\n",
       "      <td>147.514785</td>\n",
       "      <td>6.056168e-34</td>\n",
       "      <td>5.190741e-30</td>\n",
       "      <td>RV1020</td>\n",
       "      <td>RV1020</td>\n",
       "      <td>Transcription-repair coupling factor</td>\n",
       "      <td>L</td>\n",
       "      <td>1595</td>\n",
       "      <td>419947.MRA_1028</td>\n",
       "      <td>Core</td>\n",
       "      <td>Z578_02842</td>\n",
       "      <td>2721.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 1268</th>\n",
       "      <td>147.327479</td>\n",
       "      <td>6.654926e-34</td>\n",
       "      <td>5.703937e-30</td>\n",
       "      <td>RV0457C</td>\n",
       "      <td>PEPTIDASE</td>\n",
       "      <td>Prolyl endopeptidase (EC 3.4.21.26)</td>\n",
       "      <td>E</td>\n",
       "      <td>1595</td>\n",
       "      <td>419947.MRA_0463</td>\n",
       "      <td>Core</td>\n",
       "      <td>Q625_02988</td>\n",
       "      <td>1557.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 4843</th>\n",
       "      <td>147.327479</td>\n",
       "      <td>6.654926e-34</td>\n",
       "      <td>5.703937e-30</td>\n",
       "      <td>RV0012</td>\n",
       "      <td>MRA_0014</td>\n",
       "      <td>FIG215594: Membrane spanning protein</td>\n",
       "      <td>S</td>\n",
       "      <td>1595</td>\n",
       "      <td>419947.MRA_0014</td>\n",
       "      <td>Core</td>\n",
       "      <td>BA33_02036</td>\n",
       "      <td>530.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 1751</th>\n",
       "      <td>146.992047</td>\n",
       "      <td>7.878977e-34</td>\n",
       "      <td>6.753071e-30</td>\n",
       "      <td>RV0897C</td>\n",
       "      <td>MKANA1_010100023024</td>\n",
       "      <td>Beta-carotene ketolase</td>\n",
       "      <td>Q</td>\n",
       "      <td>1595</td>\n",
       "      <td>557599.MkanA1_010100023024</td>\n",
       "      <td>Core</td>\n",
       "      <td>None</td>\n",
       "      <td>896.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 1241</th>\n",
       "      <td>146.671016</td>\n",
       "      <td>9.260817e-34</td>\n",
       "      <td>7.937446e-30</td>\n",
       "      <td>RV0822C</td>\n",
       "      <td>MT0844</td>\n",
       "      <td>Cell envelope-associated transcriptional atten...</td>\n",
       "      <td>K</td>\n",
       "      <td>1595</td>\n",
       "      <td>419947.MRA_0832</td>\n",
       "      <td>Core</td>\n",
       "      <td>Z578_03954</td>\n",
       "      <td>1535.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 4936</th>\n",
       "      <td>146.363890</td>\n",
       "      <td>1.080911e-33</td>\n",
       "      <td>9.264490e-30</td>\n",
       "      <td>RV0484C</td>\n",
       "      <td>SDH</td>\n",
       "      <td>Serine 3-dehydrogenase</td>\n",
       "      <td>L</td>\n",
       "      <td>1595</td>\n",
       "      <td>419947.MRA_0491</td>\n",
       "      <td>Core</td>\n",
       "      <td>X103_00504</td>\n",
       "      <td>525.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 5551</th>\n",
       "      <td>146.363890</td>\n",
       "      <td>1.080911e-33</td>\n",
       "      <td>9.264490e-30</td>\n",
       "      <td>RV3301C</td>\n",
       "      <td>PHOU</td>\n",
       "      <td>Phosphate transport system regulatory protein ...</td>\n",
       "      <td>P</td>\n",
       "      <td>1595</td>\n",
       "      <td>83332.Rv3301c</td>\n",
       "      <td>Core</td>\n",
       "      <td>Z578_01269</td>\n",
       "      <td>470.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 6988</th>\n",
       "      <td>146.070195</td>\n",
       "      <td>1.253128e-33</td>\n",
       "      <td>1.074056e-29</td>\n",
       "      <td>RV1262C</td>\n",
       "      <td>AS9A_3300</td>\n",
       "      <td>Bis(5'-nucleosyl)-tetraphosphatase (asymmetric...</td>\n",
       "      <td>['F', 'G']</td>\n",
       "      <td>1593</td>\n",
       "      <td>443218.AS9A_3300</td>\n",
       "      <td>Core</td>\n",
       "      <td>P456_00290</td>\n",
       "      <td>125.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 3233</th>\n",
       "      <td>145.170368</td>\n",
       "      <td>1.971126e-33</td>\n",
       "      <td>1.689452e-29</td>\n",
       "      <td>RV1895</td>\n",
       "      <td>MT1946</td>\n",
       "      <td>Threonine dehydrogenase and related Zn-depende...</td>\n",
       "      <td>C</td>\n",
       "      <td>1591</td>\n",
       "      <td>419947.MRA_1907</td>\n",
       "      <td>Core</td>\n",
       "      <td>Z578_01405</td>\n",
       "      <td>841.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 1023</th>\n",
       "      <td>145.114483</td>\n",
       "      <td>2.027366e-33</td>\n",
       "      <td>1.737655e-29</td>\n",
       "      <td>RV3693</td>\n",
       "      <td>RV3692</td>\n",
       "      <td>putative membrane protein</td>\n",
       "      <td>S</td>\n",
       "      <td>1595</td>\n",
       "      <td>419947.MtubH3_010100023495</td>\n",
       "      <td>Core</td>\n",
       "      <td>AO54_00184</td>\n",
       "      <td>664.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 5365</th>\n",
       "      <td>144.856030</td>\n",
       "      <td>2.309066e-33</td>\n",
       "      <td>1.979101e-29</td>\n",
       "      <td>RV2676C</td>\n",
       "      <td>YWFI</td>\n",
       "      <td>Coproheme decarboxylase HemQ (no EC)</td>\n",
       "      <td>S</td>\n",
       "      <td>1595</td>\n",
       "      <td>83332.Rv2676c</td>\n",
       "      <td>Core</td>\n",
       "      <td>Z578_01557</td>\n",
       "      <td>525.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 5904</th>\n",
       "      <td>144.856030</td>\n",
       "      <td>2.309066e-33</td>\n",
       "      <td>1.979101e-29</td>\n",
       "      <td>RV0133</td>\n",
       "      <td>NRGA</td>\n",
       "      <td>N-acetyltransferase, GCN5-family</td>\n",
       "      <td>K</td>\n",
       "      <td>1592</td>\n",
       "      <td>419947.MRA_0140</td>\n",
       "      <td>Core</td>\n",
       "      <td>Z578_00339</td>\n",
       "      <td>460.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 5565</th>\n",
       "      <td>144.555522</td>\n",
       "      <td>2.686184e-33</td>\n",
       "      <td>2.302328e-29</td>\n",
       "      <td>RV2270</td>\n",
       "      <td>0</td>\n",
       "      <td>Lipoprotein LppN</td>\n",
       "      <td>0</td>\n",
       "      <td>1594</td>\n",
       "      <td>0</td>\n",
       "      <td>Core</td>\n",
       "      <td>T593_02377</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 7705</th>\n",
       "      <td>144.555522</td>\n",
       "      <td>2.686184e-33</td>\n",
       "      <td>2.302328e-29</td>\n",
       "      <td>RV3316</td>\n",
       "      <td>RV3316</td>\n",
       "      <td>Succinate dehydrogenase cytochrome b-556 subunit</td>\n",
       "      <td>C</td>\n",
       "      <td>1595</td>\n",
       "      <td>272631.ML0699</td>\n",
       "      <td>Core</td>\n",
       "      <td>Z578_01285</td>\n",
       "      <td>215.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 1010</th>\n",
       "      <td>144.268371</td>\n",
       "      <td>3.103959e-33</td>\n",
       "      <td>2.660403e-29</td>\n",
       "      <td>RV3481C</td>\n",
       "      <td>RV3480C</td>\n",
       "      <td>Wax ester synthase/acyl-CoA:diacylglycerol acy...</td>\n",
       "      <td>S</td>\n",
       "      <td>1595</td>\n",
       "      <td>83332.Rv3480c</td>\n",
       "      <td>Core</td>\n",
       "      <td>AO98_03593</td>\n",
       "      <td>1118.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 2554</th>\n",
       "      <td>144.268371</td>\n",
       "      <td>3.103959e-33</td>\n",
       "      <td>2.660403e-29</td>\n",
       "      <td>RV0729</td>\n",
       "      <td>SVIR_32990</td>\n",
       "      <td>Putative xylulose kinase (EC 2.7.1.17)</td>\n",
       "      <td>G</td>\n",
       "      <td>1595</td>\n",
       "      <td>471857.Svir_32990</td>\n",
       "      <td>Core</td>\n",
       "      <td>Z578_02092</td>\n",
       "      <td>145.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 3614</th>\n",
       "      <td>144.268371</td>\n",
       "      <td>3.103959e-33</td>\n",
       "      <td>2.660403e-29</td>\n",
       "      <td>RV1841C</td>\n",
       "      <td>SCLAV_0672</td>\n",
       "      <td>Magnesium and cobalt efflux protein CorC</td>\n",
       "      <td>P</td>\n",
       "      <td>1595</td>\n",
       "      <td>419947.MRA_1852</td>\n",
       "      <td>Core</td>\n",
       "      <td>AN87_01613</td>\n",
       "      <td>715.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 4656</th>\n",
       "      <td>144.268371</td>\n",
       "      <td>3.103959e-33</td>\n",
       "      <td>2.660403e-29</td>\n",
       "      <td>RV3079C</td>\n",
       "      <td>MKANA1_010100005130</td>\n",
       "      <td>Hydride transferase 1 (Fragment)</td>\n",
       "      <td>S</td>\n",
       "      <td>1594</td>\n",
       "      <td>557599.MkanA1_010100005130</td>\n",
       "      <td>Core</td>\n",
       "      <td>Z578_01971</td>\n",
       "      <td>538.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 3149</th>\n",
       "      <td>144.070393</td>\n",
       "      <td>3.429261e-33</td>\n",
       "      <td>2.939220e-29</td>\n",
       "      <td>RV0996</td>\n",
       "      <td>MRA_1003</td>\n",
       "      <td>FIG037137: Putative conserved transmembrane pr...</td>\n",
       "      <td>S</td>\n",
       "      <td>1594</td>\n",
       "      <td>419947.MRA_1003</td>\n",
       "      <td>Core</td>\n",
       "      <td>AI72_02495</td>\n",
       "      <td>793.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 4148</th>\n",
       "      <td>144.070393</td>\n",
       "      <td>3.429261e-33</td>\n",
       "      <td>2.939220e-29</td>\n",
       "      <td>RV0604</td>\n",
       "      <td>RV0604</td>\n",
       "      <td>protein of unknown function LppY and LpqO</td>\n",
       "      <td>S</td>\n",
       "      <td>1592</td>\n",
       "      <td>83332.Rv0604</td>\n",
       "      <td>Core</td>\n",
       "      <td>Z578_02958</td>\n",
       "      <td>668.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 4455</th>\n",
       "      <td>143.994123</td>\n",
       "      <td>3.563492e-33</td>\n",
       "      <td>3.054269e-29</td>\n",
       "      <td>RV2877C</td>\n",
       "      <td>RV2877C</td>\n",
       "      <td>Mercury resistance probable Hg transport protein</td>\n",
       "      <td>O</td>\n",
       "      <td>1595</td>\n",
       "      <td>83332.Rv2877c</td>\n",
       "      <td>Core</td>\n",
       "      <td>Z578_03348</td>\n",
       "      <td>577.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 6925</th>\n",
       "      <td>143.994123</td>\n",
       "      <td>3.563492e-33</td>\n",
       "      <td>3.054269e-29</td>\n",
       "      <td>RV0723</td>\n",
       "      <td>RV0723</td>\n",
       "      <td>LSU ribosomal protein L15p (L27Ae)</td>\n",
       "      <td>J</td>\n",
       "      <td>1594</td>\n",
       "      <td>419947.MRA_0731</td>\n",
       "      <td>Core</td>\n",
       "      <td>Z578_02085</td>\n",
       "      <td>306.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 2807</th>\n",
       "      <td>143.673726</td>\n",
       "      <td>4.187224e-33</td>\n",
       "      <td>3.588870e-29</td>\n",
       "      <td>RV0331</td>\n",
       "      <td>RV0331</td>\n",
       "      <td>Oxidoreductase (flavoprotein)</td>\n",
       "      <td>S</td>\n",
       "      <td>1595</td>\n",
       "      <td>83332.Rv0331</td>\n",
       "      <td>Core</td>\n",
       "      <td>AO48_03375</td>\n",
       "      <td>860.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 6872</th>\n",
       "      <td>143.673726</td>\n",
       "      <td>4.187224e-33</td>\n",
       "      <td>3.588870e-29</td>\n",
       "      <td>0</td>\n",
       "      <td>MAB_2302</td>\n",
       "      <td>membrane protein, MmpS family</td>\n",
       "      <td>S</td>\n",
       "      <td>182</td>\n",
       "      <td>561007.MAB_2302</td>\n",
       "      <td>Acces</td>\n",
       "      <td>AN87_01324</td>\n",
       "      <td>134.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 3539</th>\n",
       "      <td>143.351871</td>\n",
       "      <td>4.923757e-33</td>\n",
       "      <td>4.220152e-29</td>\n",
       "      <td>RV3200C</td>\n",
       "      <td>MRA_3237</td>\n",
       "      <td>Potassium channel protein</td>\n",
       "      <td>P</td>\n",
       "      <td>1594</td>\n",
       "      <td>419947.MRA_3237</td>\n",
       "      <td>Core</td>\n",
       "      <td>X154_02944</td>\n",
       "      <td>736.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 1742</th>\n",
       "      <td>143.073529</td>\n",
       "      <td>5.664403e-33</td>\n",
       "      <td>4.854960e-29</td>\n",
       "      <td>RV1204C</td>\n",
       "      <td>NFA53720</td>\n",
       "      <td>ATPase</td>\n",
       "      <td>K</td>\n",
       "      <td>1595</td>\n",
       "      <td>247156.nfa53720</td>\n",
       "      <td>Core</td>\n",
       "      <td>Z578_01702</td>\n",
       "      <td>276.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 2062</th>\n",
       "      <td>143.044267</td>\n",
       "      <td>5.748468e-33</td>\n",
       "      <td>4.927012e-29</td>\n",
       "      <td>RV2484C</td>\n",
       "      <td>MRA_2509</td>\n",
       "      <td>Diacyglycerol O-acyltransferase (EC 2.3.1.20)</td>\n",
       "      <td>S</td>\n",
       "      <td>1594</td>\n",
       "      <td>419947.MRA_2509</td>\n",
       "      <td>Core</td>\n",
       "      <td>AO71_03323</td>\n",
       "      <td>1078.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 4656</th>\n",
       "      <td>143.044267</td>\n",
       "      <td>5.748468e-33</td>\n",
       "      <td>4.927012e-29</td>\n",
       "      <td>RV3079C</td>\n",
       "      <td>MKANA1_010100005130</td>\n",
       "      <td>Hydride transferase 1 (Fragment)</td>\n",
       "      <td>S</td>\n",
       "      <td>1594</td>\n",
       "      <td>557599.MkanA1_010100005130</td>\n",
       "      <td>Core</td>\n",
       "      <td>Z578_01971</td>\n",
       "      <td>538.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 1053</th>\n",
       "      <td>142.750417</td>\n",
       "      <td>6.665017e-33</td>\n",
       "      <td>5.712586e-29</td>\n",
       "      <td>RV2320C</td>\n",
       "      <td>USPA4</td>\n",
       "      <td>Cationic amino acid transport integral membran...</td>\n",
       "      <td>T</td>\n",
       "      <td>1595</td>\n",
       "      <td>83332.Rv2319c</td>\n",
       "      <td>Core</td>\n",
       "      <td>X399_02426</td>\n",
       "      <td>638.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 5788</th>\n",
       "      <td>142.750417</td>\n",
       "      <td>6.665017e-33</td>\n",
       "      <td>5.712586e-29</td>\n",
       "      <td>RV3167C</td>\n",
       "      <td>RV3167C</td>\n",
       "      <td>Transcriptional regulator, AcrR family</td>\n",
       "      <td>K</td>\n",
       "      <td>1595</td>\n",
       "      <td>83332.Rv3167c</td>\n",
       "      <td>Core</td>\n",
       "      <td>Z578_03871</td>\n",
       "      <td>430.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 4975</th>\n",
       "      <td>142.631467</td>\n",
       "      <td>7.076356e-33</td>\n",
       "      <td>6.065144e-29</td>\n",
       "      <td>RV0658C</td>\n",
       "      <td>MRA_0668</td>\n",
       "      <td>putative membrane protein</td>\n",
       "      <td>S</td>\n",
       "      <td>1588</td>\n",
       "      <td>419947.MRA_0668</td>\n",
       "      <td>Acces</td>\n",
       "      <td>None</td>\n",
       "      <td>440.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 460</th>\n",
       "      <td>142.526251</td>\n",
       "      <td>7.461307e-33</td>\n",
       "      <td>6.395086e-29</td>\n",
       "      <td>RV1248C</td>\n",
       "      <td>SROT_0714</td>\n",
       "      <td>Dihydrolipoamide succinyltransferase component...</td>\n",
       "      <td>C</td>\n",
       "      <td>1595</td>\n",
       "      <td>640132.Srot_0714</td>\n",
       "      <td>Core</td>\n",
       "      <td>AN41_01826</td>\n",
       "      <td>1906.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 2192</th>\n",
       "      <td>142.526251</td>\n",
       "      <td>7.461307e-33</td>\n",
       "      <td>6.395086e-29</td>\n",
       "      <td>RV1979C</td>\n",
       "      <td>ML1974</td>\n",
       "      <td>Amino acid permease</td>\n",
       "      <td>E</td>\n",
       "      <td>1592</td>\n",
       "      <td>272631.ML1974</td>\n",
       "      <td>Core</td>\n",
       "      <td>Z550_00860</td>\n",
       "      <td>745.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 1133</th>\n",
       "      <td>142.469847</td>\n",
       "      <td>7.676225e-33</td>\n",
       "      <td>6.579292e-29</td>\n",
       "      <td>RV1326C</td>\n",
       "      <td>RV1326C</td>\n",
       "      <td>1,4-alpha-glucan (glycogen) branching enzyme, ...</td>\n",
       "      <td>G</td>\n",
       "      <td>1595</td>\n",
       "      <td>419947.MRA_1334</td>\n",
       "      <td>Core</td>\n",
       "      <td>Z578_03418</td>\n",
       "      <td>1748.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 1346</th>\n",
       "      <td>142.469847</td>\n",
       "      <td>7.676225e-33</td>\n",
       "      <td>6.579292e-29</td>\n",
       "      <td>RV3535C</td>\n",
       "      <td>RV3469C</td>\n",
       "      <td>4-hydroxy-2-oxovalerate aldolase (EC 4.1.3.39)</td>\n",
       "      <td>Q</td>\n",
       "      <td>1595</td>\n",
       "      <td>419947.MRA_3573</td>\n",
       "      <td>Core</td>\n",
       "      <td>AL99_00361</td>\n",
       "      <td>756.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 2497</th>\n",
       "      <td>142.469847</td>\n",
       "      <td>7.676225e-33</td>\n",
       "      <td>6.579292e-29</td>\n",
       "      <td>RV3170</td>\n",
       "      <td>RV3170</td>\n",
       "      <td>Amine oxidase [flavin-containing] A (EC 1.4.3.4)</td>\n",
       "      <td>E</td>\n",
       "      <td>1594</td>\n",
       "      <td>216594.MMAR_1389</td>\n",
       "      <td>Core</td>\n",
       "      <td>Z578_03874</td>\n",
       "      <td>830.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 3410</th>\n",
       "      <td>142.469847</td>\n",
       "      <td>7.676225e-33</td>\n",
       "      <td>6.579292e-29</td>\n",
       "      <td>RV0205</td>\n",
       "      <td>RV0205</td>\n",
       "      <td>FIG01121868: Possible membrane protein, Rv0205</td>\n",
       "      <td>S</td>\n",
       "      <td>1592</td>\n",
       "      <td>83332.Rv0205</td>\n",
       "      <td>Core</td>\n",
       "      <td>Z578_00414</td>\n",
       "      <td>742.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 4357</th>\n",
       "      <td>142.469847</td>\n",
       "      <td>7.676225e-33</td>\n",
       "      <td>6.579292e-29</td>\n",
       "      <td>RV1596</td>\n",
       "      <td>RV1596</td>\n",
       "      <td>Quinolinate phosphoribosyltransferase [decarbo...</td>\n",
       "      <td>H</td>\n",
       "      <td>1593</td>\n",
       "      <td>419947.MRA_1606</td>\n",
       "      <td>Core</td>\n",
       "      <td>P926_01348</td>\n",
       "      <td>596.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 4791</th>\n",
       "      <td>142.469847</td>\n",
       "      <td>7.676225e-33</td>\n",
       "      <td>6.579292e-29</td>\n",
       "      <td>RV0167</td>\n",
       "      <td>RV0167</td>\n",
       "      <td>Conserved hypothetical integral membrane prote...</td>\n",
       "      <td>Q</td>\n",
       "      <td>1594</td>\n",
       "      <td>83332.Rv0167</td>\n",
       "      <td>Core</td>\n",
       "      <td>Z578_00374</td>\n",
       "      <td>548.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 5117</th>\n",
       "      <td>142.469847</td>\n",
       "      <td>7.676225e-33</td>\n",
       "      <td>6.579292e-29</td>\n",
       "      <td>RV2927C</td>\n",
       "      <td>RV2927C</td>\n",
       "      <td>FIG00814129: Possible chaperone</td>\n",
       "      <td>D</td>\n",
       "      <td>1595</td>\n",
       "      <td>419947.MRA_2954</td>\n",
       "      <td>Core</td>\n",
       "      <td>None</td>\n",
       "      <td>506.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cluster 2335</th>\n",
       "      <td>142.202105</td>\n",
       "      <td>8.783958e-33</td>\n",
       "      <td>7.528730e-29</td>\n",
       "      <td>RV0777</td>\n",
       "      <td>RV0777</td>\n",
       "      <td>Adenylosuccinate lyase (EC 4.3.2.2) @ SAICAR l...</td>\n",
       "      <td>F</td>\n",
       "      <td>1595</td>\n",
       "      <td>419947.MRA_0786</td>\n",
       "      <td>Core</td>\n",
       "      <td>Z578_03539</td>\n",
       "      <td>1011.4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    chi2          pVal  pVal_corrected clust_to_rv  \\\n",
       "Cluster 1116  272.915547  2.627186e-61    2.251761e-57     RV1908C   \n",
       "Cluster 1116  230.246391  5.267552e-52    4.514819e-48     RV1908C   \n",
       "Cluster 2387  152.831095  4.170489e-35    3.574526e-31     RV2462C   \n",
       "Cluster 1329  152.136980  5.914091e-35    5.068967e-31     RV1736C   \n",
       "Cluster 591   151.106004  9.936176e-35    8.516296e-31     RV0932C   \n",
       "Cluster 6719  150.644727  1.253256e-34    1.074166e-30     RV2533C   \n",
       "Cluster 2133  150.229502  1.544530e-34    1.323817e-30     RV0143C   \n",
       "Cluster 1527  148.425516  3.829271e-34    3.282068e-30     RV1069C   \n",
       "Cluster 4383  148.175464  4.342859e-34    3.722264e-30      RV3551   \n",
       "Cluster 464   147.514785  6.056168e-34    5.190741e-30      RV1020   \n",
       "Cluster 1268  147.327479  6.654926e-34    5.703937e-30     RV0457C   \n",
       "Cluster 4843  147.327479  6.654926e-34    5.703937e-30      RV0012   \n",
       "Cluster 1751  146.992047  7.878977e-34    6.753071e-30     RV0897C   \n",
       "Cluster 1241  146.671016  9.260817e-34    7.937446e-30     RV0822C   \n",
       "Cluster 4936  146.363890  1.080911e-33    9.264490e-30     RV0484C   \n",
       "Cluster 5551  146.363890  1.080911e-33    9.264490e-30     RV3301C   \n",
       "Cluster 6988  146.070195  1.253128e-33    1.074056e-29     RV1262C   \n",
       "Cluster 3233  145.170368  1.971126e-33    1.689452e-29      RV1895   \n",
       "Cluster 1023  145.114483  2.027366e-33    1.737655e-29      RV3693   \n",
       "Cluster 5365  144.856030  2.309066e-33    1.979101e-29     RV2676C   \n",
       "Cluster 5904  144.856030  2.309066e-33    1.979101e-29      RV0133   \n",
       "Cluster 5565  144.555522  2.686184e-33    2.302328e-29      RV2270   \n",
       "Cluster 7705  144.555522  2.686184e-33    2.302328e-29      RV3316   \n",
       "Cluster 1010  144.268371  3.103959e-33    2.660403e-29     RV3481C   \n",
       "Cluster 2554  144.268371  3.103959e-33    2.660403e-29      RV0729   \n",
       "Cluster 3614  144.268371  3.103959e-33    2.660403e-29     RV1841C   \n",
       "Cluster 4656  144.268371  3.103959e-33    2.660403e-29     RV3079C   \n",
       "Cluster 3149  144.070393  3.429261e-33    2.939220e-29      RV0996   \n",
       "Cluster 4148  144.070393  3.429261e-33    2.939220e-29      RV0604   \n",
       "Cluster 4455  143.994123  3.563492e-33    3.054269e-29     RV2877C   \n",
       "Cluster 6925  143.994123  3.563492e-33    3.054269e-29      RV0723   \n",
       "Cluster 2807  143.673726  4.187224e-33    3.588870e-29      RV0331   \n",
       "Cluster 6872  143.673726  4.187224e-33    3.588870e-29           0   \n",
       "Cluster 3539  143.351871  4.923757e-33    4.220152e-29     RV3200C   \n",
       "Cluster 1742  143.073529  5.664403e-33    4.854960e-29     RV1204C   \n",
       "Cluster 2062  143.044267  5.748468e-33    4.927012e-29     RV2484C   \n",
       "Cluster 4656  143.044267  5.748468e-33    4.927012e-29     RV3079C   \n",
       "Cluster 1053  142.750417  6.665017e-33    5.712586e-29     RV2320C   \n",
       "Cluster 5788  142.750417  6.665017e-33    5.712586e-29     RV3167C   \n",
       "Cluster 4975  142.631467  7.076356e-33    6.065144e-29     RV0658C   \n",
       "Cluster 460   142.526251  7.461307e-33    6.395086e-29     RV1248C   \n",
       "Cluster 2192  142.526251  7.461307e-33    6.395086e-29     RV1979C   \n",
       "Cluster 1133  142.469847  7.676225e-33    6.579292e-29     RV1326C   \n",
       "Cluster 1346  142.469847  7.676225e-33    6.579292e-29     RV3535C   \n",
       "Cluster 2497  142.469847  7.676225e-33    6.579292e-29      RV3170   \n",
       "Cluster 3410  142.469847  7.676225e-33    6.579292e-29      RV0205   \n",
       "Cluster 4357  142.469847  7.676225e-33    6.579292e-29      RV1596   \n",
       "Cluster 4791  142.469847  7.676225e-33    6.579292e-29      RV0167   \n",
       "Cluster 5117  142.469847  7.676225e-33    6.579292e-29     RV2927C   \n",
       "Cluster 2335  142.202105  8.783958e-33    7.528730e-29      RV0777   \n",
       "\n",
       "                        gene_name  \\\n",
       "Cluster 1116              RV1908C   \n",
       "Cluster 1116              RV1908C   \n",
       "Cluster 2387              RV2462C   \n",
       "Cluster 1329               RV1161   \n",
       "Cluster 591               RV0931C   \n",
       "Cluster 6719              RV2533C   \n",
       "Cluster 2133            MMAR_0353   \n",
       "Cluster 1527               MT1099   \n",
       "Cluster 4383               RV3551   \n",
       "Cluster 464                RV1020   \n",
       "Cluster 1268            PEPTIDASE   \n",
       "Cluster 4843             MRA_0014   \n",
       "Cluster 1751  MKANA1_010100023024   \n",
       "Cluster 1241               MT0844   \n",
       "Cluster 4936                  SDH   \n",
       "Cluster 5551                 PHOU   \n",
       "Cluster 6988            AS9A_3300   \n",
       "Cluster 3233               MT1946   \n",
       "Cluster 1023               RV3692   \n",
       "Cluster 5365                 YWFI   \n",
       "Cluster 5904                 NRGA   \n",
       "Cluster 5565                    0   \n",
       "Cluster 7705               RV3316   \n",
       "Cluster 1010              RV3480C   \n",
       "Cluster 2554           SVIR_32990   \n",
       "Cluster 3614           SCLAV_0672   \n",
       "Cluster 4656  MKANA1_010100005130   \n",
       "Cluster 3149             MRA_1003   \n",
       "Cluster 4148               RV0604   \n",
       "Cluster 4455              RV2877C   \n",
       "Cluster 6925               RV0723   \n",
       "Cluster 2807               RV0331   \n",
       "Cluster 6872             MAB_2302   \n",
       "Cluster 3539             MRA_3237   \n",
       "Cluster 1742             NFA53720   \n",
       "Cluster 2062             MRA_2509   \n",
       "Cluster 4656  MKANA1_010100005130   \n",
       "Cluster 1053                USPA4   \n",
       "Cluster 5788              RV3167C   \n",
       "Cluster 4975             MRA_0668   \n",
       "Cluster 460             SROT_0714   \n",
       "Cluster 2192               ML1974   \n",
       "Cluster 1133              RV1326C   \n",
       "Cluster 1346              RV3469C   \n",
       "Cluster 2497               RV3170   \n",
       "Cluster 3410               RV0205   \n",
       "Cluster 4357               RV1596   \n",
       "Cluster 4791               RV0167   \n",
       "Cluster 5117              RV2927C   \n",
       "Cluster 2335               RV0777   \n",
       "\n",
       "                                                        product         cog  \\\n",
       "Cluster 1116  Catalase (EC 1.11.1.6) / Peroxidase (EC 1.11.1.7)           P   \n",
       "Cluster 1116  Catalase (EC 1.11.1.6) / Peroxidase (EC 1.11.1.7)           P   \n",
       "Cluster 2387          Cell division trigger factor (EC 5.2.1.8)           O   \n",
       "Cluster 1329  Respiratory nitrate reductase alpha chain (EC ...           C   \n",
       "Cluster 591   Serine/threonine-protein kinase PknD (EC 2.7.1...           T   \n",
       "Cluster 6719             Transcription termination protein NusB           K   \n",
       "Cluster 2133                           Chloride channel protein           P   \n",
       "Cluster 1527               Predicted membrane protein (DUF2319)           S   \n",
       "Cluster 4383  Putative CoA-transferase subunit alpha Rv3551/...           I   \n",
       "Cluster 464                Transcription-repair coupling factor           L   \n",
       "Cluster 1268                Prolyl endopeptidase (EC 3.4.21.26)           E   \n",
       "Cluster 4843               FIG215594: Membrane spanning protein           S   \n",
       "Cluster 1751                             Beta-carotene ketolase           Q   \n",
       "Cluster 1241  Cell envelope-associated transcriptional atten...           K   \n",
       "Cluster 4936                             Serine 3-dehydrogenase           L   \n",
       "Cluster 5551  Phosphate transport system regulatory protein ...           P   \n",
       "Cluster 6988  Bis(5'-nucleosyl)-tetraphosphatase (asymmetric...  ['F', 'G']   \n",
       "Cluster 3233  Threonine dehydrogenase and related Zn-depende...           C   \n",
       "Cluster 1023                          putative membrane protein           S   \n",
       "Cluster 5365               Coproheme decarboxylase HemQ (no EC)           S   \n",
       "Cluster 5904                   N-acetyltransferase, GCN5-family           K   \n",
       "Cluster 5565                                   Lipoprotein LppN           0   \n",
       "Cluster 7705   Succinate dehydrogenase cytochrome b-556 subunit           C   \n",
       "Cluster 1010  Wax ester synthase/acyl-CoA:diacylglycerol acy...           S   \n",
       "Cluster 2554             Putative xylulose kinase (EC 2.7.1.17)           G   \n",
       "Cluster 3614           Magnesium and cobalt efflux protein CorC           P   \n",
       "Cluster 4656                   Hydride transferase 1 (Fragment)           S   \n",
       "Cluster 3149  FIG037137: Putative conserved transmembrane pr...           S   \n",
       "Cluster 4148          protein of unknown function LppY and LpqO           S   \n",
       "Cluster 4455   Mercury resistance probable Hg transport protein           O   \n",
       "Cluster 6925                 LSU ribosomal protein L15p (L27Ae)           J   \n",
       "Cluster 2807                      Oxidoreductase (flavoprotein)           S   \n",
       "Cluster 6872                      membrane protein, MmpS family           S   \n",
       "Cluster 3539                          Potassium channel protein           P   \n",
       "Cluster 1742                                             ATPase           K   \n",
       "Cluster 2062      Diacyglycerol O-acyltransferase (EC 2.3.1.20)           S   \n",
       "Cluster 4656                   Hydride transferase 1 (Fragment)           S   \n",
       "Cluster 1053  Cationic amino acid transport integral membran...           T   \n",
       "Cluster 5788             Transcriptional regulator, AcrR family           K   \n",
       "Cluster 4975                          putative membrane protein           S   \n",
       "Cluster 460   Dihydrolipoamide succinyltransferase component...           C   \n",
       "Cluster 2192                                Amino acid permease           E   \n",
       "Cluster 1133  1,4-alpha-glucan (glycogen) branching enzyme, ...           G   \n",
       "Cluster 1346     4-hydroxy-2-oxovalerate aldolase (EC 4.1.3.39)           Q   \n",
       "Cluster 2497   Amine oxidase [flavin-containing] A (EC 1.4.3.4)           E   \n",
       "Cluster 3410     FIG01121868: Possible membrane protein, Rv0205           S   \n",
       "Cluster 4357  Quinolinate phosphoribosyltransferase [decarbo...           H   \n",
       "Cluster 4791  Conserved hypothetical integral membrane prote...           Q   \n",
       "Cluster 5117                    FIG00814129: Possible chaperone           D   \n",
       "Cluster 2335  Adenylosuccinate lyase (EC 4.3.2.2) @ SAICAR l...           F   \n",
       "\n",
       "              count                       ortho    pan      refseq   score  \n",
       "Cluster 1116   1592               83332.Rv1908c   Core  Z578_01418  1674.2  \n",
       "Cluster 1116   1592               83332.Rv1908c   Core  Z578_01418  1674.2  \n",
       "Cluster 2387   1595             419947.MRA_2488   Core  Z578_02233   974.6  \n",
       "Cluster 1329   1593          710685.MycrhN_6026   Core  Z578_00643   553.3  \n",
       "Cluster 591    1595             419947.MRA_0939   Core  AO93_01423  1471.6  \n",
       "Cluster 6719   1592             419947.MRA_2561   Core  AJ54_03676   283.6  \n",
       "Cluster 2133   1593            216594.MMAR_0353   Core  N126_01920   868.4  \n",
       "Cluster 1527   1594             419947.MRA_1079   Core        None  1350.9  \n",
       "Cluster 4383   1595             419947.MRA_3590   Core  Z578_00037   662.2  \n",
       "Cluster 464    1595             419947.MRA_1028   Core  Z578_02842  2721.8  \n",
       "Cluster 1268   1595             419947.MRA_0463   Core  Q625_02988  1557.4  \n",
       "Cluster 4843   1595             419947.MRA_0014   Core  BA33_02036   530.7  \n",
       "Cluster 1751   1595  557599.MkanA1_010100023024   Core        None   896.9  \n",
       "Cluster 1241   1595             419947.MRA_0832   Core  Z578_03954  1535.8  \n",
       "Cluster 4936   1595             419947.MRA_0491   Core  X103_00504   525.5  \n",
       "Cluster 5551   1595               83332.Rv3301c   Core  Z578_01269   470.5  \n",
       "Cluster 6988   1593            443218.AS9A_3300   Core  P456_00290   125.8  \n",
       "Cluster 3233   1591             419947.MRA_1907   Core  Z578_01405   841.0  \n",
       "Cluster 1023   1595  419947.MtubH3_010100023495   Core  AO54_00184   664.0  \n",
       "Cluster 5365   1595               83332.Rv2676c   Core  Z578_01557   525.5  \n",
       "Cluster 5904   1592             419947.MRA_0140   Core  Z578_00339   460.2  \n",
       "Cluster 5565   1594                           0   Core  T593_02377     0.0  \n",
       "Cluster 7705   1595               272631.ML0699   Core  Z578_01285   215.9  \n",
       "Cluster 1010   1595               83332.Rv3480c   Core  AO98_03593  1118.6  \n",
       "Cluster 2554   1595           471857.Svir_32990   Core  Z578_02092   145.6  \n",
       "Cluster 3614   1595             419947.MRA_1852   Core  AN87_01613   715.0  \n",
       "Cluster 4656   1594  557599.MkanA1_010100005130   Core  Z578_01971   538.5  \n",
       "Cluster 3149   1594             419947.MRA_1003   Core  AI72_02495   793.0  \n",
       "Cluster 4148   1592                83332.Rv0604   Core  Z578_02958   668.1  \n",
       "Cluster 4455   1595               83332.Rv2877c   Core  Z578_03348   577.1  \n",
       "Cluster 6925   1594             419947.MRA_0731   Core  Z578_02085   306.9  \n",
       "Cluster 2807   1595                83332.Rv0331   Core  AO48_03375   860.4  \n",
       "Cluster 6872    182             561007.MAB_2302  Acces  AN87_01324   134.4  \n",
       "Cluster 3539   1594             419947.MRA_3237   Core  X154_02944   736.1  \n",
       "Cluster 1742   1595             247156.nfa53720   Core  Z578_01702   276.0  \n",
       "Cluster 2062   1594             419947.MRA_2509   Core  AO71_03323  1078.6  \n",
       "Cluster 4656   1594  557599.MkanA1_010100005130   Core  Z578_01971   538.5  \n",
       "Cluster 1053   1595               83332.Rv2319c   Core  X399_02426   638.1  \n",
       "Cluster 5788   1595               83332.Rv3167c   Core  Z578_03871   430.4  \n",
       "Cluster 4975   1588             419947.MRA_0668  Acces        None   440.1  \n",
       "Cluster 460    1595            640132.Srot_0714   Core  AN41_01826  1906.6  \n",
       "Cluster 2192   1592               272631.ML1974   Core  Z550_00860   745.5  \n",
       "Cluster 1133   1595             419947.MRA_1334   Core  Z578_03418  1748.9  \n",
       "Cluster 1346   1595             419947.MRA_3573   Core  AL99_00361   756.0  \n",
       "Cluster 2497   1594            216594.MMAR_1389   Core  Z578_03874   830.5  \n",
       "Cluster 3410   1592                83332.Rv0205   Core  Z578_00414   742.3  \n",
       "Cluster 4357   1593             419947.MRA_1606   Core  P926_01348   596.3  \n",
       "Cluster 4791   1594                83332.Rv0167   Core  Z578_00374   548.6  \n",
       "Cluster 5117   1595             419947.MRA_2954   Core        None   506.5  \n",
       "Cluster 2335   1595             419947.MRA_0786   Core  Z578_03539  1011.4  "
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Chi2_excel_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Check if primary genes are in MI, ANOVA, or ANOVA F-test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "len([\"isoniazid\", \"ethambutol\", \"rifampicin\", \"pyrazinamide\", \"ethionamide\", \"ofloxacin\", \"4-aminosalicylic_acid\", \"streptomycin\", \"MDR\", \"XDR\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Load $\\fbox{GEMpro_seq_dataframes/}$ into single dataframe that integrates Nathan's pdb structures $\\fbox{nathan_structures_for_erol_45rows.csv}$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "seqDF_dict = {\"pyrazinamide\": (\"SVM\", \"SGD\"),\n",
    "             \"ethambutol\": (\"SVM\", \"SGD\"),\n",
    "             \"rifampicin\": (\"SVM\", \"SGD\"),\n",
    "             \"isoniazid\": (\"SVM\", \"SGD\"),\n",
    "             \"ethionamide\": (\"logReg\", \"GD\"), # 79% AUC\n",
    "             \"4-aminosalicylic_acid\": (\"logReg\", \"GD\"), # 80% AUC\n",
    "             \"ofloxacin\": (\"SVM\", \"SGD\"),\n",
    "             \"streptomycin\": (\"SVM\", \"SGD\"),\n",
    "             \"XDR\": (\"SVM\", \"SGD\"),\n",
    "             \"MDR\": (\"SVM\", \"SGD\"),\n",
    "             }\n",
    "\n",
    "nathan_pdb_structure_df = pd.read_csv(\"nathan_structures_for_erol_45rows.csv\", index_col=0)\n",
    "\n",
    "all_antibiotic_seqDF = pd.DataFrame()\n",
    "for antibiotic_id, (estimator_id, grad_descent_id) in seqDF_dict.iteritems():\n",
    "    \n",
    "    # seqDF_file_loc = \"GEMPro_seq_dataframes/\"+antibiotic_id+\"_\"+estimator_id+\"_\"+grad_descent_id+\"_SEQdf.csv\"\n",
    "    seqDF_file_loc = \"GEMPro_seq_dataframes/\"+antibiotic_id+\"_\"+estimator_id+\"_\"+grad_descent_id+\"_60_SEQdf.csv\"\n",
    "    print seqDF_file_loc\n",
    "    \n",
    "    antibiotic_seq_df = pd.read_csv(seqDF_file_loc, index_col=0)\n",
    "    antibiotic_seq_df[\"antibiotic\"] = antibiotic_id\n",
    "    \n",
    "    all_antibiotic_seqDF = pd.concat([all_antibiotic_seqDF, antibiotic_seq_df])\n",
    "    \n",
    "\n",
    "all_antibiotic_seqDF[\"id\"] = all_antibiotic_seqDF[\"gene_id\"].apply(lambda x: nathan_pdb_structure_df.loc[x,\"id\"] if x in nathan_pdb_structure_df.index else \"No PDB\")\n",
    "all_antibiotic_seqDF[\"is_experimental\"] = all_antibiotic_seqDF[\"gene_id\"].apply(lambda x: nathan_pdb_structure_df.loc[x,\"is_experimental\"] if x in nathan_pdb_structure_df.index else \"No PDB\")\n",
    "all_antibiotic_seqDF[\"file_type\"] = all_antibiotic_seqDF[\"gene_id\"].apply(lambda x: nathan_pdb_structure_df.loc[x,\"file_type\"] if x in nathan_pdb_structure_df.index else \"No PDB\")\n",
    "all_antibiotic_seqDF[\"structure_file\"] = all_antibiotic_seqDF[\"gene_id\"].apply(lambda x: nathan_pdb_structure_df.loc[x,\"structure_file\"] if x in nathan_pdb_structure_df.index else \"No PDB\")\n",
    "all_antibiotic_seqDF.columns = [\"gene_id\", \"name\", \"sequences\", \"uniprot_id\", \"antibiotic\", \"id\", \"is_experimental\", \"file_type\", \"structure_file\"]\n",
    "all_antibiotic_seqDF.to_csv(\"all_antibiotic_PDB_seqDF_XDR_MDR.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# ---- seq dataframes to single dataframe -----"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Build dataframe for eddy\n",
    "- map cluster gene to alleles to either being \"R\" or \"S\" - we should go in manually and make sure that this is the final solution for this problem.\n",
    "- If LOR is gereater than 3, then we should define it to be resistant.\n",
    "- elif LOR is less than -3, then we should define it to be susceptible."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def allele_cooccurence_oddsRatio(all_1, all_2, y_drg, X_all, add_small_value_to_table, log_or_not, single_or_double_tuple):\n",
    "    \"\"\"\n",
    "    Takes in two alleles, creates a cooccurence column, and computes the association of that\n",
    "    column with the AMR phenotype.\n",
    "    \"\"\"\n",
    "    # print X_all.shape, all_1, all_2, X_all.columns\n",
    "    two_allele_df = X_all.loc[:, [all_1,all_2]].copy()\n",
    "    \n",
    "    if single_or_double_tuple[0] == \"double\":\n",
    "        two_allele_df[\"cooccurence\"] = two_allele_df[all_1] + two_allele_df[all_2]\n",
    "        two_allele_df[\"cooccurence\"][two_allele_df[\"cooccurence\"]<2]=0\n",
    "        two_allele_df[\"cooccurence\"][two_allele_df[\"cooccurence\"]==2]=1\n",
    "    elif single_or_double_tuple[0] == \"single\" and all_1 == all_2:\n",
    "        two_allele_df = X_all.loc[:, [all_2]].copy()\n",
    "        two_allele_df[\"cooccurence\"] = two_allele_df[single_or_double_tuple[1]]\n",
    "    elif single_or_double_tuple[0] == \"single\":\n",
    "        two_allele_df[\"cooccurence\"] = two_allele_df[single_or_double_tuple[1]]\n",
    "    \n",
    "    two_allele_df[\"resistant\"] = y_drg\n",
    "    \n",
    "    cooccurence_resistant = float(len(two_allele_df[(two_allele_df[\"cooccurence\"]==1) & (two_allele_df[\"resistant\"]==1)].index))\n",
    "    cooccurence_susceptible = float(len(two_allele_df[(two_allele_df[\"cooccurence\"]==1) & (two_allele_df[\"resistant\"]==0)].index))\n",
    "    \n",
    "    no_cooccurence_resistant = float(len(two_allele_df[(two_allele_df[\"cooccurence\"]==0) & (two_allele_df[\"resistant\"]==1)].index))\n",
    "    no_cooccurence_susceptible = float(len(two_allele_df[(two_allele_df[\"cooccurence\"]==0) & (two_allele_df[\"resistant\"]==0)].index))\n",
    "    \n",
    "    # add_small_value_to_table = .1\n",
    "    # if one of the cells is 0, there won't be an odds ratio. To get past this, we add a small value of .5 to each value\n",
    "    if cooccurence_resistant==0 or cooccurence_susceptible==0 or no_cooccurence_resistant==0 or no_cooccurence_susceptible==0:\n",
    "        cooccurence_resistant+=add_small_value_to_table\n",
    "        cooccurence_susceptible+=add_small_value_to_table\n",
    "        no_cooccurence_resistant+=add_small_value_to_table\n",
    "        no_cooccurence_susceptible+=add_small_value_to_table\n",
    "    \n",
    "    odds_ratio = (cooccurence_resistant/cooccurence_susceptible)/(no_cooccurence_resistant/no_cooccurence_susceptible)\n",
    "    \n",
    "    if log_or_not == True:\n",
    "        odds_ratio = np.log(odds_ratio)\n",
    "    \n",
    "    return odds_ratio"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "seqDF_dict = {\n",
    "#     \"pyrazinamide\": (\"SVM\", \"SGD\"),\n",
    "#              \"ethambutol\": (\"SVM\", \"SGD\"),\n",
    "#              \"rifampicin\": (\"SVM\", \"SGD\"),\n",
    "             \"isoniazid\": (\"SVM\", \"SGD\"),\n",
    "#              \"ethionamide\": (\"SVM\", \"SGD\"), # 79% AUC\n",
    "#              \"4-aminosalicylic_acid\": (\"SVM\", \"SGD\"), # 80% AUC\n",
    "#              \"ofloxacin\": (\"SVM\", \"SGD\"),\n",
    "#              \"streptomycin\": (\"SVM\", \"SGD\"),\n",
    "#              \"XDR\": (\"SVM\", \"SGD\"),\n",
    "#              \"MDR\": (\"SVM\", \"SGD\"),\n",
    "             }\n",
    "\n",
    "all_antibiotic_seqDF = pd.DataFrame()\n",
    "\n",
    "for antibiotic_id, (estimator_id, grad_descent_id) in seqDF_dict.iteritems():\n",
    "    \n",
    "    # seqDF_file_loc = \"GEMPro_seq_dataframes/\"+antibiotic_id+\"_\"+estimator_id+\"_\"+grad_descent_id+\"_SEQdf.csv\"\n",
    "    seqDF_file_loc = \"GEMPro_seq_dataframes/\"+antibiotic_id+\"_\"+estimator_id+\"_\"+grad_descent_id+\"_60_SEQdf.csv\"\n",
    "    print seqDF_file_loc\n",
    "    \n",
    "    antibiotic_seq_df = pd.read_csv(seqDF_file_loc, index_col=0)\n",
    "    antibiotic_seq_df[\"antibiotic\"] = antibiotic_id\n",
    "    \n",
    "    all_antibiotic_seqDF = pd.concat([all_antibiotic_seqDF, antibiotic_seq_df])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "all_antibiotic_seqDF[all_antibiotic_seqDF[\"gene_id\"]==\"Rv0118c\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "all_antibiotic_seqDF.loc[\"mut_Cluster1116_2\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "for index, row in all_antibiotic_seqDF.iterrows():\n",
    "    if \"Rv1908c\" in row[\"gene_id\"] or \"Rv1908c\" in :\n",
    "        print index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "estimator_type = \"SVM\"\n",
    "SGD_or_GD = \"SGD\"\n",
    "\n",
    "All_or_OutDf = \"All\"\n",
    "LOR_abs_cutoff = 1.0\n",
    "\n",
    "allele_resist_df = pd.DataFrame()\n",
    "\n",
    "# for drug_name in [\"isoniazid\", \"ethambutol\", \"rifampicin\", \"pyrazinamide\"]:\n",
    "\n",
    "for drug_name in [\"isoniazid\", \"ethambutol\", \"rifampicin\", \"pyrazinamide\", \"ethionamide\", \"ofloxacin\", \"4-aminosalicylic_acid\",\"streptomycin\", \"MDR\", \"XDR\"]: # , \n",
    "\n",
    "    drug_top100_allele_df = pd.read_csv(\n",
    "        \"SVM_dataframes/\"+drug_name+\"_\"+estimator_type+\"_\"+SGD_or_GD+\"_top_100alleles.csv\", \n",
    "                                        index_col=0)\n",
    "    \n",
    "    \n",
    "    # Get antibiotic specific dataframe for the calculation of odds ratio.\n",
    "    if drug_name == \"MDR\" or drug_name == \"XDR\":\n",
    "        X1, y1 = get_MDR_training(resistance_data, data_to_plot, drug_name)\n",
    "    else:\n",
    "        y1, X1 = get_target_data(data_to_plot, resistance_data, drug_name)\n",
    "        X1 = remove_key_antibiotic_clusters(X1, drug_name)\n",
    "\n",
    "    uniq_clusters = list(set([str(x).split(\"_\")[0] for x in drug_top100_allele_df.iloc[:50, :].index]))\n",
    "\n",
    "    clst_allele_to_LOR_AMR_dict = {}\n",
    "    \n",
    "    for uniq_clst in uniq_clusters:\n",
    "        \n",
    "        uniq_clst_dash = uniq_clst+\"_\"\n",
    "        clst_alleles = [x for x in list(set(clust_variant_df.columns)) if uniq_clst_dash in x]\n",
    "        for uniq_allele in clst_alleles:\n",
    "            # print uniq_allele\n",
    "            allele_LOR = allele_cooccurence_oddsRatio(uniq_allele, uniq_allele, y1, X1, 0.1, True, (\"single\", uniq_allele))\n",
    "\n",
    "            if allele_LOR>LOR_abs_cutoff:\n",
    "                AMR_pheno = \"R\"\n",
    "            elif allele_LOR<-LOR_abs_cutoff:\n",
    "                AMR_pheno = \"S\"\n",
    "            else:\n",
    "                AMR_pheno = \"Cantsaydawg\"\n",
    "\n",
    "            product_name = test_df_final.loc[uniq_clst.replace(\"ter\", \"ter \"), \"clust_to_rv\"].replace(\"V\", \"v\").replace(\"C\", \"c\")\n",
    "            \n",
    "            if \"mut_\"+uniq_allele in all_antibiotic_seqDF.index:\n",
    "                allele_loc_name = \"mut_\"+uniq_allele\n",
    "            elif \"refseq_\"+uniq_allele in all_antibiotic_seqDF.index:\n",
    "                allele_loc_name = \"refseq_\"+uniq_allele\n",
    "            else:\n",
    "                allele_loc_name = \"no_seq\"\n",
    "                print uniq_allele, \"-> no seq?\"\n",
    "                \n",
    "            if allele_loc_name != \"no_seq\":\n",
    "                rv_gene_id = all_antibiotic_seqDF.loc[allele_loc_name][\"gene_id\"]\n",
    "                gene_name = all_antibiotic_seqDF.loc[allele_loc_name][\"name\"]\n",
    "                gene_sequence = all_antibiotic_seqDF.loc[allele_loc_name][\"sequences\"]\n",
    "                uniprot_id = all_antibiotic_seqDF.loc[allele_loc_name][\"uniprot_id\"]\n",
    "                \n",
    "                \n",
    "#             try:\n",
    "#                 rv_gene_id = all_antibiotic_seqDF.loc[\"mut_\"+uniq_allele][\"gene_id\"]\n",
    "#                 gene_name = all_antibiotic_seqDF.loc[\"mut_\"+uniq_allele][\"name\"]\n",
    "#                 gene_sequence = all_antibiotic_seqDF.loc[\"mut_\"+uniq_allele][\"sequences\"]\n",
    "#                 uniprot_id = all_antibiotic_seqDF.loc[\"mut_\"+uniq_allele][\"uniprot_id\"]\n",
    "#             except:\n",
    "#                 rv_gene_id = all_antibiotic_seqDF.loc[\"refseq_\"+uniq_allele][\"gene_id\"]\n",
    "#                 gene_name = all_antibiotic_seqDF.loc[\"refseq_\"+uniq_allele][\"name\"]\n",
    "#                 gene_sequence = all_antibiotic_seqDF.loc[\"refseq_\"+uniq_allele][\"sequences\"]\n",
    "#                 uniprot_id = all_antibiotic_seqDF.loc[\"refseq_\"+uniq_allele][\"uniprot_id\"]\n",
    "                \n",
    "                clst_allele_to_LOR_AMR_dict.update({uniq_allele: {\"allele_LOR\": allele_LOR, \"AMR\":AMR_pheno, \n",
    "                                                                  \"antibiotic\": drug_name,\"gene_product\": product_name,\n",
    "                                                                  \"rv_gene_id\": rv_gene_id, \"gene_name\": gene_name,\n",
    "                                                                 \"gene_sequence\": gene_sequence, \"uniprot_id\": uniprot_id}})\n",
    "                clst_allele_to_LOR_AMR_df = pd.DataFrame.from_dict(clst_allele_to_LOR_AMR_dict, orient=\"index\")\n",
    "            \n",
    "    allele_resist_df = allele_resist_df.append(clst_allele_to_LOR_AMR_df)\n",
    "\n",
    "# allele_resist_df.to_csv(\"allele_resist_seqInfo_\"+str(LOR_abs_cutoff).replace(\".\",\"point\")+\"_df.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "test_df_final.loc[\"Cluster 1116\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "allele_resist_df.loc[\"Cluster1116_1\", \"gene_sequence\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "allele_resist_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "len(all_antibiotic_seqDF.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "uniq_allele"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "try:\n",
    "    rv_gene_id = all_antibiotic_seqDF.loc[\"mut_\"+\"Cluster1153_10\"][\"gene_id\"]\n",
    "except:\n",
    "    rv_gene_id = all_antibiotic_seqDF.loc[\"ref_\"+\"Cluster1153_10\"][\"gene_id\"]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "all_antibiotic_seqDF.loc[\"mut_\"+\"Cluster1153_10\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "data_to_plot.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "allele_resist_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "allele_resist_df.to_csv(\"allele_resist_df.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "allele_resist_df[(allele_resist_df[\"antibiotic\"]==\"ethambutol\")&(allele_resist_df[\"AMR\"]==\"R\")]\n",
    "# GEMPro_final = get_GEMPRO_df(general_cluster_alleles, clust_to_rv, rv_to_name, clster_manual_map, clust_variant_df)\n",
    "# GEMPro_final.to_csv(\"GEMPro_seq_dataframes/\"+drug_name+\"_\"+estimator_type+\"_\"+SGD_or_GD+\"_\"+str(top_num)+\"_SEQdf.csv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "test_df_final.loc[\"Cluster 3289\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "clst_allele_to_LOR_AMR_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### TO DO\n",
    "1. Talk to eddy about the attached excel sheets\n",
    "2. Talk to Nathan about getting homology structures for the following structures\n",
    "3. Figure out how to write code that loops through the following structure files and takes out features that most likely relate to a causal mechanism such as embR where there is a deletion that likely reduces the binding affinity to DNA. Otherwise, ...."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 8. Look at top correlations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "top_allele_pairs = 40\n",
    "\n",
    "abs_corr_df = sim_corr_df.abs()\n",
    "os = (abs_corr_df.where(np.triu(np.ones(abs_corr_df.shape), k=1).astype(np.bool))\n",
    "                 .stack()\n",
    "                 .sort_values(ascending=False))\n",
    "\n",
    "for (allele_1, allele_2) in os.index[:top_allele_pairs]:\n",
    "    try:\n",
    "        clst_1,prod_1, clst_2,prod_2 = allele_1.split(\"Rv\")[0],allele_1.split(\", \")[1], allele_2.split(\"Rv\")[0],allele_2.split(\", \")[1]\n",
    "        print prod_1, prod_2\n",
    "        print (clst_1, clst_2), os.loc[(allele_1, allele_2)]\n",
    "        print \n",
    "    except:\n",
    "        print \"NO GOOD: \", allele_1, allele_2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 9. Create pairwise allele tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pairwise_table_groups = {\n",
    "#     \"ethambutol\": [(\"Cluster551_\", \"Cluster4244_\"), # embB / ubiA\n",
    "#                                         (\"Cluster551_\", \"Cluster3176_\"), # embB / embR\n",
    "#                                         (\"Cluster4244_\", \"Cluster4348_\"), # ubiA / proC\n",
    "#                                         (\"Cluster4244_\", \"Cluster3176_\"), # ubiA / embR\n",
    "#                                         (\"Cluster3176_\", \"Cluster4244_\"),\n",
    "#                                         (\"Cluster551_\", \"Cluster2496_\"),\n",
    "#                                         (\"Cluster551_\", \"Cluster2954_\")\n",
    "#                                        ],\n",
    "                         \"isoniazid\": [(\"Cluster1116_\", \"Cluster4738_\"), # katG / inhA\n",
    "                                        (\"Cluster1116_\", \"Cluster4498_\"), # katG / lpqR\n",
    "                                        (\"Cluster1116_\", \"Cluster1601_\"), # katG / oxcA\n",
    "                                        (\"Cluster1116_\", \"Cluster2178_\"), # katG / pbpA\n",
    "                                       \n",
    "                                       (\"Cluster4738_\",\"Cluster1116_\"), # inhA / katG \n",
    "                                        (\"Cluster4498_\", \"Cluster1116_\"), # lpqR / katG \n",
    "                                        (\"Cluster1601_\", \"Cluster1116_\"), # oxcA / katG\n",
    "                                        (\"Cluster2178_\", \"Cluster1116_\"), # pbpA / katG \n",
    "                                       (\"Cluster1237_\", \"Cluster1116_\"), # embR / katG \n",
    "                                       (\"Cluster1258_\", \"Cluster1116_\"), # priA / katG \n",
    "                                       (\"Cluster1258_\", \"Cluster1116_\"), # priA / katG \n",
    "                                       \n",
    "                                       (\"Cluster4738_\", \"Cluster4498_\"), # inhA / lpqR\n",
    "                                       (\"Cluster4738_\", \"Cluster1601_\"), # inhA / oxcA\n",
    "                                       (\"Cluster4738_\", \"Cluster2178_\"), # inhA / pbpA\n",
    "                                       (\"Cluster203_\", \"Cluster1116_\"), # inhA / pbpA\n",
    "                                       ],\n",
    "#                         \"streptomycin\": [(\"Cluster7435_\", \"Cluster2032_\"),\n",
    "#                                         (\"Cluster7435_\", \"Cluster5240_\"),\n",
    "#                                         (\"Cluster7435_\", \"Cluster4056_\"),\n",
    "#                                          (\"Cluster4961_\", \"Cluster5263_\")\n",
    "#                                        ],\n",
    "#                         \"ethionamide\": [(\"Cluster2122_\", \"Cluster5664_\"),\n",
    "#                                         (\"Cluster2122_\", \"Cluster3373_\"),\n",
    "#                                         (\"Cluster2122_\", \"Cluster6929_\"),\n",
    "#                                          (\"Cluster6929_\", \"Cluster1916_\")\n",
    "#                                        ],\n",
    "#                         \"4-aminosalicylic_acid\": [\n",
    "#                             (\"Cluster2211_\", \"Cluster1916_\"), # folC, serA1\n",
    "#                             (\"Cluster2211_\", \"Cluster459_\"), # folC, Rv0194 \"multi-drug ATP..\"\n",
    "#                             (\"Cluster2211_\", \"Cluster4486_\"), # folC, thyA\n",
    "#                             (\"Cluster4486_\", \"Cluster594_\"), # thyA, Rv0194 \"multi-drug ATP..\"\n",
    "#                             (\"Cluster1916_\", \"Cluster594_\"), # serA1, Rv0194 \"multi-drug ATP..\"\n",
    "#                             (\"Cluster3005_\", \"Cluster1031_\"), # cysK2, purL\n",
    "#                             (\"Cluster3005_\", \"Cluster2211_\") # cysK2, folC\n",
    "#                                                  ],\n",
    "#                         \"rifampicin\": [(\"Cluster491_\", \"Cluster415_\"),\n",
    "#                                        (\"Cluster491_\", \"Cluster1206_\"),\n",
    "#                                        (\"Cluster491_\", \"Cluster3849_\"),\n",
    "#                                        (\"Cluster1206_\", \"Cluster3849_\"),\n",
    "#                                       ],\n",
    "#                         \"pyrazinamide\": [(\"Cluster3930_\", \"Cluster3713_\"),\n",
    "#                                        (\"Cluster3930_\", \"Cluster3289_\"), # pncA, Rv2560 # high correlation!\n",
    "#                                        (\"Cluster3930_\", \"Cluster5659_\"), # pncA, hbhA\n",
    "#                                        (\"Cluster5659_\", \"Cluster3930_\"), # hbhA, pncA\n",
    "#                                        (\"Cluster3930_\", \"Cluster2336_\"), # pncA, pykA\n",
    "#                                        (\"Cluster521_\", \"Cluster3930_\"), # Rv1358, pncA\n",
    "#                                         (\"Cluster5659_\", \"Cluster3930_\"), # Rv1358, pncA\n",
    "#                                         (\"Cluster3348_\", \"Cluster3930_\"), # korB (Rv2454c), pncA\n",
    "#                                         (\"Cluster3289_\", \"Cluster3930_\"), # Rv2560, pncA\n",
    "#                                         (\"Cluster3713_\", \"Cluster3930_\"), # Rv2715 (hydrolase), pykA\n",
    "#                                          (\"Cluster1613_\", \"Cluster3930_\"), # pncA v2, pncA\n",
    "#                                          (\"Cluster1613_\", \"Cluster3289_\"),\n",
    "#                                          (\"Cluster1613_\", \"Cluster3713_\"),\n",
    "#                                          (\"Cluster521_\",\"Cluster1613_\"), # Rv1359, Rv2042c\n",
    "#                                       ],\n",
    "                        \"XDR\": [(\"Cluster893_\", \"Cluster1613_\"), # gyrA, nicotinamidase\n",
    "                                (\"Cluster1613_\", \"Cluster6316_\"), # nicotinamidase, mannose-6-phosphate isomerase\n",
    "                                (\"Cluster1613_\", \"Cluster6316_\"),\n",
    "                                \n",
    "                                      ],\n",
    "#                             \"ofloxacin\": [(\"Cluster893_\", \"Cluster2980_\"), # gyrA, ald\n",
    "#                                         (\"Cluster893_\", \"Cluster1645_\"), # gyrA, cydC\n",
    "#                                           (\"Cluster893_\", \"Cluster2212_\"), # gyrA, ansP2\n",
    "#                                           (\"Cluster2980_\", \"Cluster893_\"), # ald, gyrA\n",
    "#                                         (\"Cluster1645_\", \"Cluster893_\"), # cydC, gyrA\n",
    "#                                           (\"Cluster2212_\", \"Cluster893_\"), # ansP2, gyrA\n",
    "#                                           (\"Cluster2227_\", \"Cluster893_\") #phoR, gyrA\n",
    "#                                       ],\n",
    "                        }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "test_df_final.loc[\"Cluster 3005\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def make_table_figure(clust1, clust2, drg_name, two_cluster_df, two_clusterPercent_df, two_varPval_df, two_varOdds_df, clust_to_rv, rv_to_name, clster_manual_map, which_df_2_color, color_bar_or_not):\n",
    "    \"\"\"\n",
    "    Automated process for building tables for the allele co-occurences.\n",
    "    \"\"\"\n",
    "    rv_name1, rv_name2 = clust_to_rv[clust1.strip(\"_\").replace(\"ter\", \"ter \")], clust_to_rv[clust2.strip(\"_\").replace(\"ter\", \"ter \")]\n",
    "    name1, name2 = rv_to_name[rv_name1], rv_to_name[rv_name2]\n",
    "\n",
    "    if name1 == None and rv_name1 in clster_manual_map.keys():\n",
    "        name1 = clster_manual_map[rv_name1]\n",
    "    elif name1 == None:\n",
    "        name1 = rv_name1\n",
    "\n",
    "    if name2 == None and rv_name2 in clster_manual_map.keys():\n",
    "        name2 = clster_manual_map[rv_name2]\n",
    "    elif name2 == None:\n",
    "        name2 = rv_name2\n",
    "        \n",
    "    # ------ Make styled dataframe by changing the index and column names, moving shit around, etc.... ------\n",
    "    if which_df_2_color == \"percent\":\n",
    "        save_occurance_df = two_clusterPercent_df.copy()\n",
    "        vmin_bot=0.2\n",
    "        vmax_top=0.95\n",
    "    elif which_df_2_color == \"pVal\":\n",
    "        save_occurance_df = two_varPval_df.copy()\n",
    "        save_occurance_df = -np.log(save_occurance_df).replace([np.inf, -np.inf], np.nan)\n",
    "        vmin_bot=0\n",
    "        vmax_top=30\n",
    "    elif which_df_2_color == \"oddsRatio\":\n",
    "        save_occurance_df = two_varOdds_df.copy()\n",
    "        abs_max_bound = max(abs(save_occurance_df.fillna(0).values.min()), save_occurance_df.fillna(0).values.max())\n",
    "        vmin_bot = -abs_max_bound\n",
    "        vmax_top = abs_max_bound\n",
    "\n",
    "    new_cols = [str(x.split(\"_\")[1]) if len(str(x.split(\"_\")[1]))>0 else x for x in save_occurance_df.columns]\n",
    "    new_inds = [str(x.split(\"_\")[1]) if len(str(x.split(\"_\")[1]))>0 else x for x in save_occurance_df.index]\n",
    "\n",
    "    new_cols_Name = [str(x.split(\"_\")[1]) if len(str(x.split(\"_\")[1]))>0 else rv_name2 for x in save_occurance_df.columns]\n",
    "    new_inds_Name = [str(x.split(\"_\")[1]) if len(str(x.split(\"_\")[1]))>0 else rv_name1 for x in save_occurance_df.index]\n",
    "\n",
    "    allele_cols = [x for x in new_cols if \"Cluster\" not in x]\n",
    "    # cluster_col = [x for x in new_cols if \"Cluster\" in x]\n",
    "    cluster_col = [rv_name2]\n",
    "\n",
    "    allele_ind = [x for x in new_inds if \"Cluster\" not in x]\n",
    "    # cluster_ind = [x for x in new_inds if \"Cluster\" in x]\n",
    "    cluster_ind = [rv_name1]\n",
    "\n",
    "    # save_occurance_df.columns = new_cols\n",
    "    # save_occurance_df.index = new_inds\n",
    "    save_occurance_df.columns = new_cols_Name\n",
    "    save_occurance_df.index = new_inds_Name\n",
    "\n",
    "    save_occurance_df = save_occurance_df[allele_cols+cluster_col]\n",
    "    save_occurance_df = save_occurance_df.reindex(index = allele_ind+cluster_ind)\n",
    "\n",
    "    labeled_two_cluster_df = two_cluster_df.copy()\n",
    "    # labeled_two_cluster_df.columns = new_cols\n",
    "    # labeled_two_cluster_df.index = new_inds\n",
    "    labeled_two_cluster_df.columns = new_cols_Name\n",
    "    labeled_two_cluster_df.index = new_inds_Name\n",
    "\n",
    "    labeled_two_cluster_df = labeled_two_cluster_df[allele_cols+cluster_col]\n",
    "    labeled_two_cluster_df = labeled_two_cluster_df.reindex(index = allele_ind+cluster_ind)\n",
    "\n",
    "    # ---- Add the two columns and rows to both ends of \"save_occurance_df\"and \"labeled_two_cluster_df\"\n",
    "    # --- Add the two columns to right side of dataframe ----\n",
    "    \n",
    "    resist_vector_name = \"#R\"\n",
    "    total_vector_name = \"Total\"\n",
    "    \n",
    "    labeled_two_cluster_df[resist_vector_name] = labeled_two_cluster_df.fillna(\"-\")[cluster_col[0]].apply(lambda x: x.split(\":\")[0].strip(\"()\").split(\"/\")[0] if \"/\" in x else x)\n",
    "    labeled_two_cluster_df[total_vector_name] = labeled_two_cluster_df.fillna(\"-\")[cluster_col[0]].apply(lambda x: x.split(\":\")[0].strip(\"()\").split(\"/\")[1] if \"/\" in x else x)\n",
    "    \n",
    "    # --- Add the two rows to the bottom of dataframe ----\n",
    "    row_resistant = labeled_two_cluster_df.fillna(\"-\").loc[cluster_ind[0], :].apply(lambda x: x.split(\":\")[0].strip(\"()\").split(\"/\")[0] if \"/\" in x else x)\n",
    "    row_resistant.name = resist_vector_name\n",
    "    row_total = labeled_two_cluster_df.fillna(\"-\").loc[cluster_ind[0], :].apply(lambda x: x.split(\":\")[0].strip(\"()\").split(\"/\")[1] if \"/\" in x else x)\n",
    "    row_total.name = total_vector_name\n",
    "    labeled_two_cluster_df = labeled_two_cluster_df.append(row_resistant)\n",
    "    labeled_two_cluster_df = labeled_two_cluster_df.append(row_total)\n",
    "\n",
    "    # --- Drop Gene ID row ---\n",
    "    labeled_two_cluster_df.drop(cluster_col, axis=1, inplace=True)\n",
    "    labeled_two_cluster_df.drop(cluster_ind, axis=0, inplace=True)\n",
    "    \n",
    "    # ---- Change save_occurance_df -----\n",
    "    save_occurance_df[resist_vector_name] = save_occurance_df[cluster_col[0]]\n",
    "    save_occurance_df[total_vector_name] = save_occurance_df[cluster_col[0]]\n",
    "    row_occurTotal_add = save_occurance_df.loc[cluster_ind[0], :].copy()\n",
    "    row_occurTotal_add.name = total_vector_name\n",
    "    row_occurR_add = save_occurance_df.loc[cluster_ind[0], :].copy()\n",
    "    row_occurR_add.name = resist_vector_name\n",
    "    save_occurance_df = save_occurance_df.append(row_occurR_add)\n",
    "    save_occurance_df = save_occurance_df.append(row_occurTotal_add)\n",
    "    save_occurance_df.drop([cluster_col[0]], axis=1, inplace=True)\n",
    "    save_occurance_df.drop([cluster_ind[0]], inplace=True)\n",
    "\n",
    "    # ------ Make seaborn heatmap from the two dataframes that looks like a table ------\n",
    "    # cm = sns.diverging_palette(220, 10, sep=80, n=7, l=60, as_cmap=True)\n",
    "    # cm = sns.diverging_palette(220, 10, sep=1, n=7, as_cmap=True)\n",
    "    # cm = sns.light_palette(\"red\",as_cmap=True, n_colors=5)\n",
    "    # cm = sns.choose_light_palette(input='husl', as_cmap=True)\n",
    "    cm = sns.diverging_palette(236, 15, s=86, l=65, sep=10, n=9, as_cmap=True)\n",
    "    \n",
    "    # ------ reorder the columns and index based on the numeric value instead of string ------\n",
    "    ordered_cols = [int(x) if x.isdigit() else x for x in save_occurance_df.columns]\n",
    "    ordered_cols.sort()\n",
    "    ordered_cols_string = [str(x) for x in ordered_cols]\n",
    "    ordered_inds = [int(x) if x.isdigit() else x for x in save_occurance_df.index]\n",
    "    ordered_inds.sort()\n",
    "    ordered_inds_string = [str(x) for x in ordered_inds]\n",
    "\n",
    "    save_occurance_df = save_occurance_df[ordered_cols_string]\n",
    "    save_occurance_df.reindex(ordered_inds_string)\n",
    "    \n",
    "    labeled_two_cluster_df = labeled_two_cluster_df[ordered_cols_string]\n",
    "    labeled_two_cluster_df.reindex(ordered_inds_string)\n",
    "    \n",
    "    labels = np.array(labeled_two_cluster_df.fillna(\"-\").astype(str).copy())\n",
    "    only_frac_labels = []\n",
    "    for x in labels:\n",
    "        row_labels = [y.split(\":\")[0].strip(\"()\") for y in x]\n",
    "        row_labels = ['$\\\\dfrac{{{0}}}{{{1}}}$'.format(y.split(\"/\")[0], y.split(\"/\")[1]) if \"/\" in y else y for y in row_labels ]\n",
    "        only_frac_labels.append(row_labels)\n",
    "\n",
    "    # plt.figure(figsize=(len(allele_ind)+1,len(allele_cols)+1))\n",
    "    # plt.tight_layout()\n",
    "    # rowCol_ratio_in_figure = len(save_occurance_df.index)/float(len(save_occurance_df.columns))\n",
    "    if color_bar_or_not == True:\n",
    "        plt.figure(figsize=(len(save_occurance_df.columns)*.65, len(save_occurance_df.index)*.5))\n",
    "    else:\n",
    "        plt.figure(figsize=(len(save_occurance_df.columns)*.5, len(save_occurance_df.index)*.5))\n",
    "    g = sns.heatmap(save_occurance_df, annot=np.array(only_frac_labels),fmt = '',\n",
    "                    cmap=cm, vmin=vmin_bot, vmax=vmax_top,\n",
    "                    cbar=True, linewidths=0.6, linecolor=\"black\", \n",
    "                    annot_kws={\"color\":\"black\"}, \n",
    "                    # annot_kws={\"size\": rowCol_ratio_in_figure*3, \"color\":\"black\"},\n",
    "                    square=True);\n",
    "                    # \n",
    "    g.set_xlabel(name1+\" alleles\", rotation=0); # , size=16\n",
    "    g.set_ylabel(name2+\"\\nalleles\", rotation=0); # , size=16\n",
    "    g.xaxis.tick_top()\n",
    "    g.xaxis.set_label_position('top') \n",
    "    g.tick_params(axis=\"both\", which=\"major\", length=0)\n",
    "    plt.setp(g.yaxis.get_majorticklabels(), rotation=0); # , size=10\n",
    "    plt.setp(g.xaxis.get_majorticklabels(), rotation=0); # , size=10\n",
    "    plt.savefig(\"cooccurance_table_figures/\"+drg_name+\"_\"+name1+\"__\"+name2+\"_\"+which_df_2_color+\"_occurance.svg\")\n",
    "    plt.close()\n",
    "    return save_occurance_df, labeled_two_cluster_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sns.palplot(sns.diverging_palette(236, 15, s=86, l=65, sep=10, n=9))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pairwise_table_groups = {\n",
    "#     \"isoniazid\": [(\"Cluster1116_\", \"Cluster3770_\"), # katG, gpsA\n",
    "#                   (\"Cluster2698_\",\"Cluster1116_\"), # aspB, katG\n",
    "#                  (\"Cluster3438_\",\"Cluster1116_\"), # adhE1, katG\n",
    "#                  (\"Cluster3205_\",\"Cluster1116_\"),\n",
    "#                  (\"Cluster3438_\",\"Cluster1116_\")], \n",
    "    \n",
    "#     \"ofloxacin\": [(\"Cluster3438_\", \"Cluster893_\")], # gyrA, kdpC\n",
    "#     \"streptomycin\": [(\"Cluster7435_\", \"Cluster3770_\"),\n",
    "#                     (\"Cluster4056_\", \"Cluster7435_\")], # katG, gpsA\n",
    "    \n",
    "#     \"pyrazinamide\": [(\"Cluster3930_\", \"Cluster6096_\"),# pncA, kdpC\n",
    "#                     (\"Cluster3889_\", \"Cluster3930_\")], # ompA, pncA\n",
    "    \n",
    "#     \"ethambutol\": [(\"Cluster551_\", \"Cluster6096_\"), # embB, kdpC\n",
    "#                   (\"Cluster3889_\", \"Cluster551_\"),# ompA,embB\n",
    "#                   (\"Cluster3889_\", \"Cluster4244_\"), # ompA, ubiA\n",
    "#                   (\"Cluster551_\", \"Cluster3005_\")], # embB, cysK2\n",
    "    \n",
    "#     \"XDR\": [(\"Cluster893_\", \"Cluster6096_\"),\n",
    "#             (\"Cluster7926_\", \"Cluster893_\"), # mazF3, gyrA\n",
    "#             (\"Cluster7926_\", \"Cluster1277_\"), # mazF3, Rv0959\n",
    "#             (\"Cluster7926_\", \"Cluster8038_\"), # mazF3, rnpA\n",
    "#             (\"Cluster7125_\", \"Cluster893_\"), # gyrA, kdpC\n",
    "#             (\"Cluster893_\", \"Cluster7926_\"),\n",
    "#             (\"Cluster893_\", \"Cluster7125_\"),\n",
    "#             (\"Cluster7125_\", \"Cluster7926_\"),\n",
    "#            (\"Cluster3438_\",\"Cluster1116_\"),\n",
    "#            (\"Cluster1736_\",\"Cluster7125_\")],  # ctaB, katG\n",
    "    \n",
    "    \"MDR\": [(\"Cluster491_\", \"Cluster6096_\"), # rpoB, kdpC\n",
    "#             (\"Cluster6096_\", \"Cluster491_\"),# kdpC, rpoB\n",
    "#            (\"Cluster1404_\", \"Cluster491_\"), # nuoL, rpoB\n",
    "#             (\"Cluster1404_\", \"Cluster2122_\"),# nuoL, etha\n",
    "#             (\"Cluster1237_\", \"Cluster1116_\"), # Rv1877, katG\n",
    "#             (\"Cluster1237_\", \"Cluster491_\"), # Rv1877, rpoB\n",
    "#             (\"Cluster551_\", \"Cluster3005_\"), # Rv1877, cysK2\n",
    "#             (\"Cluster4056_\", \"Cluster7435_\"),\n",
    "#             (\"Cluster1916_\", \"Cluster4056_\"),\n",
    "#             (\"Cluster5665_\", \"Cluster491_\"),\n",
    "            (\"Cluster1736_\",\"Cluster893_\"),\n",
    "           ], \n",
    "    \n",
    "    # \"ethionamide\": [(\"Cluster2698_\", \"Cluster2122_\")], # aspB, ethA\n",
    "    \n",
    "#     \"4-aminosalicylic_acid\": [(\"Cluster2698_\", \"Cluster2211_\"), # aspB, folC\n",
    "#                               (\"Cluster1404_\", \"Cluster2211_\"), # nuoL, folC\n",
    "#                               (\"Cluster1404_\", \"Cluster2698_\"), # nuoL, folC\n",
    "#                             (\"Cluster2698_\", \"Cluster4486_\")], # aspB, thyA\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "clust_alleles = [\"what\"]\n",
    "color_measure = \"oddsRatio\" # \"percent\", \"oddsRatio\", \"pVal\"\n",
    "pVal_test_type = \"chi2\" # or f_classif - only useful if color_measure = \"pVal\"\n",
    "add_zeroCell = 0.5 # Adds this value to each cell in the odds ratio table. Happens when there is a 0\n",
    "\n",
    "for antibiotic_name, cluster_groups in pairwise_table_groups.iteritems():\n",
    "    \n",
    "    # \", \"ethionamide\", \"ofloxacin\", \"amikacin\" \"4-aminosalicylic_acid\"\n",
    "    if antibiotic_name == \"MDR\" or antibiotic_name == \"XDR\":\n",
    "        X1, y1 = get_MDR_training(resistance_data, data_to_plot, antibiotic_name)\n",
    "    else:\n",
    "        y1, X1 = get_target_data(data_to_plot, resistance_data, antibiotic_name)\n",
    "        X1 = remove_key_antibiotic_clusters(X1, antibiotic_name)\n",
    "    \n",
    "    print \"Before removing low variance: \", X1.shape\n",
    "    selector = VarianceThreshold(threshold=.01)\n",
    "    selector.fit_transform(X1)\n",
    "    X1 = X1[X1.columns[selector.get_support()]].copy()\n",
    "    print \"After removing low variance: \", X1.shape\n",
    "    \n",
    "    for (clust1, clust2) in cluster_groups:\n",
    "        \n",
    "        two_clust_df, two_clustPercent_df = get_2allele_resist_dataframes(clust1, clust2, antibiotic_name, \"All\",\n",
    "                                                                              y1,clust_variant_df.loc[y1.index, :], \n",
    "                                                                              clust_variant_df, clust_alleles, False)\n",
    "        if color_measure == \"oddsRatio\":\n",
    "            two_alleleOdds_df = return_2allele_oddsRatio_df(two_clustPercent_df, two_clust_df, \n",
    "                                           clust_variant_df.loc[y1.index, :], y1, add_zeroCell, log_or_no=True)\n",
    "            two_allelePval_df = pd.DataFrame()\n",
    "        elif color_measure == \"pVal\":\n",
    "            two_allelePval_df = return_2allele_Pval_df(pVal_test_type, two_clustPercent_df, two_clust_df, \n",
    "                                                        clust_variant_df.loc[y1.index, :], y1)\n",
    "            two_alleleOdds_df = pd.DataFrame()\n",
    "        \n",
    "        # note that the last input is either pVal or percent depending on how to color the cell.\n",
    "        save_occur_df, labeled_2cluster_df = make_table_figure(clust1, clust2, antibiotic_name, two_clust_df, \n",
    "                                                               two_clustPercent_df, \n",
    "                                                               two_allelePval_df, two_alleleOdds_df,\n",
    "                                                               clust_to_rv, \n",
    "                                                               rv_to_name, clster_manual_map, color_measure,\n",
    "                                                              color_bar_or_not = True)\n",
    "    # print antibiotic_name, cluster_groups"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "----"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Network Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "Corr_stack = sim_all_df_Prod.fillna(0).T.corr().where(np.triu(np.ones(sim_all_df_Prod.fillna(0).T.corr().shape)).astype(np.bool))\n",
    "Corr_stack = Corr_stack.stack().reset_index()\n",
    "Corr_stack.columns=[\"Allele_row\", \"Allele_col\", \"corr\"]\n",
    "print len(Corr_stack[\"Allele_row\"].unique())\n",
    "Corr_stack.head()\n",
    "G = nx.from_pandas_dataframe(Corr_stack, \"Allele_row\", \"Allele_col\", ['corr'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "G.edge"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pos=nx.spring_layout(G)\n",
    "plt.axis('off')\n",
    "nx.draw_networkx_nodes(G,pos,node_size=20)\n",
    "nx.draw_networkx_edges(G,pos,alpha=0.4)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# for col in sim_all_df_T.columns:\n",
    "#     print \"sum: \", sim_all_df_T[col].sum(), \n",
    "#     print \"max: \", sim_all_df_T[col].max(),\n",
    "#     print \"min: \", sim_all_df_T[col].min()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "sim_all_df_T.sort_values(\"svm_weight_sum\", ascending=False)[:top_num]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "rv_to_clust[\"Rv2764c\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "for g in tb_TF_network_dict[\"Rv1287\"]:\n",
    "    try:\n",
    "        print rv_to_name[g]\n",
    "    except:\n",
    "        print \"no name map, \", g\n",
    "        continue"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Look at the reactions in iEK1011"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "model_iEK = cobra.io.load_json_model(\"../Figure scripts/iEK1011_inVivo_media.json\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# pathwayas\n",
    "iEK_genes = [str(x.id) for x in model_iEK.genes]\n",
    "model_gene_subsystems = []\n",
    "for x in model_iEK.genes:\n",
    "    for rxn in x.reactions:\n",
    "        model_gene_subsystems.append(str(rxn.subsystem))\n",
    "allSubsystems_Counter_dict = Counter(model_gene_subsystems)\n",
    "# react_pathways = Counter(important_pathways)\n",
    "svm_gene_pathways = []\n",
    "for g in rv_list:\n",
    "    if g in iEK_genes:\n",
    "        # print g\n",
    "        all_geneRxn_pathways = []\n",
    "        for rxn in model_iEK.genes.get_by_id(g).reactions:\n",
    "            all_geneRxn_pathways.append(str(rxn.subsystem))\n",
    "        svm_gene_pathways.extend(all_geneRxn_pathways)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "for gene in [\"Rv0500\", \"Rv3795\", \"Rv3806c\", \"Rv1563c\"]:\n",
    "    print \"--------\"\n",
    "    for x in model_iEK.genes.get_by_id(gene).reactions:\n",
    "        print x, x.reaction, x.name\n",
    "        \n",
    "for metab in [\"pro__L_c\", \"prpp_c\"]:\n",
    "    print \"--------\"\n",
    "    for x in model_iEK.metabolites.get_by_id(metab).reactions:\n",
    "        if x.id != \"biomass\":\n",
    "            print x, x.name, x.reaction"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 9. Enrichments\n",
    "Write TF enrichment code to see what regulon could possibly be enriched on an SVM hyperplane."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "all_cogs_in_X1 = []\n",
    "for clst in X1.columns:\n",
    "    cluster_name = clst.split(\"_\")[0].replace(\"ter\", \"ter \")\n",
    "    # print test_df_final.loc[cluster_name, \"cog\"]\n",
    "    try:\n",
    "        cog_category = ast.literal_eval(test_df_final.loc[cluster_name, \"cog\"])\n",
    "        all_cogs_in_X1.extend(cog_category)\n",
    "    except:\n",
    "        cog_category = test_df_final.loc[cluster_name, \"cog\"]\n",
    "        all_cogs_in_X1.append(cog_category)\n",
    "\n",
    "\n",
    "rv_significant_cogs = []\n",
    "for clst in sim_all_df_T.index:\n",
    "    cluster_name = clst.split(\"_\")[0].replace(\"ter\", \"ter \")\n",
    "    # print test_df_final.loc[cluster_name, \"cog\"]\n",
    "    try:\n",
    "        cog_category = ast.literal_eval(test_df_final.loc[cluster_name, \"cog\"])\n",
    "        rv_significant_cogs.extend(cog_category)\n",
    "    except:\n",
    "        cog_category = test_df_final.loc[cluster_name, \"cog\"]\n",
    "        rv_significant_cogs.append(cog_category)\n",
    "        \n",
    "#     if cog_category == \"0\":\n",
    "#         print test_df_final.loc[cluster_name, :].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "tf_counts = []\n",
    "for rv in rv_list:\n",
    "    try:\n",
    "        tf_counts.extend(gene_to_TFs_dict[rv])\n",
    "    except:\n",
    "        continue\n",
    "\n",
    "# get pathway stuff ready for TB\n",
    "all_pathways_list = []\n",
    "for x in gene_to_pathway.values():\n",
    "    all_pathways_list.extend(x)\n",
    "all_pathways_COUNTER = Counter(all_pathways_list)\n",
    "\n",
    "pathway_counts = []\n",
    "for g in rv_list:\n",
    "    try:\n",
    "        pathway_counts.extend(gene_to_pathway[g])\n",
    "        # print \"annotation for \", g, gene_to_pathway[g]\n",
    "    except:\n",
    "        continue\n",
    "        # print \"no annotation for \", g\n",
    "        \n",
    "        \n",
    "        \n",
    "pVal_enrich_threshold = 0.005\n",
    "enrich_PATH_dict = get_enrichment_pvals(all_pathways_COUNTER,\n",
    "                                      Counter(pathway_counts), \n",
    "                                      pVal_enrich_threshold)\n",
    "\n",
    "enrich_TF_dict = get_enrichment_pvals(all_TFs_COUNTER,\n",
    "                                      Counter(tf_counts), \n",
    "                                      pVal_enrich_threshold)\n",
    "\n",
    "enrich_model_dict = get_enrichment_pvals(allSubsystems_Counter_dict,\n",
    "                                      Counter(svm_gene_pathways), \n",
    "                                      pVal_enrich_threshold)\n",
    "\n",
    "enrich_COG_dict = get_enrichment_pvals(Counter(all_cogs_in_X1),\n",
    "                                      Counter(rv_significant_cogs), \n",
    "                                      pVal_enrich_threshold)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "for tf_rv, pval in enrich_TF_dict.iteritems():\n",
    "    print tf_rv, \"| p-val:\", pval, \"| name:\", rv_to_name[tf_rv], \"| product:\", clust_to_prod[rv_to_clust[tf_rv]]\n",
    "    print \"| # regulated genes:\", len(tb_TF_network_dict[tf_rv])\n",
    "    \n",
    "    tf_AMR_gene_set = list(set(tb_TF_network_dict[tf_rv]).intersection(set(rv_list)))\n",
    "    print [(x, rv_to_name[x]) for x in tf_AMR_gene_set]\n",
    "    print \"---------\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "for tf, genes in tb_TF_network_dict.iteritems():\n",
    "    if \"Rv3148\" in genes:\n",
    "        print tf, rv_to_name[tf], clust_to_prod[rv_to_clust[tf]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "for (tf_common, count) in Counter(tf_counts).most_common():\n",
    "    print tf_common, rv_to_name[tf_common], clust_to_prod[rv_to_clust[tf_common]], count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_2allele_resist_dataframes(clust1, clust2, drug_name, All_or_OutDf, y1, cv_df_y1, cv_df, clust_alleles, verbose_out):\n",
    "    \n",
    "    double_allele_resist_dict, double_allele_resistPercent_dict = explain_TWOcluster_AMR(clust1, clust2, \n",
    "                                   All_or_OutDf, y1, cv_df_y1, cv_df, clust_alleles, verbose_out)\n",
    "    \n",
    "    clust_1_resist_dict, clust_1_resistPercent_dict = explain_cluster_AMR(clust1,All_or_OutDf, y1, cv_df_y1, \n",
    "                        cv_df, clust_alleles, verbose_out)\n",
    "    clust_1_df = pd.DataFrame.from_dict(clust_1_resist_dict, orient=\"index\")\n",
    "    clust_1_df.columns = [clust1]\n",
    "    clust_1percent_df = pd.DataFrame.from_dict(clust_1_resistPercent_dict, orient=\"index\")\n",
    "    clust_1percent_df.columns = [clust1]\n",
    "\n",
    "    clust_2_resist_dict, clust_2_resistPercent_dict = explain_cluster_AMR(clust2,All_or_OutDf, y1, cv_df_y1, \n",
    "                        cv_df, clust_alleles, verbose_out)\n",
    "    clust_2_df = pd.DataFrame.from_dict(clust_2_resist_dict, orient=\"index\")\n",
    "    clust_2_df.columns = [clust2]\n",
    "    clust_2percent_df = pd.DataFrame.from_dict(clust_2_resistPercent_dict, orient=\"index\")\n",
    "    clust_2percent_df.columns = [clust2]\n",
    "\n",
    "    two_cluster_df = pd.DataFrame.from_dict(double_allele_resist_dict)\n",
    "    two_cluster_df = pd.concat([two_cluster_df, clust_2_df], axis=1)\n",
    "    two_cluster_df = two_cluster_df.append(clust_1_df.T)\n",
    "\n",
    "    two_clusterPercent_df = pd.DataFrame.from_dict(double_allele_resistPercent_dict)\n",
    "    two_clusterPercent_df = pd.concat([two_clusterPercent_df, clust_2percent_df], axis=1)\n",
    "    two_clusterPercent_df = two_clusterPercent_df.append(clust_1percent_df.T)\n",
    "    \n",
    "    output = drug_name+\"__\"+clust1+clust2\n",
    "    print output\n",
    "\n",
    "    writer = pd.ExcelWriter(\"cooccurence_table_excel/\"+output+\".xlsx\")\n",
    "    two_cluster_df.to_excel(writer, sheet_name='resist_fraction', index=True)\n",
    "    two_clusterPercent_df.to_excel(writer, sheet_name='resist_percent', index=True)\n",
    "    writer.save()\n",
    "    return two_cluster_df, two_clusterPercent_df\n",
    "\n",
    "\n",
    "def allele_cooccurence_pValue(all_1, all_2, y_drg, X_all, association_test):\n",
    "    \"\"\"\n",
    "    Takes in two alleles, creates a cooccurence column, and computes the association of that\n",
    "    column with the AMR phenotype.\n",
    "    \"\"\"\n",
    "    two_allele_df = X_all.loc[:, [all_1,all_2]].copy()\n",
    "    two_allele_df[\"cooccurence\"] = two_allele_df[all_1] + two_allele_df[all_2]\n",
    "    two_allele_df[\"cooccurence\"][two_allele_df[\"cooccurence\"]<2]=0\n",
    "    two_allele_df[\"cooccurence\"][two_allele_df[\"cooccurence\"]==2]=1\n",
    "    \n",
    "    if association_test == \"chi2\":\n",
    "        test_stats, pVals = chi2(two_allele_df, y_drg)\n",
    "    elif association_test == \"f_classif\":\n",
    "        test_stats, pVals = f_classif(two_allele_df, y_drg)\n",
    "        \n",
    "    allele_cooccurence_pVal_dict = {}\n",
    "    for col_ind in range(len(two_allele_df.columns)):\n",
    "        allele_cooccurence_pVal_dict.update({two_allele_df.columns[col_ind]: \n",
    "                                             {\"test_stat\": test_stats[col_ind], \"pVal\": pVals[col_ind]}})\n",
    "    \n",
    "    return allele_cooccurence_pVal_dict\n",
    "\n",
    "\n",
    "def return_2allele_Pval_df(assoc_test_type, two_clusterPercent_df, two_clst_df, X_allez, y1):\n",
    "    two_clusterPval_df = two_clusterPercent_df.copy()\n",
    "    for cl1, row in two_clusterPercent_df.iterrows():\n",
    "        for cl2 in row.index: \n",
    "    #         print cl1, cl2\n",
    "    #         print len(cl1.split(\"_\")), len(cl2.split(\"_\")[1])\n",
    "\n",
    "            if cl1.split(\"_\")[0] != cl2.split(\"_\")[0]:\n",
    "                if len(cl1.split(\"_\")[1])==0:\n",
    "                    if len(cl2.split(\"_\")[1])>0:\n",
    "                        allele_pVal_dict = allele_cooccurence_pValue(cl1, cl2, y1, \n",
    "                                                                     X_allez, assoc_test_type)\n",
    "                        two_clusterPval_df.loc[cl1, cl2] = allele_pVal_dict[cl2][\"pVal\"]\n",
    "\n",
    "                elif len(cl2.split(\"_\")[1])==0:\n",
    "                    if len(cl1.split(\"_\")[1])>0:\n",
    "                        allele_pVal_dict = allele_cooccurence_pValue(cl1, cl2, y1, \n",
    "                                                                     X_allez,assoc_test_type)\n",
    "                        two_clusterPval_df.loc[cl1, cl2] = allele_pVal_dict[cl1][\"pVal\"]\n",
    "\n",
    "                else:\n",
    "                    if two_clst_df.loc[cl1, cl2] != \"-\":\n",
    "                        allele_pVal_dict = allele_cooccurence_pValue(cl1, cl2, y1, \n",
    "                                                                     X_allez,assoc_test_type) \n",
    "                        two_clusterPval_df.loc[cl1, cl2] = allele_pVal_dict[\"cooccurence\"][\"pVal\"]\n",
    "\n",
    "            elif cl1.split(\"_\")[0] == cl2.split(\"_\")[0] and len(cl2.split(\"_\")[1])>0:\n",
    "                if assoc_test_type == \"chi2\":\n",
    "                    test_single, pVal_single = chi2(X_allez[[cl2, cl2]], y1)\n",
    "                elif assoc_test_type == \"f_classif\":\n",
    "                    test_single, pVal_single = f_classif(X_allez[[cl2, cl2]], y1)\n",
    "                two_clusterPval_df.loc[cl1, cl2] = pVal_single[0]\n",
    "\n",
    "            elif cl1.split(\"_\")[0] == cl2.split(\"_\")[0] and len(cl1.split(\"_\")[1])>0:\n",
    "                if assoc_test_type == \"chi2\":\n",
    "                    test_single, pVal_single = chi2(X_allez[[cl1, cl1]], y1)\n",
    "                elif assoc_test_type == \"f_classif\":\n",
    "                    test_single, pVal_single = f_classif(X_allez[[cl1, cl1]], y1)\n",
    "                two_clusterPval_df.loc[cl1, cl2] = pVal_single[0]\n",
    "                \n",
    "    return two_clusterPval_df\n",
    "\n",
    "\n",
    "def allele_cooccurence_oddsRatio(all_1, all_2, y_drg, X_all, add_small_value_to_table, log_or_not, single_or_double_tuple):\n",
    "    \"\"\"\n",
    "    Takes in two alleles, creates a cooccurence column, and computes the association of that\n",
    "    column with the AMR phenotype.\n",
    "    \"\"\"\n",
    "    # print X_all.shape, all_1, all_2, X_all.columns\n",
    "    two_allele_df = X_all.loc[:, [all_1,all_2]].copy()\n",
    "    \n",
    "    \n",
    "    if single_or_double_tuple[0] == \"double\":\n",
    "        two_allele_df[\"cooccurence\"] = two_allele_df[all_1] + two_allele_df[all_2]\n",
    "        two_allele_df[\"cooccurence\"][two_allele_df[\"cooccurence\"]<2]=0\n",
    "        two_allele_df[\"cooccurence\"][two_allele_df[\"cooccurence\"]==2]=1\n",
    "    elif single_or_double_tuple[0] == \"single\":\n",
    "        two_allele_df[\"cooccurence\"] = two_allele_df[single_or_double_tuple[1]]\n",
    "    \n",
    "    two_allele_df[\"resistant\"] = y_drg\n",
    "    \n",
    "    cooccurence_resistant = float(len(two_allele_df[(two_allele_df[\"cooccurence\"]==1) & (two_allele_df[\"resistant\"]==1)].index))\n",
    "    cooccurence_susceptible = float(len(two_allele_df[(two_allele_df[\"cooccurence\"]==1) & (two_allele_df[\"resistant\"]==0)].index))\n",
    "    \n",
    "    no_cooccurence_resistant = float(len(two_allele_df[(two_allele_df[\"cooccurence\"]==0) & (two_allele_df[\"resistant\"]==1)].index))\n",
    "    no_cooccurence_susceptible = float(len(two_allele_df[(two_allele_df[\"cooccurence\"]==0) & (two_allele_df[\"resistant\"]==0)].index))\n",
    "    \n",
    "    # add_small_value_to_table = .1\n",
    "    # if one of the cells is 0, there won't be an odds ratio. To get past this, we add a small value of .5 to each value\n",
    "    if cooccurence_resistant==0 or cooccurence_susceptible==0 or no_cooccurence_resistant==0 or no_cooccurence_susceptible==0:\n",
    "        cooccurence_resistant+=add_small_value_to_table\n",
    "        cooccurence_susceptible+=add_small_value_to_table\n",
    "        no_cooccurence_resistant+=add_small_value_to_table\n",
    "        no_cooccurence_susceptible+=add_small_value_to_table\n",
    "    \n",
    "    odds_ratio = (cooccurence_resistant/cooccurence_susceptible)/(no_cooccurence_resistant/no_cooccurence_susceptible)\n",
    "    \n",
    "    if log_or_not == True:\n",
    "        odds_ratio = np.log(odds_ratio)\n",
    "    \n",
    "    return odds_ratio\n",
    "\n",
    "\n",
    "def return_2allele_oddsRatio_df(two_clusterPercent_df, two_clst_df, X_allez, y1, add_smallval_table, log_or_no):\n",
    "    two_clusterOddsRatio_df = two_clusterPercent_df.copy()\n",
    "    for cl1, row in two_clusterPercent_df.iterrows():\n",
    "        for cl2 in row.index: \n",
    "            if cl1.split(\"_\")[0] != cl2.split(\"_\")[0]:\n",
    "            \n",
    "            # ----- case: (clusterX_1, clusterY_1) -----\n",
    "                if two_clst_df.loc[cl1, cl2] != \"-\" and (len(cl1.split(\"_\")[1])>0 or len(cl2.split(\"_\")[1])>0):\n",
    "                    # print cl1, cl2, \"_3\"\n",
    "                    allele_oddsRatio = allele_cooccurence_oddsRatio(cl1, cl2, y1, clust_variant_df.loc[y1.index, :], \n",
    "                                                                    add_smallval_table, log_or_no, (\"double\", \"-\"))\n",
    "                    two_clusterOddsRatio_df.loc[cl1, cl2] = allele_oddsRatio\n",
    "\n",
    "            elif cl1.split(\"_\")[0] == cl2.split(\"_\")[0] and len(cl2.split(\"_\")[1])>0:\n",
    "                \n",
    "                # ----- case: (clusterX_, clusterX_1) -----\n",
    "                # print cl1, cl2, \"_4\"\n",
    "                allele_oddsRatio = allele_cooccurence_oddsRatio(cl1, cl2, y1, clust_variant_df.loc[y1.index, :], \n",
    "                                                                add_smallval_table, log_or_no, (\"single\", cl2))\n",
    "                two_clusterOddsRatio_df.loc[cl1, cl2] = allele_oddsRatio\n",
    "\n",
    "            elif cl1.split(\"_\")[0] == cl2.split(\"_\")[0] and len(cl1.split(\"_\")[1])>0:\n",
    "                \n",
    "                # ----- case: (clusterY_1, clusterY_) -----\n",
    "                # print cl1, cl2, \"_5\"\n",
    "                allele_oddsRatio = allele_cooccurence_oddsRatio(cl1, cl2, y1, clust_variant_df.loc[y1.index, :], \n",
    "                                                                add_smallval_table, log_or_no, (\"single\", cl1))\n",
    "                two_clusterOddsRatio_df.loc[cl1, cl2] = allele_oddsRatio\n",
    "                \n",
    "    return two_clusterOddsRatio_df\n",
    "\n",
    "\n",
    "def make_table_figure(clust1, clust2, drg_name, two_cluster_df, two_clusterPercent_df, \n",
    "                      two_varPval_df, two_varOdds_df, \n",
    "                      clust_to_rv, rv_to_name, clster_manual_map, which_df_2_color):\n",
    "    \"\"\"\n",
    "    Automated process for building tables for the allele co-occurences.\n",
    "    \"\"\"\n",
    "    rv_name1, rv_name2 = clust_to_rv[clust1.strip(\"_\").replace(\"ter\", \"ter \")], clust_to_rv[clust2.strip(\"_\").replace(\"ter\", \"ter \")]\n",
    "    name1, name2 = rv_to_name[rv_name1], rv_to_name[rv_name2]\n",
    "\n",
    "    if name1 == None and rv_name1 in clster_manual_map.keys():\n",
    "        name1 = clster_manual_map[rv_name1]\n",
    "    elif name1 == None:\n",
    "        name1 = rv_name1\n",
    "\n",
    "    if name2 == None and rv_name2 in clster_manual_map.keys():\n",
    "        name2 = clster_manual_map[rv_name2]\n",
    "    elif name2 == None:\n",
    "        name2 = rv_name2\n",
    "        \n",
    "    # ------ Make styled dataframe by changing the index and column names, moving shit around, etc.... ------\n",
    "    if which_df_2_color == \"percent\":\n",
    "        save_occurance_df = two_clusterPercent_df.copy()\n",
    "        vmin_bot=0.2\n",
    "        vmax_top=0.95\n",
    "    elif which_df_2_color == \"pVal\":\n",
    "        save_occurance_df = two_varPval_df.copy()\n",
    "        save_occurance_df = -np.log(save_occurance_df).replace([np.inf, -np.inf], np.nan)\n",
    "        vmin_bot=0\n",
    "        vmax_top=30\n",
    "    elif which_df_2_color == \"oddsRatio\":\n",
    "        save_occurance_df = two_varOdds_df.copy()\n",
    "        abs_max_bound = max(abs(save_occurance_df.fillna(0).values.min()), save_occurance_df.fillna(0).values.max())\n",
    "        vmin_bot = -abs_max_bound\n",
    "        vmax_top = abs_max_bound\n",
    "\n",
    "    new_cols = [str(x.split(\"_\")[1]) if len(str(x.split(\"_\")[1]))>0 else x for x in save_occurance_df.columns]\n",
    "    new_inds = [str(x.split(\"_\")[1]) if len(str(x.split(\"_\")[1]))>0 else x for x in save_occurance_df.index]\n",
    "\n",
    "    new_cols_Name = [str(x.split(\"_\")[1]) if len(str(x.split(\"_\")[1]))>0 else rv_name2 for x in save_occurance_df.columns]\n",
    "    new_inds_Name = [str(x.split(\"_\")[1]) if len(str(x.split(\"_\")[1]))>0 else rv_name1 for x in save_occurance_df.index]\n",
    "\n",
    "    allele_cols = [x for x in new_cols if \"Cluster\" not in x]\n",
    "    # cluster_col = [x for x in new_cols if \"Cluster\" in x]\n",
    "    cluster_col = [rv_name2]\n",
    "\n",
    "    allele_ind = [x for x in new_inds if \"Cluster\" not in x]\n",
    "    # cluster_ind = [x for x in new_inds if \"Cluster\" in x]\n",
    "    cluster_ind = [rv_name1]\n",
    "\n",
    "    # save_occurance_df.columns = new_cols\n",
    "    # save_occurance_df.index = new_inds\n",
    "    save_occurance_df.columns = new_cols_Name\n",
    "    save_occurance_df.index = new_inds_Name\n",
    "\n",
    "    save_occurance_df = save_occurance_df[allele_cols+cluster_col]\n",
    "    save_occurance_df = save_occurance_df.reindex(index = allele_ind+cluster_ind)\n",
    "\n",
    "    labeled_two_cluster_df = two_cluster_df.copy()\n",
    "    # labeled_two_cluster_df.columns = new_cols\n",
    "    # labeled_two_cluster_df.index = new_inds\n",
    "    labeled_two_cluster_df.columns = new_cols_Name\n",
    "    labeled_two_cluster_df.index = new_inds_Name\n",
    "\n",
    "    labeled_two_cluster_df = labeled_two_cluster_df[allele_cols+cluster_col]\n",
    "    labeled_two_cluster_df = labeled_two_cluster_df.reindex(index = allele_ind+cluster_ind)\n",
    "\n",
    "    # ---- Add the two columns and rows to both ends of \"save_occurance_df\"and \"labeled_two_cluster_df\"\n",
    "    # --- Add the two columns to right side of dataframe ----\n",
    "    \n",
    "    resist_vector_name = \"#R\"\n",
    "    total_vector_name = \"Total\"\n",
    "    \n",
    "    labeled_two_cluster_df[resist_vector_name] = labeled_two_cluster_df.fillna(\"-\")[cluster_col[0]].apply(lambda x: x.split(\":\")[0].strip(\"()\").split(\"/\")[0] if \"/\" in x else x)\n",
    "    labeled_two_cluster_df[total_vector_name] = labeled_two_cluster_df.fillna(\"-\")[cluster_col[0]].apply(lambda x: x.split(\":\")[0].strip(\"()\").split(\"/\")[1] if \"/\" in x else x)\n",
    "    \n",
    "    # --- Add the two rows to the bottom of dataframe ----\n",
    "    row_resistant = labeled_two_cluster_df.fillna(\"-\").loc[cluster_ind[0], :].apply(lambda x: x.split(\":\")[0].strip(\"()\").split(\"/\")[0] if \"/\" in x else x)\n",
    "    row_resistant.name = resist_vector_name\n",
    "    row_total = labeled_two_cluster_df.fillna(\"-\").loc[cluster_ind[0], :].apply(lambda x: x.split(\":\")[0].strip(\"()\").split(\"/\")[1] if \"/\" in x else x)\n",
    "    row_total.name = total_vector_name\n",
    "    labeled_two_cluster_df = labeled_two_cluster_df.append(row_resistant)\n",
    "    labeled_two_cluster_df = labeled_two_cluster_df.append(row_total)\n",
    "\n",
    "    # --- Drop Gene ID row ---\n",
    "    labeled_two_cluster_df.drop(cluster_col, axis=1, inplace=True)\n",
    "    labeled_two_cluster_df.drop(cluster_ind, axis=0, inplace=True)\n",
    "    \n",
    "    # ---- Change save_occurance_df -----\n",
    "    save_occurance_df[resist_vector_name] = save_occurance_df[cluster_col[0]]\n",
    "    save_occurance_df[total_vector_name] = save_occurance_df[cluster_col[0]]\n",
    "    row_occurTotal_add = save_occurance_df.loc[cluster_ind[0], :].copy()\n",
    "    row_occurTotal_add.name = total_vector_name\n",
    "    row_occurR_add = save_occurance_df.loc[cluster_ind[0], :].copy()\n",
    "    row_occurR_add.name = resist_vector_name\n",
    "    save_occurance_df = save_occurance_df.append(row_occurR_add)\n",
    "    save_occurance_df = save_occurance_df.append(row_occurTotal_add)\n",
    "    save_occurance_df.drop([cluster_col[0]], axis=1, inplace=True)\n",
    "    save_occurance_df.drop([cluster_ind[0]], inplace=True)\n",
    "\n",
    "    # ------ Make seaborn heatmap from the two dataframes that looks like a table ------\n",
    "    # cm = sns.diverging_palette(220, 10, sep=80, n=7, l=60, as_cmap=True)\n",
    "    cm = sns.diverging_palette(220, 10, sep=1, n=7, as_cmap=True)\n",
    "    # cm = sns.light_palette(\"red\",as_cmap=True, n_colors=5)\n",
    "    # cm = sns.choose_light_palette(input='husl', as_cmap=True)\n",
    "\n",
    "    labels = np.array(labeled_two_cluster_df.fillna(\"-\").astype(str).copy())\n",
    "    only_frac_labels = []\n",
    "    for x in labels:\n",
    "        row_labels = [y.split(\":\")[0].strip(\"()\") for y in x]\n",
    "        row_labels = ['$\\\\dfrac{{{0}}}{{{1}}}$'.format(y.split(\"/\")[0], y.split(\"/\")[1]) if \"/\" in y else y for y in row_labels ]\n",
    "        only_frac_labels.append(row_labels)\n",
    "\n",
    "    # plt.figure(figsize=(len(allele_ind)+1,len(allele_cols)+1))\n",
    "    # plt.tight_layout()\n",
    "    # rowCol_ratio_in_figure = len(save_occurance_df.index)/float(len(save_occurance_df.columns))\n",
    "    plt.figure(figsize=(len(save_occurance_df.columns)*.5, len(save_occurance_df.index)*.5))\n",
    "    g = sns.heatmap(save_occurance_df, annot=np.array(only_frac_labels),fmt = '',\n",
    "                    cmap=cm, vmin=vmin_bot, vmax=vmax_top,\n",
    "                    cbar=False, linewidths=0.6, linecolor=\"black\", \n",
    "                    annot_kws={\"color\":\"black\"}, \n",
    "                    # annot_kws={\"size\": rowCol_ratio_in_figure*3, \"color\":\"black\"},\n",
    "                    square=True);\n",
    "                    # \n",
    "    g.set_xlabel(name1+\" alleles\", rotation=0); # , size=16\n",
    "    g.set_ylabel(name2+\"\\nalleles\", rotation=0); # , size=16\n",
    "    g.xaxis.tick_top()\n",
    "    g.xaxis.set_label_position('top') \n",
    "    g.tick_params(axis=\"both\", which=\"major\", length=0)\n",
    "    plt.setp(g.yaxis.get_majorticklabels(), rotation=0); # , size=10\n",
    "    plt.setp(g.xaxis.get_majorticklabels(), rotation=0); # , size=10\n",
    "    plt.savefig(\"cooccurance_table_figures/\"+drg_name+\"_\"+name1+\"__\"+name2+\"_\"+which_df_2_color+\"_occurance.svg\")\n",
    "    plt.close()\n",
    "    return save_occurance_df, labeled_two_cluster_df\n",
    "# allele_pVal_dict = allele_cooccurence_pValue(\"Cluster551_9\",\"Cluster3176_1\", y1, \n",
    "#                                              clust_variant_df.loc[y1.index, :], \"f_classif\") "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "sim_allDrugs_df = pd.DataFrame()\n",
    "top_num = 60\n",
    "training_size_percents = [.8]\n",
    "score_threshold_test = .80 # .80\n",
    "n_classes = 2\n",
    "simulation_iterations = 200\n",
    "verbose_sim = False\n",
    "\n",
    "fpr, tpr, roc_auc = dict(), dict(), dict()\n",
    "estimator_type = \"SVM\" #logReg or SVM\n",
    "SGD_or_GD = \"SGD\" #SGD or GD\n",
    "\n",
    "# kf = StratifiedKFold(n_splits=2, shuffle=False)\n",
    "\n",
    "antibiotic_clusters = []\n",
    "# Done: Ethambutol, Isoniazid, Pyrazinamide, Rifampicin, Streptomycin, 4-aminosalicylic_acid\n",
    "# [\"isoniazid\", \"ethambutol\", \"rifampicin\", \"pyrazinamide\", \"ethionamide\", \"ofloxacin\", \"4-aminosalicylic_acid\", \"MDR\", \"XDR\"]\n",
    "for drug_name in [\"isoniazid\"]:# [\"isoniazid\", \"ethambutol\", \"rifampicin\", \"pyrazinamide\", \"ethionamide\", \"ofloxacin\", \"4-aminosalicylic_acid\", \"MDR\", \"XDR\"]: # \", \"ethionamide\", \"ofloxacin\", \"amikacin\" \"4-aminosalicylic_acid\"\n",
    "    \n",
    "    if drug_name == \"MDR\" or drug_name == \"XDR\":\n",
    "        X1, y1 = get_MDR_training(resistance_data, data_to_plot, drug_name)\n",
    "    else:\n",
    "        y1, X1 = get_target_data(data_to_plot, resistance_data, drug_name)\n",
    "        X1 = remove_key_antibiotic_clusters(X1, drug_name)\n",
    "    \n",
    "    print \"Before removing low variance: \", X1.shape\n",
    "    selector = VarianceThreshold(threshold=.01)\n",
    "    selector.fit_transform(X1)\n",
    "    X1 = X1[X1.columns[selector.get_support()]].copy()\n",
    "    print \"After removing low variance: \", X1.shape\n",
    "    \n",
    "    # threshold_range = threshold_range_dict[drug_name]\n",
    "\n",
    "    sim_all_df = pd.DataFrame()\n",
    "\n",
    "    for t in tqdm(range(simulation_iterations)):\n",
    "\n",
    "        # randomly choose the threshold value.\n",
    "        # threshold_val = random.randint(threshold_range[0], threshold_range[1])\n",
    "        if estimator_type == \"SVM\":\n",
    "            if SGD_or_GD == \"SGD\":\n",
    "                clf = SGDClassifier(loss=\"hinge\", penalty=\"l1\", shuffle=True, class_weight=\"balanced\" , alpha=0.0001) \n",
    "                if drug_name==\"ethionamide\":\n",
    "                    threshold_val = 160\n",
    "                elif drug_name==\"XDR\":\n",
    "                    threshold_val = 230\n",
    "                elif drug_name==\"capreomycin\":\n",
    "                    threshold_val = 190\n",
    "                elif drug_name==\"moxifloxacin\":\n",
    "                    threshold_val = 190\n",
    "                elif drug_name==\"ciprofloxacin\":\n",
    "                    threshold_val = 140\n",
    "                elif drug_name==\"rifabutin\":\n",
    "                    threshold_val = 100\n",
    "                else:\n",
    "                    threshold_val = 170\n",
    "            elif SGD_or_GD == \"GD\":\n",
    "                clf = LinearSVC(penalty = \"l1\", loss=\"squared_hinge\", dual=False, class_weight=\"balanced\")\n",
    "                threshold_val = .3\n",
    "        elif estimator_type == \"logReg\":\n",
    "            if SGD_or_GD == \"SGD\":\n",
    "                clf = SGDClassifier(loss=\"log\", penalty=\"l1\", shuffle=True, class_weight=\"balanced\" , alpha=0.0001) \n",
    "                threshold_val = 190\n",
    "            elif SGD_or_GD == \"GD\":\n",
    "                clf = LogisticRegression(penalty = \"l1\", class_weight=\"balanced\")\n",
    "                if drug_name==\"ethionamide\":\n",
    "                    threshold_val = .9\n",
    "                elif drug_name==\"cycloserine\":\n",
    "                    threshold_val = .4\n",
    "                elif drug_name==\"rifabutin\":\n",
    "                    threshold_val = .4\n",
    "                else:\n",
    "                    threshold_val = .7    \n",
    "    \n",
    "                    \n",
    "        percent_of_train = random.choice(training_size_percents)\n",
    "        train_x, X_test, train_y, y_test = train_test_split(X1, y1, \n",
    "                                                            train_size = percent_of_train, \n",
    "                                                            stratify=None) # stratify=y1\n",
    "    \n",
    "        # train_x, train_y = resample(X1, y1, n_samples=int(percent_of_train*len(X1.index)))\n",
    "\n",
    "        # feature selection w/ parameter \"threshold\"\n",
    "        feat_selection = SelectFromModel(clf, threshold=threshold_val)\n",
    "\n",
    "        # Setup pipeline for SGD-SVM feature selection\n",
    "        model = Pipeline([\n",
    "                  ('fs', feat_selection),  # SelectFromModel using SGD-SVM\n",
    "                  ('clf', clf), \n",
    "                ])\n",
    "        \n",
    "        # model.fit(X1, y1)\n",
    "        model.fit(train_x, train_y)\n",
    "\n",
    "        sim_t_df = pd.DataFrame([model.named_steps[\"clf\"].coef_[0]], \n",
    "                                columns=X1.columns[model.named_steps[\"fs\"].get_support()])\n",
    "        \n",
    "        model_score_on_test = model.score(X_test, y_test)\n",
    "        \n",
    "        if verbose_sim == True:\n",
    "            print \"\",\n",
    "            print \"# of features in simulation \"+str(t)+ \":\"+ str(len(sim_t_df.columns)),\n",
    "            print \"#R: \" + str(train_y.sum()) + \"#S: \" + str(len(train_y.index)-train_y.sum()),\n",
    "            print \"% of samples: \" + str(percent_of_train),\n",
    "            print \"Prediction Score: \", model_score_on_test\n",
    "        \n",
    "        # if model_score_on_test > score_threshold_test:\n",
    "        sim_all_df = pd.concat([sim_all_df, sim_t_df], ignore_index=True)\n",
    "        \n",
    "        # y_score = model.predict(X_test)\n",
    "        y_score = model.decision_function(X_test)\n",
    "\n",
    "        fpr[t], tpr[t], _ = roc_curve(y_test.values, y_score)\n",
    "        roc_auc[t] = auc(fpr[t], tpr[t])\n",
    "\n",
    "    \n",
    "    print \"# of cluster features...\", len(list(set(sim_all_df.columns)))\n",
    "\n",
    "    sim_all_df_T = sim_all_df.transpose().copy()\n",
    "    sim_all_df_T[\"svm_weight_sum\"] = sim_all_df_T.apply(lambda x: abs(x).sum(), axis=1)\n",
    "    sim_all_df_T_top = sim_all_df_T.sort_values(\"svm_weight_sum\", ascending=False)[:top_num]\n",
    "    \n",
    "    sim_allDrugs_df = pd.concat([sim_allDrugs_df, sim_all_df[sim_all_df_T_top.index]])\n",
    "\n",
    "    amr_clsts = []\n",
    "    for x in list(sim_all_df_T_top.index):\n",
    "        amr_clsts.append(str(x).split(\"_\")[0].replace(\"ter\", \"ter \"))\n",
    "\n",
    "    amr_clsts_set = list(set(amr_clsts))\n",
    "    antibiotic_clusters = antibiotic_clusters+amr_clsts_set\n",
    "\n",
    "    \n",
    "    clster_var_mapping_prod, clster_var_mapping_Rv = {}, {}\n",
    "    for i in sim_all_df_T.index:\n",
    "        var_index = i.split(\"_\")[1]\n",
    "        # print i\n",
    "        # clster_var_mapping_prod.update({i: clust_to_prod[str(i).split(\"_\")[0].replace(\"ter\", \"ter \")]+var_index})\n",
    "\n",
    "        try:\n",
    "            clster_var_mapping_prod.update({i: i+clust_to_prod[str(i).split(\"_\")[0].replace(\"ter\", \"ter \")]+\"_\"+var_index})\n",
    "            # clster_var_mapping_Rv.update({i: clust_to_rv[str(i).split(\"_\")[0].replace(\"ter\", \"ter \")]+var_index})\n",
    "        except:\n",
    "            clster_var_mapping_prod.update({i: i+clust_to_prod[str(i).split(\"_\")[0].replace(\"ter\", \"ter \")]+\"_\"+var_index})\n",
    "            # clster_var_mapping_Rv.update({i: clust_to_prod[str(i).split(\"_\")[0].replace(\"ter\", \"ter \")]+var_index})\n",
    "        \n",
    "    clster_to_IDENTIFIER, rv_list = clster_identifier_func(clster_manual_map, sim_all_df_T, \n",
    "                                                       clust_to_rv, rv_to_name, clust_to_prod)\n",
    "    \n",
    "    save_ROC_curve_figure(drug_name, fpr, tpr, roc_auc, simulation_iterations, estimator_type)\n",
    "    \n",
    "    sim_all_df_Prod, allele_to_colormap, clust_alleles = svm_top_weights_colormap(sim_all_df_T, clster_to_IDENTIFIER, \n",
    "                                                                              top_num, X1, y1, \n",
    "                                                                              clster_var_mapping_prod)\n",
    "    \n",
    "    sim_all_df_Prod.to_csv(\"SVM_dataframes/\"+drug_name+\"_\"+estimator_type+\"_\"+SGD_or_GD+\"_top_clusts.csv\")\n",
    "    sim_all_df_T.sort_values([\"svm_weight_sum\"], ascending=False)[:100].to_csv(\"SVM_dataframes/\"+drug_name+\"_\"+estimator_type+\"_\"+SGD_or_GD+\"_top_100alleles.csv\")\n",
    "    \n",
    "    sim_corr_df = sim_all_df_Prod.fillna(0).T.corr().copy()\n",
    "    sim_corr_df = sim_corr_df[abs(sim_corr_df)>.05].fillna(0)\n",
    "\n",
    "    g_plot = plot_svm_heatmap(sim_corr_df, (15, 15), allele_to_colormap, \"seismic\") # \"seismic\"\n",
    "    g_plot.savefig(\"SVM_figures/\"+drug_name+\"_\"+estimator_type+\"_\"+SGD_or_GD+\"_correlation.png\")\n",
    "    \n",
    "    sim_corr_df.to_csv(\"SVM_dataframes/\"+drug_name+\"_\"+estimator_type+\"_\"+SGD_or_GD+\"_correlation_matrix.csv\")\n",
    "\n",
    "    g = sns.clustermap(sim_all_df_Prod.fillna(0), \n",
    "                       col_cluster=False, row_cluster=True, cmap=\"seismic\",\n",
    "                      figsize=(18,14), xticklabels=False)\n",
    "    plt.setp(g.ax_heatmap.yaxis.get_majorticklabels(), rotation=0);\n",
    "    plt.setp(g.ax_heatmap.yaxis.get_majorticklabels(), rotation=0);\n",
    "    g.savefig(\"SVM_figures/\"+drug_name+\"_\"+estimator_type+\"_\"+SGD_or_GD+\"_iterations.png\")\n",
    "    \n",
    "    general_cluster_alleles = list(set([clst_al.split(\"_\")[0].replace(\"ter\", \"ter \") for clst_al in clust_alleles]))\n",
    "\n",
    "    GEMPro_final = get_GEMPRO_df(general_cluster_alleles, clust_to_rv, rv_to_name, clster_manual_map, clust_variant_df)\n",
    "    GEMPro_final.to_csv(\"GEMPro_seq_dataframes/\"+drug_name+\"_\"+estimator_type+\"_\"+SGD_or_GD+\"_\"+str(top_num)+\"_SEQdf.csv\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
